---
title: "Requested Gene Distribution in Single Cells of Pyelonephrtis"
author: "Xin Wang"
date: "2024-06-17"
description: 
output: html_document
---
# Loading the packages
```{r}
# check the session information
sessionInfo()
library(dplyr)
library(Seurat)
library(patchwork)

#install.packages('devtools') #assuming it is not already installed

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
library(devtools)
#install_github('andreacirilloac/updateR')
# install.packages("sctransform")
# install.packages("ggthemes")
# loading the libarary
library(updateR)

library("sctransform")
library("ggthemes")
library(Seurat)

# BiocManager::install("limma")
# BiocManager::install("ComplexHeatmap")
# BiocManager::install("biomaRt")
# BiocManager::install("methrix")BiocManager::install("org.Hs.eg.db")
# devtools::install_github("saeyslab/nichenetr")
library(limma)
library(nichenetr)
library(tidyverse)
library(dplyr)

library(biomaRt)
library("data.table")


library(umap)
library(patchwork)
library(cowplot)
library(installr)

#install.packages("Matrix", repos = "http://cran.r-project.org")
### we used discre color palettes from ggsci

library("ggsci")
library("ggplot2")
library("gridExtra")

#options(buildtools.check = function(action) TRUE )
#devtools::install_url('https://cran.r-project.org/src/contrib/Archive/Matrix.utils/Matrix.utils_0.9.7.tar.gz')
#devtools::install_github('cole-trapnell-lab/monocle3')
#library(monocle3)

library(harmony)
library(Seurat)
#library(SeuratData)
library(tidyverse)

#library(BPCells)
library(ggplot2)
library("devtools")
library("AnnotationDbi")
library("org.Hs.eg.db")
#library(tibble)
#library(future)
library(here)
library(patchwork)
library(future)
library(monocle3)
#devtools::install_dev("remotes")
#remotes::install_github('chris-mcginnis-ucsf/DoubletFinder')
#remotes::install_github(repo ='chris-mcginnis-ucsf/DoubletFinder' )
#install.packages("DoubletFinder")
library(DoubletFinder)
library(future.callr)
library(ComplexHeatmap)

library(RColorBrewer)
library(circlize)
library(monocle3)
#install.packages("magick")
library(magick)
#K means with 6 groups

```


## Step 2: Set up the workspace environment:
```{r}
#rm(list=ls())
Indir =c("/Users/XXW004/Documents/Projects/RuizRosado/PIPseq/Results/Evaluation_V022723/CellEvaluation_Subclustering_V0803/PT_Subclusters/OutputRDS/")
Outdir =c("/Users/XXW004/Documents/Projects/RuizRosado/SpatialTranscriptomic/Results/BioMarkers_V08072025/")
# set up the color scale:
myPalette <- colorRampPalette(rev(brewer.pal(11, "Spectral")))

# we already generate the list and will ignore the transfer 
# read the mouse and human homologous genes
mh_data<- read.csv("/Users/XXW004/Documents/Projects/Databases/HumanMouseTransfer/HOM_MouseHuman_Symbol.csv")
mh_data %>% head()

```


# generate the health cluster of PT cells

```{r}
setwd(Outdir)
# Read the preclustered object

#scrna <- readRDS(file = paste0(Indir,"Pyelonephritis_singlecell_harmony_withAnnotation_v0803.RDS"))
# single cell directory:
Scindir = "/Users/XXW004/Documents/Projects/RuizRosado/PIPseq/Results/Evaluation_V022723/CellEvaluation_Subclustering_V0803/FinalSingleCellObj_V08012024/"

# immune single cell directory:
ImmScindir = "/Users/XXW004/Documents/Projects/RuizRosado/SpatialTranscriptomic/Results/Deconvolution_Six_V07022025/"

# spatial transcript directory:
Spindir ="/Users/XXW004/Documents/Projects/RuizRosado/SpatialTranscriptomic/Results/Deconvolution_Six_V07022025/"

# Output of each pathways and 
Interestedpathwaysdir = "/Users/XXW004/Documents/Projects/BrianBecknell/UrinePyelonephritis/Results/Result_v0715_newsets/InterestedGOTerms/"
# read the single cell
SingleCellObj <- readRDS(file = paste0(Scindir,"PIPSeq_Final_Annotated_v08012024.rds"))

# read the immune cells
ImmuneSingleCellObj<- readRDS(file = paste0(ImmScindir,"PIPSeq_Final_CD45ImmuneCells_UsedForScanpy.RDS"))

# read the spatial 
SpatialObj <- readRDS(file = paste0(Spindir,"Pyelonephritis_Infection_ATLAS_RCTD_Deco_UsingEpiStroImm.rds"))
DefaultAssay(SpatialObj) <- "SCT"
```

# chagne the image background
```{r}
#BiocManager::install("EBImage")
#BiocManager::install("ggimage")
library(ggimage)
library(EBImage) # for image manipulation
names(SpatialObj@images)
# Access the image
set_spatial_background_white_all_images <- function(seurat_obj, threshold = 0.05) {
  
  for (img_name in names(seurat_obj@images)) {
    message("Processing image: ", img_name)

    img <- seurat_obj@images[[img_name]]@image
    img_norm <- normalize(img)
    
    if (length(dim(img_norm)) == 2) {
      img_rgb <- array(rep(img_norm, each = 3), dim = c(dim(img_norm), 3))
    } else {
      img_rgb <- img_norm
    }
    
    background_mask <- apply(img_rgb, c(1, 2), function(x) all(x < threshold))
    
    img_rgb[,,1][background_mask] <- 1
    img_rgb[,,2][background_mask] <- 1
    img_rgb[,,3][background_mask] <- 1
    
    seurat_obj@images[[img_name]]@image <- img_rgb
  }
  
  return(seurat_obj)
}

# Apply to all images in a single Seurat object
SpatialObj <- set_spatial_background_white_all_images(SpatialObj)
set_tissue_areas_light_blue <- function(seurat_obj, threshold = 0.05, color = c(0.678, 0.847, 0.902)) {
  for (img_name in names(seurat_obj@images)) {
    img <- seurat_obj@images[[img_name]]@image
    img_norm <- normalize(img)

    # Convert to RGB if image is grayscale
    if (length(dim(img_norm)) == 2) {
      img_rgb <- array(rep(img_norm, each = 3), dim = c(dim(img_norm), 3))
    } else {
      img_rgb <- img_norm
    }

    # Create a mask for tissue (areas that are not background)
    tissue_mask <- apply(img_rgb, c(1, 2), function(x) any(x > threshold))
    
    # Set tissue areas to light blue, leave the background unchanged
    img_rgb[tissue_mask, 1] <- color[1]
    img_rgb[tissue_mask, 2] <- color[2]
    img_rgb[tissue_mask, 3] <- color[3]

    # Assign modified image back
    seurat_obj@images[[img_name]]@image <- img_rgb
  }
  return(seurat_obj)
}
# Light blue tint
SpatialObj <- set_tissue_areas_light_blue(SpatialObj)

```

# Generate the DotPlot for the stromal, epithelial cell and immune cell across the spatial, single cell including the immune cells
```{r}
#1. The kidney immune profiling

CellCategories<- c("Immune Cells")
RequestedGenes3<- c("Ptprc", "Cd3e", "Cd4", "Cd8a", "Foxp3", "Il2ra", "Cd19", "Cd79a", "Ms4a1", "Prdm1", "Ccr2", "Cx3cr1", "Adgre1", "Mrc1", "Arg1", "Nos2",
"Ly6g", "S100a8", "S100a9", "Elane", "Ncr1", "Gzmb", "Prf1")

# 2. The kidney stromal cells
CellCategories<- c("Stromal Cells")
RequestedGenes2<- c("Pdgfra", "Pdgfrb", "Col1a1", "Col3a1","Dcn", "Postn", "Acta2", "Cspg4", "Rgs5", "Pecam1", "Cdh5", "Lyve1", "Prox1", "Itga8", "Cxcl12", "Myh11", "Tagln")
# 3. The kidney epithelial cells
CellCategories<- c("Epithelial Cells")
RequestedGenes1 <- c("Lrp2","Slc34a1","Slc5a12","Slc13a3","Slc7a12","Pdpn","Cldn1","Aqp1","Trpv5","Umod", "Slc12a1","Slc12a3", "Pvalb","Aqp2", "Aqp3", "Atp6v1b1", "Slc4a1","Slc26a4", "Slc4a9","Krt5","Upk1b","Nphs1","Wt1")

length(RequestedGenes1)
length(RequestedGenes2)
length(RequestedGenes3)
# Generate the DotPlots in spatial

P1<-DotPlot(SpatialObj, features =  RequestedGenes1,col.min = -2, col.max = 3,cols = "Spectral") +RotatedAxis() + NoLegend()+ RotatedAxis()+  theme(axis.title.x = element_blank())
P2<-DotPlot(SpatialObj, features =  RequestedGenes2,col.min = -2, col.max = 3,cols = "Spectral") +RotatedAxis()  + NoLegend() + RotatedAxis()  + theme(
    axis.text.y = element_blank(),
    axis.title.x = element_blank(),
    #axis.ticks.y = element_blank(),
    axis.title.y = element_blank()
  )

P3<-DotPlot(SpatialObj, features =  RequestedGenes3,col.min = -2, col.max = 3,cols = "Spectral") +RotatedAxis()+ ylab(NULL) +guides(color = guide_colorbar(title = 'AverExp'),
         size = guide_legend(title = "PercExp")) + RotatedAxis() +theme(
    axis.text.y = element_blank(),
    axis.title.x = element_blank(),
    #axis.ticks.y = element_blank(),
    axis.title.y = element_blank()
  )

pdf("SpatialDotPlot_BioMarkerDistribution.pdf", height = 5, width = 18)
P1 + P2 +P3 + plot_layout(widths = c(1.4, 0.9, 1.2))
dev.off()

# The overall marker expression across single cells and spatial at different timepoints
# we set up identity to date
SpatialObj@meta.data $group %>% table()

Idents(SpatialObj) <- SpatialObj@meta.data $group
P3<-DotPlot(SpatialObj, features =  RequestedGenes1,col.min = -2, col.max = 3,cols = "Spectral") +RotatedAxis() + NoLegend()+ RotatedAxis()+  theme(axis.title.x = element_blank())
P4<-DotPlot(SpatialObj, features =  RequestedGenes2,col.min = -2, col.max = 3,cols = "Spectral") +RotatedAxis()  + NoLegend() + RotatedAxis()  + theme(
    axis.text.y = element_blank(),
    axis.title.x = element_blank(),
    #axis.ticks.y = element_blank(),
    axis.title.y = element_blank()
  )

P5<-DotPlot(SpatialObj, features =  RequestedGenes3,col.min = -2, col.max = 3,cols = "Spectral") +RotatedAxis()+ ylab(NULL) +guides(color = guide_colorbar(title = 'AverExp'),
         size = guide_legend(title = "PercExp")) + RotatedAxis() +theme(
    axis.text.y = element_blank(),
    axis.title.x = element_blank(),
    #axis.ticks.y = element_blank(),
    axis.title.y = element_blank()
  )

pdf("SpatialDotPlot_BioMarkerDistribution_infecteddate.pdf", height = 5, width = 18)
P3 + P4 +P5 + plot_layout(widths = c(1.4, 0.9, 1.2))
dev.off()

```

# create a function to buld the spatial distribution of each genes and create folder for each types
```{r}


SpatialColors <- colorRampPalette(colors = rev(x = brewer.pal(n = 11, name = "Spectral")))
myPalette <- colorRampPalette(rev(brewer.pal(11, "Spectral")))
PlotSpatialAndSingleCellFeatures <-
  function(RequestedGenes,
           # the list contain cell types and their correpsonding genes
           SpatialObj,
           # spatial object
           SingleCellObj,
           # single cell object
           ImmuneSingleCellObj ,
           # immune single cell object
           RequestedDir,
           SpatialColors) {
    for (i in names(RequestedGenes)) {
      # create folder to store the categories
      #i="PanTCell"
      Catrequestedir <- paste0(RequestedDir, i,"/")
      dir.create(Catrequestedir)
      setwd(Catrequestedir)
     base_width_spatial <- max(4, length(RequestedGenes[[i]]))        # spatial plots scale with gene count
    base_width_singlecell <- max(4, length(RequestedGenes[[i]]))   # single cell plot width
      # generate the dotplots for  spatial 
      SpatialDotPltCateGenes <-
        DotPlot(
          SpatialObj,
          features =  RequestedGenes[[i]],
          col.min = -2,
          col.max = 3,
          cols = "Spectral"
        ) + RotatedAxis() + ylab(NULL) + guides(color = guide_colorbar(title = 'AverExp'),
                                                size = guide_legend(title = "PercExp")) + RotatedAxis()

  ## 
      ggsave(paste0(Catrequestedir,i,"_MarkerSpatial_DotPlot.pdf"), plot =SpatialDotPltCateGenes, height = 6, width = base_width_spatial )
       # generate the dotplots for  single cell 
        SinglecellDotPltCateGenes <-
        DotPlot(
          SingleCellObj,
          features =  RequestedGenes[[i]],
          col.min = -2,
          col.max = 3,
          cols = "Spectral"
        ) + RotatedAxis() + ylab(NULL) + guides(color = guide_colorbar(title = 'AverExp'),
                                                size = guide_legend(title = "PercExp")) + RotatedAxis()

  
      ggsave(paste0(Catrequestedir,i,"_MarkerSingleCell_DotPlot.pdf"), plot =SinglecellDotPltCateGenes, height = 9, width = base_width_singlecell )
      
      # generate the dotplots for  single cell 
        ImmSinglecellDotPltCateGenes <-
        DotPlot(
          ImmuneSingleCellObj,
          features =  RequestedGenes[[i]],
          col.min = -2,
          col.max = 3,
          cols = "Spectral"
        ) + RotatedAxis() + ylab(NULL) + guides(color = guide_colorbar(title = 'AverExp'),
                                                size = guide_legend(title = "PercExp")) + RotatedAxis()

  
      ggsave(paste0(Catrequestedir,i,"_MarkerImmuneSingleCell_DotPlot.pdf"), plot =ImmSinglecellDotPltCateGenes, height = 4, width = base_width_singlecell )
      
      # create genes distribution from
      for (j in RequestedGenes[[i]]) {
        #---- Spatial Plot ----#
        skip_to_next <- FALSE
        tryCatch(
          print(SpatialObj@assays$SCT@data[j, ]),
          error = function(e) {
            skip_to_next <<- TRUE
          }
        )
        if (skip_to_next) {
          next
        }
        Maxvalue <- ceiling(max(SpatialObj@assays$SCT@data[j, ]))
        Minvalue <- floor(min(SpatialObj@assays$SCT@data[j, ]))
        GenesInRequestedMouseSpatialFeature <-
          SpatialFeaturePlot(
            SpatialObj,
            features = j,
            pt.size.factor = 1500,
            #alpha = c(0.8, 1),
            crop = TRUE
          ) &
          #ggplot2::scale_fill_gradientn(colours =  SpatialColors(n = 60), limits = c(0, 0.6))
          ggplot2::scale_fill_gradientn(colours = myPalette(4),
                                        limits = c(Minvalue, Maxvalue))
        ggsave(
          filename = paste0(Catrequestedir, "/", "Spatial_", j, "_featureplot.pdf"),
          plot = GenesInRequestedMouseSpatialFeature,
          height = 3,
          width = 15
        )
        #---- Single Cell Plot ----#
        skip_to_next_sc <- FALSE
        
        tryCatch(
          print(SingleCellObj@assays$SCT@data[j, ]),
          error = function(e) {
            skip_to_next_sc <<- TRUE
          }
        )
        if (skip_to_next_sc) {
          next
        }
        
        ScMaxvalue <-
          ceiling(max(SingleCellObj@assays$SCT@data[j, ]))
        ScMinvalue <- floor(min(SingleCellObj@assays$SCT@data[j, ]))
        
        GenesInGOtermMouseSingleCellFeature <-
          FeaturePlot(object = SingleCellObj,
                      features = j,
                      split.by = "DataSet")
        
        ggsave(
          filename = paste0(Catrequestedir, "/", "SingleCell_", j, "_feature.pdf"),
          plot = GenesInGOtermMouseSingleCellFeature,
          height = 3,
          width = 12
        )
        
        #---- Immune Single Cell Plot ----#
        skip_to_next_imsc <- FALSE
        
        tryCatch(
          print(GetAssayData(ImmuneSingleCellObj, assay = "RNA", slot = "data")[j, ]),
          error = function(e) {
            skip_to_next_imsc <<- TRUE
          }
        )
        
        if (skip_to_next_imsc) {
          next
        }
        ImmuneSingleCellObj
        ScMaxvalue <-
          ceiling(max( GetAssayData(ImmuneSingleCellObj, assay = "RNA", slot = "data")[j, ]))
        ScMinvalue <-
          floor(min(GetAssayData(ImmuneSingleCellObj, assay = "RNA", slot = "data")[j, ]))
        
        GenesInGOtermMouseImmSingleCellFeature <-
          FeaturePlot(object = ImmuneSingleCellObj,
                      features = j,
                      split.by = "orig.ident")
        
        ggsave(
          filename = paste0(
            Catrequestedir,
            "/",
            "ImmuneSingleCell_",
            j,
            "_feature.pdf"
          ),
          plot = GenesInGOtermMouseImmSingleCellFeature,
          height = 3,
          width = 12
        )
      }
    }
  }
```


# requested Genes put them into the folder
```{r}
# check the monocyte, macrophage, macrophage genes:
#RequestedGenes<- c("Ly6c2","Itgm","Ccr2","C1qa","Cd68","Adgre1","Csf1r","S100a8","Ly6g")
# 
# # check the ligand and receptors found in the ATLAS
# RequestedGenes<- c("C3","Ccl3","Ccl3","Ccl4","Ccl5","Ccl6","Ccl7","Ccl9","Ccrl2","Cxcl1","Cxcl10","Cxcl14","Cxcl16","Cxcl3","Cxcl5","Cxcr2")
# 
# # Fig S2 for requested of tissue-specific gene expression markers (identities) for the renal urothelium (Upk3a), pelvis (Lc4a11 and Nfat5), medulla (Slc3a1 and Prlr), and cortex (Renin and Nephrin) on slides collected from infected kidney
# RequestedGenes<- c("Upk3a","Psca","Slc4a11","Nfat5","Slc3a1","Prlr","Ren1","Nphs1")
# # 

### The kidney immune profiling:
# RequestedGenes<- c("Ptprc", "Cd3e", "Cd4", "Cd8a", "Foxp3", "Il2ra", "Cd19", "Cd79a", "Ms4a1", "Prdm1",
# "Ly6c2", "Ccr2", "Cx3cr1", "Adgre1", "Mrc1", "Arg1", "Nos2", "Cd11b", "Cd11c",
# "Ly6g", "S100a8", "S100a9", "Elane", "Ncr1", "Klrb1c", "Gzmb", "Prf1")

ImmuneGenes <- list()
ImmuneGenes[["PanTCell"]] <- c("Cd3e", "Cd3d", "Cd3g", "Trac")
ImmuneGenes[["CD4T"]] <- c("Cd4")
ImmuneGenes[["CD8T"]] <- c("Cd8a", "Cd8b1")
ImmuneGenes[["Tregs"]] <-c("Cd4", "Il2ra", "Foxp3", "Ctla4", "Ikzf2")
ImmuneGenes[["PanTCell"]] <- c("Cd3e", "Cd3d", "Cd3g", "Trac")
ImmuneGenes[["Th1"]] <-c("Tbx21", "Ifng")
ImmuneGenes[["Th2"]] <-c("Gata3", "Il4", "Il5", "Il13")
ImmuneGenes[["Th17"]] <- c("Rorc", "Il17a", "Il17f")
ImmuneGenes[["Tfh"]] <- c("Bcl6", "Cxcr5", "Il21")
ImmuneGenes[["ExhaustedT"]] <- c("Pdcd1", "Lag3", "Havcr2" , "Tigit")
ImmuneGenes[["Trm"]] <- c("Itgae", "Cd69", "Cxcr6")


# B cells
ImmuneGenes[["PanB"]] <-c("Cd19", "Ms4a1", "Cd79a", "Cd79b")
ImmuneGenes[["MatureB"]] <-c("Cd19", "Cd23", "Cd21")
ImmuneGenes[["PlasmaCell"]] <-c("Prdm1", "Xbp1", "Sdc1")
ImmuneGenes[["GeminalCenterB"]] <-c("Bcl6", "Aicda")

# Monocyte 
ImmuneGenes[["Monocyte"]] <-c("Cd11b", "Ly6c2", "Ccr2", "Fcgr3")
ImmuneGenes[["ResidentMacrophages"]] <-c("Adgre1", "Cd11b", "Cd64", "Mertk")
ImmuneGenes[["InfiltratingMacrophages"]] <-c("Ly6c", "Ccr2", "Cd11b")
ImmuneGenes[["M1like"]] <-c("Nos2", "Il1b", "Tnf")
ImmuneGenes[["M2like"]] <-c("Mrc1", "Arg1", "Chil3")

# neutrophil
ImmuneGenes[["Neutrophil"]] <-c("Ly6g", "S100a8", "S100a9", "Mpo", "Elane", "Cxcr2")

# Dendritic
ImmuneGenes[["cDC1"]] <-c("Xcr1", "Batf3", "Clec9a", "Cd8a")
ImmuneGenes[["cDC2"]] <-c("Cd11b", "Irf4", "Sirpa")
ImmuneGenes[["pDCs"]] <-c("Siglech", "Bst2", "Tcf4")

# NK cell
ImmuneGenes[["NK"]] <-c("Ncr1", "Klrb1c", "Gzmb", "Prf1")


# Innate lymphoid cells
ImmuneGenes[["ILC1"]] <-c("Tbx21", "Ifng", "Ncr1")
ImmuneGenes[["ILC2"]] <-c("Gata3", "Il5", "Il13", "Rora")
ImmuneGenes[["ILC3"]] <-c("Rorc", "Il17a", "Il22")
# Eosinophil
ImmuneGenes[["Eosinophil"]] <-c("Siglecf", "Gata1", "Prg2")

# create a function to generate the spatial, single cell for these immune cells
# setup the directory
ImmunDir<- paste0(Outdir,"ImmuneCells/")
dir.create(ImmunDir)
setwd(ImmunDir)
# run the function
PlotSpatialAndSingleCellFeatures(ImmuneGenes,SpatialObj, SingleCellObj, ImmuneSingleCellObj , ImmunDir, SpatialColors )

# Stroma cells


# RequestedGenes<- c("Pdgfra", "Pdgfrb", "Col1a1", "Col3a1", "Col6a1", "Dcn", "Lum", "Postn", "Acta2", "Gpx3", "Tcf21", "Meis1", "Rgs5", "Cspg4", "Des", "Pecam1", "Cdh5", "Kdr", "Emcn", "Plvap", "Lyve1", "Prox1", "Pdpn", "Nr2f2", "Efnb2", "Gja5", "Itga8", "Cxcl12", "Gata3", "Myh11", "Tagln", "Cnn1", "Notch3")

# shorten version:
# RequestedGenes2<- c("Pdgfra", "Pdgfrb", "Col1a1", "Col3a1","Dcn", "Postn", "Acta2", "Cspg4", "Rgs5", "Pecam1", "Cdh5", "Lyve1", "Prox1", "Itga8", "Cxcl12", "Myh11", "Tagln")
StromalGenes<- list()
StromalGenes[["Fibroblast"]] <-c("Pdgfra", "Col1a1", "Col3a1","Dcn", "Postn")
StromalGenes[["Myofibroblast"]] <-c("Acta2", "Postn", "Tagln")
StromalGenes[["Pericyte"]] <-c("Pdgfrb", "Rgs5", "Cspg4")
StromalGenes[["Endothelial"]] <-c("Pecam1", "Cdh5", "Kdr")
StromalGenes[["LymphaticEC"]] <-c("Lyve1", "Pdpn", "Prox1")
StromalGenes[["vSMC"]] <-c("Acta2", "Myh11", "Tagln","Cnn1")
StromalGenes[["Mesangial"]] <-c("Itga8", "Pdgfrb", "Cxcl12")
# setup the directory
StromalDir<- paste0(Outdir,"StromalCells/")
dir.create(StromalDir)
setwd(StromalDir)
# run the function
PlotSpatialAndSingleCellFeatures(StromalGenes,SpatialObj, SingleCellObj, ImmuneSingleCellObj , StromalDir, SpatialColors )


# Epithelial cells
EpithelialGenes <-list()
EpithelialGenes[["PanPT"]] <-c("Lrp2","Slc34a1","Slc13a3")
EpithelialGenes[["PTS1"]] <-c("Slc5a2", "Slc34a1", "Lrp2", "Cubn", "Slc22a6", "Slc22a8", "Aqp1", "Gatm","Slc5a12")
EpithelialGenes[["PTS2"]] <-c("Acsm3","Cyp4a14", "Slc22a13", "Abcc4", "Slc7a13", "Gstm1","Slc13a3")
EpithelialGenes[["PTS3"]] <-c("Slc7a12", "Akr1c18", "Gsta1", "C", "Gstm1")
EpithelialGenes[["TDL"]] <-c("Aqp1")
EpithelialGenes[["CNT"]] <-c("Trpv5")
EpithelialGenes[["TAL"]] <-c("Umod", "Slc12a1", "Kcnj1", "Claudin10")
EpithelialGenes[["DCT"]] <-c("Slc12a3", "Pvalb", "Trpv5", "Calb1")
EpithelialGenes[["PC"]] <-c("Aqp2", "Aqp3", "Aqp4",  "Scnn1a")
EpithelialGenes[["ICA"]] <-c( "Atp6v1b1", "Slc4a1")
EpithelialGenes[["ICB"]] <-c( "Slc26a4", "Slc4a9", "Foxi1")
EpithelialGenes[["URO"]] <-c("Krt5","Upk1b","Krt14")
EpithelialGenes[["POD"]] <-c("Nphs1","Nphs2","Wt1")
EpithelialGenes[["PEC"]] <-c("Pdpn","Cldn1")

FinalEpithelialGenes<- EpithelialGenes %>% unlist()

EpithelialDir<- paste0(Outdir,"EpithelialGenes/")
dir.create(EpithelialDir)
setwd(EpithelialDir)
# run the function
PlotSpatialAndSingleCellFeatures(EpithelialGenes,SpatialObj, SingleCellObj, ImmuneSingleCellObj , EpithelialDir, SpatialColors )
```




```{r}
Finalcelltypemarkers<- c("Lrp2","Slc34a1","Slc13a3", # Proximal tubule cells (PT)
                      "Slc12a3","Pvalb","Wnk1",# Distal connecting tublue (DCT)

                     "Umod","Slc12a1","Cldn10", # Thick ascending limb cells (TAL)
                      "Klk1","Slc8a1","Calb1", # Connecting tubule (CNT)
                     "Slc14a2","Fst","Bst1", # descending thin  limb cells (DTL)
                     #"Akr1b1","Sh3gl3","Prox1", # ascending thin limb cells (ATL)
                     #"Enox1","Thsd4", #Macula densa
                      "Aqp2","Hsd11b2","Scnn1g", # principal cells (PC)
                    "Atp6v1g3","Aqp6","Slc26a7", # Intercalated cells
                    "Krt5","Upk1b","Krt14", # Urothelium
                    "Cdh5","Igfbp3","Pecam1", # Endothelium
                    "Col3a1","Fbn1","Lum","Col1a1", # Fibroblast
                    "Ireb2","Alas2", # Plasmacytoid dentritic cells
                    "S100a8","S100a9","Il1b", # Neutrophil
                    "C1qa","C1qb","Aif1", # Macrophage
                    "Cxcr6","Cd247","Cd3e", # T cells
                    "Igkc","Cd79a","Cd79b", # B cells
                    "Ccl5","Nkg7","Cd7" # NK/Cd8 cells
)

# Kidney epithelial cells
RequestedGenes<- c("Lrp2",
    "Slc34a1",
    "Slc5a12",
    # PTS1
    "Cyp4a14",
    # PTS2
    "Slc7a12",
    # PTS3
    "Acsm1",
    "Cubn",
    "Havcr1",
    "Slc12a3",
    "Slc8a1",
    "Calb1",
    "Trpm6",
    "Pvalb",
    "Wnk1",
    "Trpv5",
    "Fst",
    "Slc14a2",
    "Slc12a1",
    "Cldn10",
    "Atp6v1g3",
    "Aqp6",
    "Slc26a4",
    "Aqp2",
    "Hsd11b2" ,
    "Scnn1g",
    "Krt5",
    "Krt14",
    "Upk1b")


RequestedGenes <-
  c(
    # Epithelial cells
    "Lrp2",
    "Slc34a1",
    "Slc5a12",
    # PTS1
    "Cyp4a14",
    # PTS2
    "Slc7a12",
    # PTS3
    "Acsm1",
    "Cubn",
    "Havcr1",
    "Slc12a3",
    "Slc8a1",
    "Calb1",
    "Trpm6",
    "Pvalb",
    "Wnk1",
    "Trpv5",
    "Fst",
    "Slc14a2",
    "Slc12a1",
    "Cldn10",
    "Atp6v1g3",
    "Aqp6",
    "Slc26a4",
    "Aqp2",
    "Hsd11b2" ,
    "Scnn1g",
    "Krt5",
    "Krt14",
    "Upk1b",
    
 # stromal cell   
    "Col3a1",
    "Lum",
    "Fbn1",
    "Pecam1",
    "Igfbp3",
    "Cdh5",

  )

#
RequestedGenes<- c( "Ptprc",
      # neutrophil:
      "Itgam",
      "Ly6g",
      "S100a8",
    "S100a9",
    "Lyz2",
    "Lst1",
    "Elane",
    "Mpo",
    "Cxcr2",
    "Camp",
    "Ltf",

    # monocyte and macrophage:

    "C1qa",
    "C1qb",
    "Fcgr1",
    "Ly6c",
    "H2-A2",
    "Cx3cr1",
    "Adgre1",
    "Mrc1",
    "Arg1",
    "Nos2",
    
    #  T cells
    "Cd3e",
    "Cd3d",
    "Cd3g",
    "Trac",
    "Trdc",
    
    # CD4+ T cell
    "Cd4",
    "Il2ra",
    "Foxp3",
    "Cxcr5",
    "Tbx21",
    
    # CD8 T cell
    "Cd8a",
    "Cd8b",
    "Gzmb",
     "Prf1",
    
    # Activation
    "Cd44",
    # naive and central memory marker
    "Sell",
     # Exhuastion
    "Il2ra",
    "Pdcd1",
    "Havcr2",
    #Tissue resident memory T cell
    "Cxcr6", # trm and inflammation T cells
    "Cd69", # Early activation
    
    # 
    
    
    "Cd14",
    "Cxcr6",
    "Cd247",
    "Cd79b",
    "Ms4a1",
    "Igkc",
    "Nkg7",
    "Cd7",
    "Ccl5",
    "Ireb2",
    "Alas2",
    "C78334",
    "Cx3cr1",
    "Fcgr1",
    "H2-Aa",
    "Adgre1",
    "Cxcr2",
    "Ly6g",
    "S100a8",
    "S100a9",
    "Il1b")
# For receptors
#RequestedGenes<-c("C3ar1","Ccr1","Ccr2","Ccr5")

#RequestedDir <- c("/Users/XXW004/Documents/Projects/RuizRosado/SpatialTranscriptomic/Results/Cell_Cell_Communication_V08012024/Chemotacticfactors/")
#RequestedDir <- c("/Users/XXW004/Documents/Projects/RuizRosado/SpatialTranscriptomic/Results/Deconvolution_Six_V07292024/Features_ForPublications/StructuralTissueFeatures/")
RequestedDir <- c("/Users/XXW004/Documents/Projects/RuizRosado/SpatialTranscriptomic/Results/Deconvolution_Six_V07022025/CellTypesBiomarkers/")

Name<- "RegionalTissueMarker"
setwd(RequestedDir)
# requested genes in the spatial 

# Generate the Dotplot of the requested genes
Idents(SingleCellObj)

DimPlot(SingleCellObj, label = T)
RequestedGenesDotPlot <-
    DotPlot(SingleCellObj, features = RequestedGenes) + guides(color = guide_colorbar(title = 'AverExp'),
         size = guide_legend(title = "PercExp")) + RotatedAxis() 

 ggsave(paste0(Name,"_SingleCell_DotPlot.pdf"), plot = RequestedGenesDotPlot, height = 7, width = 6)

SpatialColors <- colorRampPalette(colors = rev(x = brewer.pal(n = 8, name = "Spectral")))

for (j in RequestedGenes) {

    # Function to generate DimPlot plot
    #  j="Calr"
    
    # we use this details to escape the figures that did not show up in spatial
    skip_to_next <- FALSE
    
    # Note that print(b) fails since b doesn't exist
    
    tryCatch(
      print(SpatialObj@assays$SCT@data[j, ]),
      error = function(e) {
        skip_to_next <<- TRUE
      }
    )
    
    if (skip_to_next) {
      next
    }
    # get the maxium and minvalue of the gene
    Maxvalue <- ceiling(max(SpatialObj@assays$SCT@data[j, ]))
    Minvalue <- floor(min(SpatialObj@assays$SCT@data[j, ]))
    GenesInRequestedMouseSpatialFeature <-
      SpatialFeaturePlot(
        SpatialObj,
        features = j,
        pt.size.factor = 1800,
        alpha = c(0.8, 1),
        crop = TRUE
      ) &
      ggplot2::scale_fill_gradientn(colours = SpatialColors(n=60),
                                    limits = c(Minvalue-1, Maxvalue)) 
  #   #& theme(
  # plot.background = element_rect(fill = "lightblue"),
  # panel.background = element_rect(fill = "lightblue"),
  # panel.grid = element_blank()
#)
    
    # save the figure into the folder
    ggsave(
      filename = paste0(RequestedDir, "/", "Spatial_", j, "_featureplot.pdf"),
      plot = GenesInRequestedMouseSpatialFeature,
      height = 3,
      width = 15
    )
    
    
    # generate the distribution from the single cells
    skip_to_next_sc <-FALSE
    tryCatch(
      print(SingleCellObj@assays$SCT@data[j, ]),
      error = function(e) {
        skip_to_next_sc <<- TRUE
      }
    )

    if (skip_to_next_sc) {
      next
    }
    # get the maxium and minvalue of the gene
    ScMaxvalue <- ceiling(max(SingleCellObj@assays$SCT@data[j, ]))
    ScMinvalue <- floor(min(SingleCellObj@assays$SCT@data[j, ]))
    #SingleCellObj @meta.data %>% head()
    #FeaturePlot(object = SingleCellObj, features = j,split.by = "DataSet")
    GenesInGOtermMouseSingleCellFeature <-
      FeaturePlot(object = SingleCellObj,
                  features = j,
                  split.by = "DataSet")

    # save the figure into the folder
    ggsave(
      filename = paste0(RequestedDir, "/", "SingleCell_", j, "_feature.pdf"),
      plot = GenesInGOtermMouseSingleCellFeature,
      height = 3,
      width = 12
    )

}

# requested genes in the single cells


```


# requested GO enrichments
```{r}
RequestedGOenrichments<- read.csv(paste0(Outdir,"PyelonephritisUrineDifferentiallyExpressedProteins_GOenrichment_NoSimplified_v0716.csv"))
# extract the gene list 
# generate the single cell figure for each gene within each pathway
AllGOterms<-RequestedGOenrichments$Description
# generate the GO terms that are interested 
AllGOterms
# instead of the all GO terms, we only generate the requested GO terms for top GO terms
RequestedGOTerms<- AllGOterms
RequestedGOTerms %>% tail()
library("biomaRt")
for(i in RequestedGOTerms) {
  # create a folder to store the
  # change the name of GO terms and join with _
  #i=c("maintenance of protein location in cell")
  GOtermName <- gsub(" ", "_", i)
  GOtermName
  GODir <- paste0(Interestedpathwaysdir, GOtermName)
  dir.create(GODir)
  setwd(GODir)
  
  
  # select the human urinary genes
  GenesInGOterm <-
    RequestedGOenrichments %>% filter(Description == i) %>% pull (geneID) %>% strsplit(split =
                                                                                         "/", fixed = T) %>% unlist()
  #mh_data %>% filter(Symbol.y == "LYZ")
  
  
  # transfer into mouse gene
  GenesInGOtermMouse <-
    mh_data %>% filter(Symbol.y %in% GenesInGOterm) %>% pull(Symbol.x) %>% unique()
  
  GenesInGOtermMouse
  # set the single cell identity into Date
  Idents(SingleCellObj) <- "DataSet"
  SingleCellObj@meta.data %>% head()
  
  # define the dotsize
  DotSize <-3
  length(GenesInGOtermMouse)
  #DotSize <- (ceiling(length(GenesInGOtermMouse) / 2) + 1)
  if (length(GenesInGOtermMouse) <= 3) {
    DotSize = 3
  } else if(length(GenesInGOtermMouse) >= 20){
    DotSize = 9
  }else{
    DotSize <- (ceiling(length(GenesInGOtermMouse)* 0.5) + 1)
  }
  
  # 12 genes we set up 6 # 18 gene set up 8, 24 gene set up 10
  #length(GenesInGOtermMouse) / 3
  # generate the DotPlot for different timepoints
  ScDateDotPlot <-
    DotPlot(SingleCellObj, features = GenesInGOtermMouse) + RotatedAxis()
  #
  ggsave(
    filename = paste0(GODir, "/", "VariousDate_", GOtermName, "_DotPlot.pdf"),
    plot = ScDateDotPlot,
    height = 3.5,
    width = DotSize
  )
  
  #
  Idents(SingleCellObj) <- "Raw_cell_type"
  # generate the DotPlot for different cell types
  ScCelltypeDotPlot <-
    DotPlot(SingleCellObj, features = GenesInGOtermMouse) + RotatedAxis()
  #
  ggsave(
    filename = paste0(GODir, "/", "VariousCelltype_", GOtermName, "_DotPlot.pdf"),
    plot = ScCelltypeDotPlot,
    height = 5,
    width = DotSize
  )
  
  # generate the spatial distribution and single cell distribution

  for (j in GenesInGOtermMouse) {
    # Generate DimPlot or SpatialDimPlot based on user's choice
    # Function to generate DimPlot plot
    #  j="Calr"

    # we use this details to escape the figures that did not show up in spatial
    skip_to_next <- FALSE

    # Note that print(b) fails since b doesn't exist

    tryCatch(
      print(SpatialObj@assays$SCT@data[j, ]),
      error = function(e) {
        skip_to_next <<- TRUE
      }
    )

    if (skip_to_next) {
      next
    }
    # get the maxium and minvalue of the gene
    Maxvalue <- ceiling(max(SpatialObj@assays$SCT@data[j, ]))
    Minvalue <- floor(min(SpatialObj@assays$SCT@data[j, ]))
    GenesInGOtermMouseSpatialFeature <-
      SpatialFeaturePlot(
        SpatialObj,
        features = j,
        pt.size.factor = 1500,
        crop = TRUE
      ) &
      ggplot2::scale_fill_gradientn(colours = myPalette(4),
                                    limits = c(Minvalue, Maxvalue-0.5))

    # save the figure into the folder
    ggsave(
      filename = paste0(GODir, "/", "Spatial_", j, "_In_", i, ".pdf"),
      plot = GenesInGOtermMouseSpatialFeature,
      height = 3,
      width = 15
    )

    # generate the distribution from the single cells
    skip_to_next_sc <-FALSE
    tryCatch(
      print(SingleCellObj@assays$SCT@data[j, ]),
      error = function(e) {
        skip_to_next_sc <<- TRUE
      }
    )

    if (skip_to_next_sc) {
      next
    }
    # get the maxium and minvalue of the gene
    ScMaxvalue <- ceiling(max(SingleCellObj@assays$SCT@data[j, ]))
    ScMinvalue <- floor(min(SingleCellObj@assays$SCT@data[j, ]))
    #SingleCellObj @meta.data %>% head()
    #FeaturePlot(object = SingleCellObj, features = j,split.by = "DataSet")
    GenesInGOtermMouseSingleCellFeature <-
      FeaturePlot(object = SingleCellObj,
                  features = j,
                  split.by = "DataSet")

    # save the figure into the folder
    ggsave(
      filename = paste0(GODir, "/", "SingleCell_", j, "_In_", i, ".pdf"),
      plot = GenesInGOtermMouseSingleCellFeature,
      height = 3,
      width = 12
    )


  }
  #
  
}




```


```{r}

# change the levels of the immune cells
Idents(IMMcells)<-ordered(factor(Idents(IMMcells)),levels=c("4","5","0","6","1","2","3"))

# FinalMarkers <- c("Il1b","S100a8","S100a9", "Cd79a","Cd79b","C1qa","C1qb","Itgam","Cd81","Ctsh","Mgl2","Cx3cr1", "Ccr2", "Nos2","Tnf","Ifngr1","Arg1","Mrc1","Il13","Cd163","Cd3g", "Cxcr6","Cd247","Cd3e","Capg","Lef1","Tcf7","Gzmk","Gimap7","Ms4a4b","Ccl5","Nkg7","Cd7","Lrp2","Umod","Aqp2","Igfbp7","Havcr1","Lcn2","Ly6c1")
CFilePath<- paste0(Outdir,"Candidatemarkers")

dir.create(file.path(CFilePath))
setwd(file.path(CFilePath))

# 
FinalMarkers <- c("Ccl5","Nkg7","Cd7","Cd79a","Cd79b","Ms4a1","Cd3g", "Cxcr6","Cd247","Cd3e","Il1b","S100a8","S100a9", "C1qa","C1qb","Itgam","Ctsh","Cd81","Mgl2","H2-Ab1","Cx3cr1", "Ccr2", "Nos2","Tnf","Arg1","Mrc1","Il13","Lrp2","Umod","Aqp2","Igfbp7","Havcr1","Lcn2","Ly6c1")

MYEGenesDotPlots<- DotPlot(IMMcells, features =  FinalMarkers, col.min = -1) +RotatedAxis()
MYEGenesDotPlots
ggsave(filename = "MYEGenesDotPlots.pdf", plot = MYEGenesDotPlots, width = 10, height = 5)
?SpatialFeaturePlot
library(ggplot2)
SpatialFeaturePlot(
        SpatialObj,
        features = "Lrp2",
        pt.size.factor = 1.6,
        crop = TRUE,image.alpha = 1,slot="data"
      ) &
      ggplot2::scale_fill_gradientn(colours = myPalette(4),
                                    limits = c(Minvalue, Maxvalue))
    

# 
# FinalMarkers <- c("S100a8","S100a9","Il1b","Itgam","Ly6g","Ly6c1", "Cd79a","Cd79b","C1qa","C1qb","Ctsh","Selenop","Cd81","Mgl2","H2-Ab1","Cx3cr1", "Ccr2", "Nos2","Tnf","Ifngr1","Arg1","Lrp2","Umod","Aqp2","Havcr1","Lcn2","Igfbp7")
# 
# 
# # Cfh: Parietal epithelial cells
# Finalcelltypemarkers<- c("Lrp2","Slc34a1","Slc13a3", # Proximal tubule cells (PT)
#                       "Slc12a3","Pvalb","Wnk1",# Distal connecting tublue (DCT)
# 
#                      "Umod","Slc12a1","Cldn10", # Thick ascending limb cells (TAL)
#                       "Klk1","Slc8a1","Calb1", # Connecting tubule (CNT)
#                      "Slc14a2","Fst","Bst1", # descending thin  limb cells (DTL)
#                      #"Akr1b1","Sh3gl3","Prox1", # ascending thin limb cells (ATL)
#                      #"Enox1","Thsd4", #Macula densa
#                       "Aqp2","Hsd11b2","Scnn1g", # principal cells (PC)
#                     "Atp6v1g3","Aqp6","Slc26a7", # Intercalated cells
#                     "Krt5","Upk1b","Krt14", # Urothelium
#                     "Cdh5","Igfbp3","Pecam1", # Endothelium
#                     "Col3a1","Fbn1","Lum","Col1a1", # Fibroblast
#                     "Ireb2","Alas2","C78334", # Plasmacytoid dentritic cells
#                     "S100a8","S100a9","Il1b","Lcn2", # Neutrophil
#                     "C1qa","C1qb","Aif1", # Macrophage
#                     "Cxcr6","Cd247","Cd3e", # T cells
#                     "Igkc","Cd79a","Cd79b", # B cells
#                     "Ccl5","Nkg7","Cd7", # NK/Cd8 cells
#                     "Hdc","Ifitm1","Ncr",
#                     "Mmp9","Ly6c1",
#                     "Havcr1")
# 
# MarkerGenesDotPlots<- DotPlot(IMMcells, features =  Finalcelltypemarkers, col.min = -0.5) +RotatedAxis()
# MarkerGenesDotPlots
# ggsave(filename = "MarkerGenesExpression_ClusterAnnotate.pdf", plot = MarkerGenesDotPlots, width = 15, height = 8)
# 
# For different type of cells



# For all the immune cell markers
for (i in FinalMarkers){
# 
  FeaturePlots<-FeaturePlot(IMMcells, features =  i, split.by = "DataSet" ) +RotatedAxis()
  ggsave(paste0(Outdir,"/CandidateMarkers/", i, "_Feature_GeneMarkers_Annoatation.pdf"), plot =FeaturePlots, height =4, width = 14)
}

# heatmap for all the cells

DoHeatmap(IMMcells, features = FinalMarkers)

```

```{r}
# T cells and others
# https://www.biocompare.com/Editorial-Articles/597618-A-Guide-to-Naive-T-Cell-Markers/
# https://www.cellsignal.com/pathways/immune-cell-markers-mouse
# Put the marker gene into a list

ImmuneDiffType <- list()

ImmuneDiffType[["Cd4"]]<- c("Cd4","Cd27","Ptprc","Il7r","Il2ra") # CD4 Naive cells: Cd4+, CD27+, CD45RA+, CD127+, CD25-, CD45RO-


ImmuneDiffType[["Cd8"]] <- c( "Cd8a","Cd28","Sell","Ccr7" ) # CD8 Naive T cells: Cd8+, CD27+, CD28+, CD45RA+, CD62L+, CCR7+, CD45RO-
ImmuneDiffType[["Th1"]] <-c('Stat1', 'Stat4', 'Tbx21', 'Il12a', 'Ifng','Tnf', 'Il2', 'Lta')
ImmuneDiffType[["Th2"]] <-c('Il4', 'Il2', 'Il10','Il13', 'Il25', 'Stat5a','Irf4','Stat6', 'Gata3', 'Il5')
ImmuneDiffType[["Th17"]] <-c('Tgfb1', 'Il6', 'Rorc', 'Stat3', 'Il17a', 'Il17f', 'Il23a', 'Irf4', 'Maf', 'Il21', 'Il22')
ImmuneDiffType[["Treg"]] <-c('Tgfb1', 'Il10', 'Foxp3', 'Stat5a', 'Il2')
ImmuneDiffType[["Th9"]] <-c('Stat6', 'Irf4', 'Tgfb1', 'Il4', 'Il9', 'Il10')
ImmuneDiffType[["Th22"]]<-c("Il22", 'Il23a', 'Il6', 'Ahr')
ImmuneDiffType[["Innate"]]<-c('Tlr4', 'Tlr5', 'Tlr11', 'Il6', 'Cxcr1', 'Cxcr2', 'Il1a', 'Il1b', 'Ccl5', 'Tnf', 'Ifng', 'Ccl2')
ImmuneDiffType[["AMP"]]<-c('Rnase4', 'Rnase6', 'Mbd1', 'Mbd2', 'Umod', 'Ltf', 'Camp')
ImmuneDiffType[["Bcells"]] <-c('Jchain', 'Pax5', 'Ms4a1', 'Cd19', 'Ighd', 'Cd27', 'Cd24a', 'Ptprc',  'Cd38', 'Cd79a', 'Ssr1', 'Fkbp11', 'Sec11c', 'Derl3', 'Prdx4', 'Cd79b', 'Vpreb3', 'Sdc1')
ImmuneDiffType[["Bcellsp"]]<-c("Cd19", "Jchain", "Pax5", "Ms4a1", "Ighd", "Cd79a", "Cd79b", "Ptprc", "H2-Ab1")
ImmuneDiffType[["neutrophil"]] <-c('Fut4', 'Fcgr3','Cxcr2', 'Mpo', 'Elane', 'Prtn3', 'Lyz1', 'S100a8', 'S100a9',  'Chil1', 'Anxa3', 'Cxcl1', 'Tgm3', 'Mmp9',  'Bpi',  'Ltf', "Gca", 'Camp', 'Ngp', "Chil3", 'Ly6g')
ImmuneDiffType[["neutros"]] <-c("Itgam", "Csf1r", "Ly6g")
ImmuneDiffType[["monocytes_macrophages"]] <-c('Itgam', 'Csf1r', 'Adgre1', 'Cd68', 'H2-Ab1', 'Cx3cr1')
ImmuneDiffType[["DC"]]<-c("Itgax", 'H2-Ab1', 'Bst2')
ImmuneDiffType[["NKcells"]]<-c('Nkg7', 'Klrk1', "Ncr1")
ImmuneDiffType[["Mono_Macro"]]<-c('Cd14', 'Fcgr3', 'Cst3', 'C1qc', 'Il1b','Mrc1', 'Msr1', 'Gca', "Ccr2", "Adgre1", 'Ly6c1')
ImmuneDiffType[["Mast_cells"]]<-c("Slc18a2", 'Fcer1a', 'Hpgds', 'Enpp3', 'Fcgr2b')
ImmuneDiffType[["Eosinophils"]]<-c('Ifng', 'Tnf')

length(ImmuneDiffType)
for (i in 1:length(ImmuneDiffType)) {
  # cell type name
  ImmuneCellType <- names(ImmuneDiffType[i])
  #ImmuneCellType <- "Mast_cells"
  # create a director for the cell type
  ImmuneCellTypeDir <- paste0(Outdir,
        "/CandidateMarkers/",
        ImmuneCellType,"/")
  dir.create(ImmuneCellTypeDir,showWarnings = FALSE)
  setwd(ImmuneCellTypeDir)
  # generate each feature
  #ImmuneDiffType[[ImmuneCellType]]
  for (j in ImmuneDiffType[[ImmuneCellType]]) {
    
    
    FeaturePlots <-
      FeaturePlot(IMMcells, features =  j, split.by = "DataSet") + RotatedAxis()
    ggsave(
      paste0(
        ImmuneCellTypeDir,
        ImmuneCellType,
        "_",
        j,
        "_Feature_GeneMarkers_Annoatation.pdf"
      ),
      plot = FeaturePlots,
      height = 4,
      width = 14
    )
  }
  
}

```

# add the Immune cell names
```{r}
# add the annotation using case_when
setwd(Outdir)
scrna@meta.data %>% head()
IMMcells@meta.data %>% head() 
  
levels(IMMcells@meta.data$SCT_snn_res.0.3)
IMMcells@meta.data <- IMMcells@meta.data %>%   mutate(ImmuneIntegrateAnnot = 
           case_when(SCT_snn_res.0.3 %in% c("0") ~ "T", # T cells
                     SCT_snn_res.0.3 %in% c("1") ~ "RecMAC", # Recruited MAC
                     SCT_snn_res.0.3 %in% c("2") ~ "M1-M2", # M1/M2 Macrophage
                     SCT_snn_res.0.3 %in% c("3") ~ "ResIMM", # Resident immune
                     SCT_snn_res.0.3 %in% c("4") ~ "NK", # NK cell
                     SCT_snn_res.0.3 %in% c("5") ~ "B", # B 
                     SCT_snn_res.0.3 %in% c("6") ~ "NEU", # Neutrophil
                 )) 

IMMcells@meta.data %>% head()

Idents(IMMcells) <- IMMcells@meta.data$ImmuneIntegrateAnnot
Idents(IMMcells)

# Reorder the levels:
Idents(IMMcells)<-ordered(factor(Idents(IMMcells)),levels=c("NK","B","T","NEU","RecMAC","M1-M2","ResIMM"))

FinalMarkersWithAnn <- c("Ccl5","Nkg7","Cd7","Cd79a","Cd79b","Ms4a1","Cd3g", "Cxcr6","Cd247","Cd3e","Il1b","S100a8","S100a9", "C1qa","C1qb","Itgam","Ctsh","Cd81","Mgl2","H2-Ab1","Cx3cr1", "Ccr2", "Nos2","Tnf","Arg1","Mrc1","Il13","Lrp2","Umod","Aqp2","Igfbp7","Havcr1","Lcn2","Ly6c1")

MYEGenesDotPlotsWithAnn<- DotPlot(IMMcells, features =  FinalMarkersWithAnn, col.min = -1) +RotatedAxis()
MYEGenesDotPlotsWithAnn
ggsave(filename = "MYEGenesDotPlots_WithAnn.pdf", plot = MYEGenesDotPlotsWithAnn, width = 12, height = 5)

IMMcellsDimWithAnno<- DimPlot(IMMcells,label = T)
IMMcellsDimWithAnno
ggsave(filename = "IMMcellsDimWithAnno.pdf", plot = IMMcellsDimWithAnno, width = 6, height =4)

```

# add the Immune cells types into the orginal datasets
```{r}
IMMcells@meta.data %>% head()

ImmuneSubAnnotation <- IMMcells@meta.data %>% dplyr::select(ImmuneIntegrateAnnot)
ImmuneSubAnnotation %>% head()
## add this annotation into the meta dataset
scrna@meta.data<- scrna@meta.data %>% rownames_to_column('CellID') %>% 
  left_join( ImmuneSubAnnotation %>% rownames_to_column('CellID')) %>% 
# then add a new column that change the NA into Raw_cell_type
  # mutate(NewAnnoation = coalesce(PTIntegrateAnnot,Raw_cell_type))
  mutate(ImmuneSub = ifelse(is.na(ImmuneIntegrateAnnot), A_new,ImmuneIntegrateAnnot))
### join the meta data to meta annoation

Idents(scrna) <- scrna@meta.data$ImmuneSub

scrna@meta.data %>% head()

row.names(scrna@meta.data) <- scrna@meta.data$CellID
 DimPlot(scrna, reduction = "umap", label = TRUE, pt.size = 0.5)

 saveRDS(scrna, file = paste0(Outdir,"OutputRDS/","PIPSeq_ImmuneSub_Annotated_v0823.RDS"))

 levels(Idents(scrna))
 
```

# add the frequency
```{r}
table(IMMcells@meta.data$DataSet)/table(scrna@meta.data$DataSet)

IMMcellsPropotionWithinTime <- table(IMMcells@meta.data$DataSet,Idents(IMMcells)) %>% as.data.frame() %>%
  group_by(Var1) %>%
  mutate(pct= prop.table(Freq) )  %>% ggplot(aes(x=Var1, y=pct, fill=Var2)) +geom_bar(position="stack", stat="identity") + theme_classic() +scale_y_continuous(labels = scales::percent)
ggsave("IMMcellsPropotionWithinTime.pdf", plot = IMMcellsPropotionWithinTime, width = 4, height = 3)

```

# check the markers
```{r}
# macrophage 1 and macrophage 2:
CFilePath<- paste0(Outdir,"Candidatemarkers")

dir.create(file.path(CFilePath))
setwd(file.path(CFilePath))

# M1 can choose CD80, CD86, CD64, CD16 and CD32 as markers. 
Macrophage2TypeMarkers <- c("Nos2","Cd79a","Cxcr6","Lef1","Ccl5","Capg","Ccl3","Ccl4","Ly6d","Lyz2","C1qa","Ifitm1","Ifngr1","Il17a","Cd80","Cd86","Fcgr1","Fcgr3","Fcgr2b","Cd163","Mrc1","Arg1", "Cd68","Havcr1","Lcn2")

IMMcells2BiomarkerDotPlot <- DotPlot(IMMcells, features = Macrophage2TypeMarkers) +RotatedAxis()
ggsave("IMMcells2BiomarkerDotPlot.pdf", plot = IMMcells2BiomarkerDotPlot, height = 5, width = 8)

Macrophage1Markers <- c("Nos2","Ifngr1","Il17a","Cd80","Cd86","Fcgr1","Fcgr3","Fcgr2b")

IMMcellsT1BiomarkerDotPlot<- DotPlot(IMMcells, features = Macrophage1Markers) +RotatedAxis()
ggsave("IMMcellsT1BiomarkerDotPlot.pdf", plot = IMMcellsT1BiomarkerDotPlot, height = 4, width = 6)

# M2 can choose CD163 and CD206 are major markers, Arg1 DECTIN
Macrophage2Markers <- c("Il13","Cd163","Mrc1","Arg1", "Cd68")
IMMcellsT2BiomarkerDotPlot<- DotPlot(IMMcells, features = Macrophage2Markers) +RotatedAxis()
ggsave("IMMcellsT2BiomarkerDotPlot.pdf", plot = IMMcellsT2BiomarkerDotPlot, height = 4, width = 6)


for (i in Macrophage1Markers){

  FeaturePlots<-FeaturePlot(IMMcells, features =  i, split.by = "DataSet" ) +RotatedAxis()

  ggsave(paste0(Outdir,"/CandidateMarkers/", i, "_M1_Feature_GeneMarkers_Annoatation.pdf"), plot =FeaturePlots, height =6, width = 14)

}

for (i in Macrophage2Markers){

  FeaturePlots<-FeaturePlot(IMMcells, features =  i, split.by = "DataSet" ) +RotatedAxis()

  ggsave(paste0(Outdir,"/CandidateMarkers/", i, "_M2_Feature_GeneMarkers_Annoatation.pdf"), plot =FeaturePlots, height =6, width = 14)

}
#
# setwd(file.path(Outdir))

```

```{r}
# contribution of monocyte-derived macrophage to inflammation

InflammationMarkers<- c("Cxcl1","Ccl4", "Ccl2", "Cxcl2")
GrowthFactors<- c("Csf3","Csf2")
InjuredMarkers <- c("Havcr1","Lcn2")
CytokinesMarkers<- c("Il23a","Ifngr1","Ifngr2","Il12","Il1a","Tnf","Il6","Il1f6")
Signaling <- c("Myd88","Tlr2","Tlr4","Nlrp3","Nfkb1","Nos2","Cybb")
Profibrotic <- c("Mmp2","Mmp9","Tgfb1","Col4a1","Col1a1","Fn1","Col3a1")

InflammationMarkersDotPlot<- DotPlot(IMMcells, features = InflammationMarkers) +RotatedAxis()
GrowthFactorsDotPlot<- DotPlot(IMMcells, features = GrowthFactors) +RotatedAxis()
InjuredMarkersDotPlot<- DotPlot(IMMcells, features = InjuredMarkers) +RotatedAxis()
CytokinesMarkersDotPlot<- DotPlot(IMMcells, features = CytokinesMarkers) +RotatedAxis()
SignalingDotPlot<- DotPlot(IMMcells, features = Signaling) +RotatedAxis()
ProfibroticDotPlot<- DotPlot(IMMcells, features = Profibrotic) +RotatedAxis()
ggsave("InflammationMarkersDotPlot.pdf", plot = InflammationMarkersDotPlot, height = 4, width = 6)
ggsave("GrowthFactorsDotPlot.pdf", plot = GrowthFactorsDotPlot, height = 4, width = 6)
ggsave("InjuredMarkersDotPlot.pdf", plot = InjuredMarkersDotPlot, height = 4, width = 6)
ggsave("CytokinesMarkersDotPlot.pdf", plot = CytokinesMarkersDotPlot, height = 4, width = 6)
ggsave("SignalingDotPlot.pdf", plot = SignalingDotPlot, height = 4, width = 6)
ggsave("ProfibroticDotPlot.pdf", plot = ProfibroticDotPlot, height = 4, width = 6)

```


```{r}
# the healthy cells
HealthMarkers<- c("Slc22a30", "Atp11a", "Trim7", "Slc22a6", "Slc7a13", "Slc5a2", "Slc5a10", "Spp2",  "Gatm", "Dcxr")

DotPlot(IMMcells, features =  HealthMarkers, col.min = -0.5) +RotatedAxis()

CyclingMarker <- "Mki67"
DotPlot(IMMcells, features =  CyclingMarker, col.min = -0.5) +RotatedAxis()

EarlyInjuryMarkers <- c("Cdh13", "Cdh6", "Havcr1")
DotPlot(IMMcells, features =  EarlyInjuryMarkers, col.min = -0.5) +RotatedAxis()
LateInjuryMarkers <-c( "Ccl2", "Gm17268", "Cald1")
DotPlot(IMMcells, features =  LateInjuryMarkers, col.min = -0.5) +RotatedAxis()

ReparingMarkers <- "Top2a"

DotPlot(IMMcells, features =  ReparingMarkers, col.min = -0.5) +RotatedAxis()

FailedFepairMarkers <- c("Ypel2","Kcnip4")

DotPlot(IMMcells, features =  FailedFepairMarkers, col.min = -0.5) +RotatedAxis()
# 
MaladptiveMarkers <- c("Vcam1", "Havcr1",  "Icam1", "Prom1","Cxcl2","Ccl3","Myc")

DotPlot(IMMcells, features =  MaladptiveMarkers, col.min = -0.5) +RotatedAxis()


# Combined 
#"Lrp2","Slc34a1","Slc13a1"
KidneyStatuesMarkers<- c( "Slc5a2","Slc34a1","Inmt","Apom", "Slc7a13","Slc22a6", "Prom1","Top2a","Cyp7b1","Mki67","Cdh13", "Cdh6", "Havcr1","Ccl2", "Gm17268", "Cald1","Vcam1", "Ypel2","Kcnip4","Hsp90aa1","Col27a1","Plin2")

 DotPlot(IMMcells, features =  KidneyStatuesMarkers,col.min = -1) +RotatedAxis()

KidneyStatuesTypeMarkers<- c("Il1b", "Mlkl", "Gsdmd","Ripk3","Fn1", "Col1a1", "Col3a1", "Tgfb1","Vcam1", "Havcr1", "Icam1", "Prom1", "Spp1", "Cst3","Gdf15", "Clu", "Igfbp7", "Apoe", "Aldob", "B2m", "S100a6","S100a8","S100a9","Pdgfb","Bak1","Bad","Casp8","Ptgs2","Casp1","Hmox1")
 DotPlot(IMMcells, features =  KidneyStatuesTypeMarkers) +RotatedAxis()
 
 KidneyStatuesTypeMarkers2<- c("Il1b", "Mlkl", "Gsdmd","Ripk3","Fn1", "Col1a1", "Col3a1", "Tgfb1","Vcam1", "Havcr1", "Icam1", "Prom1", "S100a6","S100a8","S100a9","Pdgfb","Bak1","Bad","Casp8","Ptgs2","Casp1","Hmox1","Nfkbia")
  DotPlot(IMMcells, features =  KidneyStatuesTypeMarkers2) +RotatedAxis()

 #, cluster 5: Fn1, Havcr1, Cst3, Igfbp7, S100a8, Il1b, Slc7a13 : Profibrotic PT
 # cluster 1: Spp1, Prom1, Kcnip4, Col27a1,Top2a,Nfkbia: Cycling PT
 # cluster 8: Cdh13, Hsp90aa1,Inmt, Mki67 : Proliferating PT
 
 # cluster 1: ASpp1, Prom1, Kcnipp4: Progenitor PT
 
 # cluster 0: Aldob
```

# the biomarkers for the 
```{r}
CandidateMarkers <- c("Aldob","Slc34a1","Spp1","Spp2","Lcn2", "Slc22a8","Slc5a2","Inmt","Prom1","Defb1","Top2a","Nfkbia", "Kcnip4", "Slc22a6", "Apoe","Havcr1", "Fn1","C3", "Il1b","S100a8", "Igfbp7", "Cst3", "Col27a1", "Hsp90aa1","Cdh13")

CandidateMarkersDotPlot<-  DotPlot(IMMcells, features =  CandidateMarkers, col.min = 0) +RotatedAxis()
CandidateMarkersDotPlot
ggsave("IMM_CandidateMarkersDotPlot.pdf", plot = CandidateMarkersDotPlot, height = 6, width = 12)
```

# #############################################
## generate the Trajectories of Macrophage cells
# #############################################
# 2.2 Prepared cds the object for Macrophage cells
```{r}
# build cds object for MYOLEOD
setwd(PyseudotimeDir)
Idents(IMMcells)
IMMcells@meta.data %>% head()

MACCells <- 
subset(IMMcells, subset =  ImmuneIntegrateAnnot == "M1_M2" | ImmuneIntegrateAnnot == "RecMAC")



MACCells_expr_matrix = GetAssayData(MACCells, slot = "counts")

# Cell meta annotation
p_data <- MACCells@meta.data 
head(p_data)

# Gene meta annotation
f_data<- data.frame(gene_short_name = row.names(MACCells), row.names = row.names(MACCells))


pre_cds <- new_cell_data_set(MACCells_expr_matrix,
                         cell_metadata = p_data,
                         gene_metadata = f_data)

pData(pre_cds) %>% colnames()

cds <- preprocess_cds(pre_cds) #标准化+PCA降维

```

# 2.2 Umap by Monocle3
```{r}
cds <- reduce_dimension(cds, preprocess_method = "PCA")
head(cds@int_colData$reducedDims$UMAP)

plot_cells(cds, reduction_method="UMAP", 
           show_trajectory_graph = FALSE,
           label_cell_groups = FALSE, 
           color_cells_by="DataSet")
dev.off()
# UMAP by Seurat
seurat_umap <- Embeddings(MACCells, reduction = "umap")[colnames(cds),]
cds@int_colData$reducedDims$UMAP <- seurat_umap
head(cds@int_colData$reducedDims$UMAP)
plot_cells(cds, reduction_method="UMAP",
           show_trajectory_graph = FALSE,
           label_cell_groups = FALSE,
           color_cells_by="DataSet")

```

# Draw the monocle state
```{r}
#分群(类似monocle的State)

cds <- cluster_cells(cds)

#预测轨迹
cds <- learn_graph(cds)
#交互式确定root节点，可以选择多个。我这里选择了一个
#cds <- order_cells(cds) 
# a helper function to identify the root principal points:
get_earliest_principal_node <- function(cds, time_bin="0DPI"){
  cell_ids <- which(colData(cds)[, "DataSet"] == time_bin)
  closest_vertex <-
  cds@principal_graph_aux[["UMAP"]]$pr_graph_cell_proj_closest_vertex
  closest_vertex <- as.matrix(closest_vertex[colnames(cds), ])
  root_pr_nodes <-
  igraph::V(principal_graph(cds)[["UMAP"]])$name[as.numeric(names
  (which.max(table(closest_vertex[cell_ids,]))))]
  
  root_pr_nodes
}
cds <- order_cells(cds, root_pr_nodes=get_earliest_principal_node(cds))


MACCells_pseudotime <-plot_cells(cds, color_cells_by = "pseudotime", 
           label_cell_groups = FALSE, 
           label_leaves = TRUE,  
           label_branch_points = FALSE,graph_label_size=1.5)
MACCells_pseudotime
ggsave(filename = "MACCells_pseudotime_MonocaCluster.pdf", plot = MACCells_pseudotime, width = 4, height = 3)

MACCells_pseudotime_Celltype <- plot_cells(cds, reduction_method="UMAP", 
           show_trajectory_graph = FALSE,
           label_cell_groups = FALSE, 
           color_cells_by="ImmuneIntegrateAnnot")
ggsave(filename = "MACCells_pseudotime_Celltype.pdf", plot = MACCells_pseudotime_Celltype, width = 5, height = 3)
MACCells_pseudotime_Celltype
```

```{r}
# Determine the psedudotime related genes
Track_genes <- graph_test(cds, neighbor_graph="principal_graph")

Track_genes %>% head()

```

# plot the expression of interesting genes during the PT pseudotime
```{r}

# re-order the tracked genes by the q value and morans I
Track_genes <- Track_genes[,c(5,2,3,4,1,6)] %>% 
  dplyr::arrange(desc(morans_I),q_value)

# Plot some interesting genes for the PT pseudotime : Top 2 genes, Napsa, Kap, Ttr, Igfbp7, Havcr1 and Prom1
KapTrack <- plot_genes_in_pseudotime(cds[Track_genes$gene_short_name[1],] , 
                         min_expr=0.5)
ggsave(filename = "MAC_pseudotime_KapTrack.pdf", plot = KapTrack, width = 5, height = 3)
NapsaTrack <- plot_genes_in_pseudotime(cds[Track_genes$gene_short_name[2],] , 
                         min_expr=0.5)
ggsave(filename = "MAC_pseudotime_NapsaTrack.pdf", plot = NapsaTrack, width = 5, height = 3)

TtrTrack <- plot_genes_in_pseudotime(cds["Ttr",] , 
                         min_expr=0.5)
ggsave(filename = "MAC_pseudotime_TtrTrack.pdf", plot = TtrTrack, width = 5, height = 3)

ApoeTrack <- plot_genes_in_pseudotime(cds["Apoe",] , 
                         min_expr=0.5)
ggsave(filename = "MAC_pseudotime_ApoeTrack.pdf", plot = ApoeTrack, width = 5, height = 3)

Fn1Track <- plot_genes_in_pseudotime(cds["Fn1",] , 
                         min_expr=0.5)

ggsave(filename = "MAC_pseudotime_Fn1Track.pdf", plot = Fn1Track, width = 5, height = 3)

# For the 
Cst3Track <- plot_genes_in_pseudotime(cds["Cst3",],min_expr=0.5 )
ggsave(filename = "MAC_pseudotime_Cst3Track.pdf", plot = Cst3Track, width = 5, height = 3)

H2Ab1Track <- plot_genes_in_pseudotime(cds["H2-Ab1",] , 
                         min_expr=0.5)
ggsave(filename = "MAC_pseudotime_H2Ab1Track.pdf", plot = H2Ab1Track, width = 5, height = 3)

```

# choose the top signifcant for visualization
```{r}
diff_pseudo_gene <- Track_genes %>% 
  dplyr::arrange(q_value) %>%
  # remove the genes with mitochondrion
  filter(!grepl('^mt-|Hbb-', gene_short_name)) %>%
  rownames() %>% head(40)

# we ignore the mitochondria genes

diff_pseudo_gene

# generate the heatmap for those genes

diff_pseudo_gene.matrix <- exprs(cds)[match(diff_pseudo_gene,rownames(rowData(cds))),order(pseudotime(cds))]

diff_pseudo_gene.matrix <- t(apply(diff_pseudo_gene.matrix,1,function(x){smooth.spline(x,df=3)$y}))
diff_pseudo_gene.matrix <- t(apply(diff_pseudo_gene.matrix,1,function(x){(x-mean(x))/sd(x)}))
rownames(diff_pseudo_gene.matrix) <- diff_pseudo_gene;

#pheatmap(diff_pseudo_gene.matrix)

# because the plot_pseudotime_heatmap no longer have, we used the complexheatmap to draw the figure
# Reference: https://github.com/cole-trapnell-lab/monocle-release/issues/295
pdf("MAC_pseudotime_psudotimeGenes.pdf", height = 6, width = 4)
htkm <- Heatmap(
  diff_pseudo_gene.matrix,
  name    = "z-score",
  col     = colorRamp2(seq(from=-2,to=2,length=11),rev(brewer.pal(11, "Spectral"))),
  show_row_names  = TRUE,
  show_column_names = FALSE,
  row_names_gp  = gpar(fontsize = 6),
  km = 8,
  row_title_rot = 0,
  use_raster=TRUE,
  cluster_rows  = TRUE,
  cluster_row_slices = FALSE,
  cluster_columns   = FALSE,show_row_dend =FALSE)

print (htkm)
dev.off()
#ggsave(filename = "PT_pseudotime_psudotimeGenes.pdf", plot = htkm, width = 8, height = 6)

?Heatmap
#Ward.D2 Hierarchical Clustering
pdf("MAC_pseudotime_psudotimeGenes_Hierarchicalcluster.pdf", height = 6, width = 4)
hthc <- Heatmap(
  diff_pseudo_gene.matrix,
  name                         = "z-score",
  col                          = colorRamp2(seq(from=-2,to=2,length=11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 6),
  clustering_method_rows = "ward.D2",
  clustering_method_columns = "ward.D2",
  row_title_rot                = 0,
  use_raster= TRUE,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE,show_row_dend=FALSE)

print(hthc)
dev.off()

cds["Havcr1",] %>% head()
# Generate the Dotplot for these genes
plot_genes_in_pseudotime(cds[c("Havcr1"),] , 
                         min_expr=0.5)

DotPlot(IMMcells, features =diff_pseudo_gene ) +RotatedAxis()

DotPlot(IMMcells, features =c("Spp1","Havcr1","Itgb1","Top2a",
                             "Prom1","Pdgfb","Defb1","Fn1","Col6a1"), col.min = -0.5 ) +RotatedAxis()

FeaturePlot(IMMcells, features = c("Lcn2","Defb1","Egf","Havcr1"))
# the following no longer have 

Track_genes
# p = plot_pseudotime_heatmap(cds[diff_pseudo_gene,], 
#                         num_clusters = 3,  # default 6
#                         return_heatmap=T)
# 

```


# #############################################
## generate the Trajectories of T cells
# #############################################
# 2.3 Prepared cds the object for T cells
```{r}
# build cds object for T
setwd(PyseudotimeDir)
Idents(IMMcells)
IMMcells@meta.data %>% head()

TCells <- 
subset(IMMcells, subset =   ImmuneIntegrateAnnot == "T")



TCells_expr_matrix = GetAssayData(TCells, slot = "counts")

# Cell meta annotation
p_data <- TCells@meta.data 
head(p_data)

# Gene meta annotation
f_data<- data.frame(gene_short_name = row.names(TCells), row.names = row.names(TCells))


pre_cds <- new_cell_data_set(TCells_expr_matrix,
                         cell_metadata = p_data,
                         gene_metadata = f_data)

pData(pre_cds) %>% colnames()

cds <- preprocess_cds(pre_cds) #标准化+PCA降维

```

# 2.2 Umap by Monocle3
```{r}
cds <- reduce_dimension(cds, preprocess_method = "PCA")
head(cds@int_colData$reducedDims$UMAP)

plot_cells(cds, reduction_method="UMAP", 
           show_trajectory_graph = FALSE,
           label_cell_groups = FALSE, 
           color_cells_by="DataSet")
dev.off()
# UMAP by Seurat
seurat_umap <- Embeddings(TCells, reduction = "umap")[colnames(cds),]
cds@int_colData$reducedDims$UMAP <- seurat_umap
head(cds@int_colData$reducedDims$UMAP)
plot_cells(cds, reduction_method="UMAP",
           show_trajectory_graph = FALSE,
           label_cell_groups = FALSE,
           color_cells_by="DataSet")

```

# Draw the monocle state
```{r}
#分群(类似monocle的State)

cds <- cluster_cells(cds)

#预测轨迹
cds <- learn_graph(cds)
#交互式确定root节点，可以选择多个。我这里选择了一个
#cds <- order_cells(cds) 
# a helper function to identify the root principal points:
get_earliest_principal_node <- function(cds, time_bin="0DPI"){
  cell_ids <- which(colData(cds)[, "DataSet"] == time_bin)
  
  closest_vertex <-
  cds@principal_graph_aux[["UMAP"]]$pr_graph_cell_proj_closest_vertex
  closest_vertex <- as.matrix(closest_vertex[colnames(cds), ])
  root_pr_nodes <-
  igraph::V(principal_graph(cds)[["UMAP"]])$name[as.numeric(names
  (which.max(table(closest_vertex[cell_ids,]))))]
  
  root_pr_nodes
}
cds <- order_cells(cds, root_pr_nodes=get_earliest_principal_node(cds))


TCells_pseudotime <-plot_cells(cds, color_cells_by = "pseudotime", 
           label_cell_groups = FALSE, 
           label_leaves = TRUE,  
           label_branch_points = FALSE,graph_label_size=1.5)
TCells_pseudotime
ggsave(filename = "TCells_pseudotime_MonocaCluster.pdf", plot = TCells_pseudotime, width = 4, height = 3)

TCells_pseudotime_Celltype <- plot_cells(cds, reduction_method="UMAP", 
           show_trajectory_graph = FALSE,
           label_cell_groups = FALSE, 
           color_cells_by="ImmuneIntegrateAnnot")
ggsave(filename = "TCells_pseudotime_Celltype.pdf", plot = TCells_pseudotime_Celltype, width = 5, height = 3)
TCells_pseudotime_Celltype
```

```{r}
# Determine the psedudotime related genes
Track_genes <- graph_test(cds, neighbor_graph="principal_graph")

Track_genes %>% head()
# re-order the tracked genes by the q value and morans I
Track_genes <- Track_genes[,c(5,2,3,4,1,6)] %>% 
  dplyr::arrange(desc(morans_I),q_value)

```

# choose the top signifcant for visualization
```{r}
diff_pseudo_gene <- Track_genes %>% 
  dplyr::arrange(q_value) %>%
  # remove the genes with mitochondrion
  filter(!grepl('^mt-|Hbb-', gene_short_name)) %>%
  rownames() %>% head(40)

# we ignore the mitochondria genes

diff_pseudo_gene

# generate the heatmap for those genes

diff_pseudo_gene.matrix <- exprs(cds)[match(diff_pseudo_gene,rownames(rowData(cds))),order(pseudotime(cds))]

diff_pseudo_gene.matrix <- t(apply(diff_pseudo_gene.matrix,1,function(x){smooth.spline(x,df=3)$y}))
diff_pseudo_gene.matrix <- t(apply(diff_pseudo_gene.matrix,1,function(x){(x-mean(x))/sd(x)}))
rownames(diff_pseudo_gene.matrix) <- diff_pseudo_gene;

#pheatmap(diff_pseudo_gene.matrix)

# because the plot_pseudotime_heatmap no longer have, we used the complexheatmap to draw the figure
# Reference: https://github.com/cole-trapnell-lab/monocle-release/issues/295
pdf("Tcells_pseudotime_psudotimeGenes.pdf", height = 6, width = 4)
htkm <- Heatmap(
  diff_pseudo_gene.matrix,
  name    = "z-score",
  col     = colorRamp2(seq(from=-2,to=2,length=11),rev(brewer.pal(11, "Spectral"))),
  show_row_names  = TRUE,
  show_column_names = FALSE,
  row_names_gp  = gpar(fontsize = 6),
  km = 8,
  row_title_rot = 0,
  use_raster=TRUE,
  cluster_rows  = TRUE,
  cluster_row_slices = FALSE,
  cluster_columns   = FALSE,
  show_row_dend=FALSE)

print (htkm)
dev.off()
#ggsave(filename = "PT_pseudotime_psudotimeGenes.pdf", plot = htkm, width = 8, height = 6)


#Ward.D2 Hierarchical Clustering
pdf("Tcells_pseudotime_psudotimeGenes_Hierarchicalcluster.pdf", height = 6, width = 4)
hthc <- Heatmap(
  diff_pseudo_gene.matrix,
  name                         = "z-score",
  col                          = colorRamp2(seq(from=-2,to=2,length=11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 6),
  clustering_method_rows = "ward.D2",
  clustering_method_columns = "ward.D2",
  row_title_rot                = 0,
  use_raster= TRUE,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE,
  show_row_dend=FALSE)

print(hthc)
dev.off()
?Heatmap
cds["Havcr1",] %>% head()
# Generate the Dotplot for these genes
plot_genes_in_pseudotime(cds[c("Havcr1"),] , 
                         min_expr=0.5)

DotPlot(IMMcells, features =diff_pseudo_gene ) +RotatedAxis()

DotPlot(IMMcells, features =c("Spp1","Havcr1","Itgb1","Top2a",
                             "Prom1","Pdgfb","Defb1","Fn1","Col6a1"), col.min = -0.5 ) +RotatedAxis()

FeaturePlot(IMMcells, features = c("Lcn2","Defb1","Egf","Havcr1"))
# the following no longer have 

Track_genes
# p = plot_pseudotime_heatmap(cds[diff_pseudo_gene,], 
#                         num_clusters = 3,  # default 6
#                         return_heatmap=T)
# 

```


# plot the expression of interesting genes during the T cell pseudotime
```{r}

# Plot some interesting genes for the PT pseudotime : Top 2 genes, Napsa, Kap, Ttr, Igfbp7, Havcr1 and Prom1
KapTrack <- plot_genes_in_pseudotime(cds[Track_genes$gene_short_name[1],] , 
                         min_expr=0.5)
ggsave(filename = "Tcells_pseudotime_KapTrack.pdf", plot = KapTrack, width = 5, height = 3)
NapsaTrack <- plot_genes_in_pseudotime(cds[Track_genes$gene_short_name[2],] , 
                         min_expr=0.5)
ggsave(filename = "Tcells_pseudotime_NapsaTrack.pdf", plot = NapsaTrack, width = 5, height = 3)

TtrTrack <- plot_genes_in_pseudotime(cds["Ttr",] , 
                         min_expr=0.5)
ggsave(filename = "Tcells_pseudotime_TtrTrack.pdf", plot = TtrTrack, width = 5, height = 3)

Igfbp7Track <- plot_genes_in_pseudotime(cds["Igfbp7",] , 
                         min_expr=0.5)
ggsave(filename = "Tcells_pseudotime_Igfbp7Track.pdf", plot = Igfbp7Track, width = 5, height = 3)

Track_genes$gene_short_name[1:5]


Track_genes$gene_short_name[2]
Track_genes$gene_short_name[1]
Defb1Track <- plot_genes_in_pseudotime(cds["Defb1",] , 
                         min_expr=0.5)
Defb1Track
ggsave(filename = "Tcells_pseudotime_Defb1Track.pdf", plot = Defb1Track, width = 5, height = 3)

# For the 
Ly6c2Track <- plot_genes_in_pseudotime(cds["Ly6c2",],min_expr=1 )
ggsave(filename = "Tcells_pseudotime_Ly6c2Track.pdf", plot = Ly6c2Track, width = 5, height = 3)

Il17aTrack <- plot_genes_in_pseudotime(cds["Il17a",],min_expr=0.5 )
ggsave(filename = "Tcells_pseudotime_Il17aTrack.pdf", plot = Il17aTrack, width = 5, height = 3)


Ccl5Track <- plot_genes_in_pseudotime(cds["Ccl5",],min_expr=0.5 )
ggsave(filename = "Tcells_pseudotime_Ccl5Track.pdf", plot = Ccl5Track, width = 5, height = 3)
Cdh16Track
#FeaturePlot(IMMcells, features = "Havcr1")

```


# #####################################
## generate the Progeny signaling pathways
# #####################################

```{r}
# set up the directory
setwd(ProgenyDir)
Idents(IMMcells)
IMMcells@meta.data %>% head()

options(future.globals.maxSize = 891289600000) 
## We compute the Progeny activity scores and add them to our Seurat object
## as a new assay called Progeny. 

# create cluster data frame with the cells, Celltypes
IMMCluster<- data.frame(Cell = names(Idents(IMMcells)),   
                              CellType = as.character(Idents(IMMcells)),
                              stringsAsFactors = FALSE)
IMMCluster %>% head()


IMMcells <- progeny(IMMcells, scale=FALSE, organism="Mouse", top=500, perm=1, return_assay = TRUE)
IMMcells
# setup the default assay as progeny
DefaultAssay(object = IMMcells) <- "progeny"

IMMcells <- Seurat::ScaleData(IMMcells, assay = "progeny") 
## We can now directly apply Seurat functions in our Progeny scores. 
## For instance, we scale the pathway activity scores. 
progeny_scores_df <- 
  as.data.frame(t(GetAssayData(IMMcells, slot = "scale.data", 
                               assay = "progeny"))) %>%
  rownames_to_column("Cell") %>%
  gather(Pathway, Activity, -Cell) 

progeny_scores_df %>% head()

#progeny_scores_df %>% filter(Cell =="0DPI_AAAAAAAACCACCAGA")
## We match Progeny scores with the cell clusters.
progeny_scores_df <- inner_join(progeny_scores_df, IMMCluster)

progeny_scores_df %>% head()
## We summarize the Progeny scores by cellpopulation
summarized_progeny_scores <- progeny_scores_df %>% 
  group_by(Pathway, CellType) %>%
  summarise(avg = mean(Activity), std = sd(Activity))
summarized_progeny_scores


## We prepare the data for the plot
summarized_progeny_scores_df <- summarized_progeny_scores %>%
  dplyr::select(-std) %>%   
  spread(Pathway, avg) %>%
  data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 

paletteLength = 100
myColor = colorRampPalette(c("royalblue", "white","lightcoral"))(paletteLength)

progenyBreaks = c(seq(min(summarized_progeny_scores_df), 0, 
                      length.out=ceiling(paletteLength/2) + 1),
                  seq(max(summarized_progeny_scores_df)/paletteLength, 
                      max(summarized_progeny_scores_df), 
                      length.out=floor(paletteLength/2)))

summarized_progeny_scores_df
ImmuneCellForHeatmap<-as.data.frame(t(summarized_progeny_scores_df[,-1]))
ImmuneCellForHeatmap
# generate the heatmap for ImmuneCells
pdf(file="ImmuneCells_Progeny_ForHeatmap.pdf",height = 5, width = 6)
library(pheatmap)
progeny_hmap = pheatmap(ImmuneCellForHeatmap,fontsize=12, 
                        fontsize_row = 10, 
                        color=myColor, breaks = progenyBreaks, 
                        #main = "PROGENy signaling", 
                        angle_col = 45,
                        treeheight_col = 0,  border_color = NA, treeheight_row=0)
print(progeny_hmap)
dev.off()
test<- FeaturePlot(PTcells, features = "NFkB")
ggsave("test_NFKB.pdf", plot = test, width = 5, height = 5)
# generate the featureplot 
ProgenyFeatures <- summarized_progeny_scores$Pathway %>% unique()
ProgenyFeatures
for (i in ProgenyFeatures ){
  p2 <-(FeaturePlot(IMMcells, features = i,split.by = "DataSet") & 
  scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red'))& 
  ggtitle(paste0(i," activity"))
  
  ggsave(paste0(i, "_Immune_cluster_featurePlot.pdf"), plot = p2, width = 20, height = 5)
}

for (i in ProgenyFeatures ){
  p2 <-(FeaturePlot(IMMcells, features = i) & 
  scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red'))& 
  ggtitle(paste0(i," activity"))
  
  ggsave(paste0(i, "_Immune_cluster_featurePlot_Nosplit.pdf"), plot = p2, width = 4, height = 3)
}

```

# Using PROGENy model from DecoupleR
```{r}
library(decoupleR)
setwd(ProgenyDir)
net <- get_progeny(organism = 'mouse', top = 500)

# Activity inference with Multivariate Linear Model (MLM)
# Extract the normalized log-transformed counts
mat <- as.matrix(IMMcells@assays$RNA@data)

# Run mlm
acts <- run_mlm(mat=mat, net=net, .source='source', .target='target',
                .mor='weight', minsize = 5)

acts 
# Visualization:
# Extract mlm and store it in pathwaysmlm in data
IMMcells[['pathwaysmlm']] <- acts %>%
  pivot_wider(id_cols = 'source', names_from = 'condition',
              values_from = 'score') %>%
  column_to_rownames('source') %>%
  Seurat::CreateAssayObject(.)


# Change assay
DefaultAssay(object = IMMcells) <- "pathwaysmlm"

# Scale the data
IMMcells <- ScaleData(IMMcells)
IMMcells@assays$pathwaysmlm@data <- IMMcells@assays$pathwaysmlm@scale.data

# run the source of pathways
ProgenyFeatures <- levels(factor(acts$source))
ProgenyFeatures
for (i in ProgenyFeatures ){
  p2 <-FeaturePlot(IMMcells, features = i) & 
 scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red')& 
  ggtitle(paste0(i," activity"))
  
  ggsave(paste0(i, "_IMMcells_cluster_featurePlot_DecoupleR.pdf"), plot = p2, width = 6, height = 5)
}

#(c("royalblue", "white","lightcoral"))
FeaturePlot(IMMcells, features = "Hypoxia", min.cutoff = -4, max.cutoff = 6, split.by = "DataSet") & 
 scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red')& 
  ggtitle(paste0("Hypoxia"," activity"))

FeaturePlot(IMMcells, features = "NFkB", min.cutoff = -4, max.cutoff = 6, split.by = "DataSet") & 
 scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red')& 
  ggtitle(paste0("NFkB"," activity"))


FeaturePlot(IMMcells, features = "TNFa", min.cutoff = -4, max.cutoff = 6, split.by = "DataSet") & 
 scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red')& 
  ggtitle(paste0("TNFa"," activity"))

p1 <- DimPlot(IMMcells, reduction = "umap", label = TRUE, pt.size = 0.5) + 
  NoLegend() + ggtitle('Cell types')
p2 <- (FeaturePlot(IMMcells, features = c("Trail")) & 
  scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red')) +
  ggtitle('Trail activity')



# Expore the mean activiey per group across pathways:
# Extract activities from object as a long dataframe
df <- t(as.matrix(IMMcells@assays$pathwaysmlm@data)) %>%
  as.data.frame() %>%
  mutate(cluster = Idents(IMMcells)) %>%
  pivot_longer(cols = -cluster, names_to = "source", values_to = "score") %>%
  group_by(cluster, source) %>%
  summarise(mean = mean(score))

# Transform to wide matrix
top_acts_mat <- df %>%
  pivot_wider(id_cols = 'cluster', names_from = 'source',
              values_from = 'mean') %>%
  column_to_rownames('cluster') %>%
  as.matrix()

# Choose color palette
palette_length = 100
my_color = colorRampPalette(c("Darkblue", "white","red"))(palette_length)

my_breaks <- c(seq(-0.5, 0, length.out=ceiling(palette_length/2) + 1),
               seq(0.05, 0.5, length.out=floor(palette_length/2)))

# Plot
IMMcellsHeatmap<- pheatmap(t(top_acts_mat), border_color = NA, color=my_color, breaks = my_breaks, angle_col = 45) 
ggsave("IMMcellscellsHeatmap_Progeny_Decouple.pdf", plot = IMMcellsHeatmap, width = 6, height = 6)

```

# Transcription factor activity inference from Immune scRNA-seq
```{r}
# CollecTRI network
# set up the directory
setwd(TFDir)
netTF <-  readRDS("/Users/XXW004/Documents/Projects/RuizRosado/PIPseq/Datasets/get_collectri_human_mouse20230716/get_collectri_mouse_process_table.rds")

mat <- as.matrix(IMMcells@assays$RNA@data)
# Run ulm
acts <- run_ulm(mat=mat, net=netTF, .source='source', .target='target',
                .mor='mor', minsize = 5)

saveRDS(acts, file="Immune_Transcriptfactor.RDS")
# Activity inference with Univariate Linear Model (ULM)
# Extract ulm and store it in tfsulm in pbmc
IMMcells[['tfsulm']] <- acts %>%
  pivot_wider(id_cols = 'source', names_from = 'condition',
              values_from = 'score') %>%
  column_to_rownames('source') %>%
  Seurat::CreateAssayObject(.)

# Change assay
DefaultAssay(object = IMMcells) <- "tfsulm"

# Scale the data
IMMcells <- ScaleData(IMMcells)
IMMcells@assays$tfsulm@data <- IMMcells@assays$tfsulm@scale.data

# Explore the top 20 more varaible TFs
n_tfs <- 25

t(as.matrix(IMMcells@assays$tfsulm@data))
# Extract activities from object as a long dataframe
df <- t(as.matrix(IMMcells@assays$tfsulm@data)) %>%
  as.data.frame() %>%
  mutate(cluster = Idents(IMMcells)) %>%
  pivot_longer(cols = -cluster, names_to = "source", values_to = "score") %>%
  group_by(cluster, source) %>%
  summarise(mean = mean(score))

# Get top tfs with more variable means across clusters
tfs <- df %>%
  group_by(source) %>%
  summarise(std = sd(mean)) %>%
  arrange(-abs(std)) %>%
  head(n_tfs) %>%
  pull(source)
# drop "Znf335" due to not exist in the expression
tfs <- tfs[!tfs == "Znf335"]

#tfs[!tfs == "Znf335"]
# Subset long data frame to top tfs and transform to wide matrix
top_acts_mat <- df %>%
  filter(source %in% tfs) %>%
  pivot_wider(id_cols = 'cluster', names_from = 'source',
              values_from = 'mean') %>%
  column_to_rownames('cluster') %>%
  as.matrix()

# Choose color palette
palette_length = 100
my_color = colorRampPalette(c("Darkblue", "white","red"))(palette_length)

my_breaks <- c(seq(-3, 0, length.out=ceiling(palette_length/2) + 1),
               seq(0.05, 3, length.out=floor(palette_length/2)))

# Plot
IMMTFheatmap<- pheatmap(t(top_acts_mat),border_color = NA, cluster_cols = F, 
                        color=my_color,fontsize=12, fontsize_row = 10, 
                        breaks = my_breaks, 
                        #main = "Predicted TF activities", 
                        angle_col = 45,
                        treeheight_col = 0, treeheight_row=0)


ggsave("IMMcellsHeatmap_Transcriptfactor_Progeny_Decouple.pdf", plot = IMMTFheatmap, width = 4, height = 5)

# draw the transcriptor factor activity and their expression
# generate the figure of transcript factor activity:
for (m in tfs){
  p2 <- (FeaturePlot(IMMcells, features = m) & 
  scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red')) +
  ggtitle(paste0(m, ' activity'))
  
  ggsave(paste0("TFActivity_",m,"_IMMcells_FeaturePlot.pdf"), plot = p2, width = 4, height = 3)
}

DefaultAssay(object = IMMcells) <- "RNA"

for (n in tfs){
  p3 <- (FeaturePlot(IMMcells, features = n) & 
  scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red')) +
  ggtitle(paste0(n, ' expression'))
  
  ggsave(paste0("TFExpression_",n,"_IMMcells_FeaturePlot.pdf"), plot = p3, width = 4, height = 3)

}

# add their activity in Dotplot
Idents(IMMcells)
# get the order of row name
GeneOrderByPheatmap <- rownames(t(top_acts_mat))[IMMTFheatmap$tree_row$order]
#IMMTFheatmap$tree_row$order

?DotPlot
TFDotPlot <- DotPlot(IMMcells, features = rev(GeneOrderByPheatmap)) +
  RotatedAxis() + coord_flip()+ 
  #ggtitle("Predicted TF expression")+
   theme(legend.position = "right")
ggsave("Immune_TFDotPlot.pdf", plot = TFDotPlot, width = 5, height = 6)

```




```{r}
# dividing into cycling, transitioning, adap- tive (successful or maladaptive repair) and degenerative (damaged or stressed). 
# cycling: Mki67
# Early Injury: Cdh13, Cdh6, Havcr1
# Late Injury: Ccl2, Gm17268, Cald1
# conserved: Slc13a1, Slc5a2,  Slc7a13, Cyp7b1



# Panoptosis
# Il1b, MLKL, Gsdmd,Ripk3

# Fibrosis 
# Fn1, Col1a1, Col3a1, Tgfb1
# adaptive
# VCAM1, DCDC2 and HAVCR1, C3, ICAM1
# VCAM1 PROM1 ITGB8
# Vcam1, Havcr1, Dcdc2, Icam1, Prom1, ITgb8

#degenerative
# SPP1, CST3, CLU and IGFBP7, APOE ALDOB
# B2M S100A6
# Spp1, Cst3, GDF15,Clu, Igfbp7, Apoe, Aldob, B2m, S100a6
# agionegious PEC,
# EGFR, VEGF, Cdh5

# Egfr, Vegf, Cdh5
KidneyStatuesMarkers<- c("Slc13a1", "Slc5a2", "Slc7a13", "Cyp7b1","Mki67","Cdh13", "Cdh6", "Havcr1","Ccl2", "Gm17268", "Cald1")

KidneyStatuesTypeMarkers<- c("Il1b", "Mlkl", "Gsdmd","Ripk3","Fn1", "Col1a1", "Col3a1", "Tgfb1","Vcam1", "Havcr1", "Icam1", "Prom1", "Spp1", "Cst3","Gdf15", "Clu", "Igfbp7", "Apoe", "Aldob", "B2m", "S100a6")

DotPlot(scrna, features =  KidneyStatuesMarkers, col.min = -0.5) +RotatedAxis()
DotPlot(scrna, features =  KidneyStatuesTypeMarkers, col.min = -0.5) +RotatedAxis()


for (i in KidneyStatuesMarkers){
# 
  FeaturePlots<-FeaturePlot(scrna, features =  i, split.by = "DataSet" ) +RotatedAxis()
  ggsave(paste0(Outdir,"/KidneyStatuesMarkers/", i, "_Feature_GeneMarkers_Annoatation.pdf"), plot =FeaturePlots, height =6, width = 14)
}

for (i in KidneyStatuesTypeMarkers){
# 
  FeaturePlots<-FeaturePlot(scrna, features =  i, split.by = "DataSet" ) +RotatedAxis()
  ggsave(paste0(Outdir,"/KidneyStatuesTypeMarkers/", i, "_Feature_GeneMarkers_Annoatation.pdf"), plot =FeaturePlots, height =6, width = 14)
}


# with the inflammation markers

# with the fibrosis markers

# with the apoptosis markers

# with the necrosis markers

# with the endothelial markers


```


