---
title: "PipseqPTSubcluster_V0822"
author: "Xin Wang"
date: "2023-08-22"
description: Proximal tubule subclusters analyses:
  - Subclusters of PT
  - Annotation of subclusters using the candidate markers
  - Psudotime of PT using monocle2, instead of monocle3
  - Basic statistics of PT
output: html_document
---
# Loading the packages
```{r}
# check the session information
sessionInfo()
library(dplyr)
library(Seurat)
library(patchwork)

#install.packages('devtools') #assuming it is not already installed

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
library(devtools)
#install_github('andreacirilloac/updateR')
# install.packages("sctransform")
# install.packages("ggthemes")
# loading the libarary
library(updateR)

library("sctransform")
library("ggthemes")
library(Seurat)

# BiocManager::install("limma")
# BiocManager::install("ComplexHeatmap")
# BiocManager::install("biomaRt")
# BiocManager::install("methrix")BiocManager::install("org.Hs.eg.db")
# devtools::install_github("saeyslab/nichenetr")
library(limma)
library(nichenetr)
library(tidyverse)
library(dplyr)

library(biomaRt)
library("data.table")


library(umap)
library(patchwork)
library(cowplot)
library(installr)

#install.packages("Matrix", repos = "http://cran.r-project.org")
### we used discre color palettes from ggsci

library("ggsci")
library("ggplot2")
library("gridExtra")

#options(buildtools.check = function(action) TRUE )
#devtools::install_url('https://cran.r-project.org/src/contrib/Archive/Matrix.utils/Matrix.utils_0.9.7.tar.gz')
#devtools::install_github('cole-trapnell-lab/monocle3')
#library(monocle3)

library(harmony)
library(Seurat)
#library(SeuratData)
library(tidyverse)

#library(BPCells)
library(ggplot2)
library("devtools")
library("AnnotationDbi")
library("org.Hs.eg.db")
#library(tibble)
#library(future)
library(here)
library(patchwork)
library(future)
#source("http://bioconductor.org/biocLite.R")
#biocLite("monocle")

#library(monocle3)

# install monocle

#BiocManager::install("monocle",force = TRUE)
library(monocle)
#devtools::install_dev("remotes")
#remotes::install_github('chris-mcginnis-ucsf/DoubletFinder')
#remotes::install_github(repo ='chris-mcginnis-ucsf/DoubletFinder' )
#install.packages("DoubletFinder")
library(DoubletFinder)
library(future.callr)
library(ComplexHeatmap)

library(RColorBrewer)
library(circlize)
library(monocle3)
#install.packages("magick")
library(magick)
#K means with 6 groups
library("scCustomize")

# install dyno

#devtools::install_github("dynverse/dyno")
library(dyno)
library(tidyverse)
```


## Step 2: Set up the workspace environment:
```{r}
#rm(list=ls())
setwd("/Users/XXW004/Documents/Projects/RuizRosado/PIPseq/Results/Evaluation_V022723/CellEvaluation_Subclustering_V0803/PT_Subclusters/")
Indir <-c("/Users/XXW004/Documents/Projects/RuizRosado/PIPseq/Results/Evaluation_V022723/CellEvaluation_Subclustering_V0803/")
Outdir <-c( "/Users/XXW004/Documents/Projects/RuizRosado/PIPseq/Results/Evaluation_V022723/CellEvaluation_Subclustering_V0803/PT_Subclusters/")
# setting up the seed 
set.seed(10000)
```

# Create 
```{r}
# create folders for DEGs, Progeny, GO, KEGG and TF analyses
# directory for DEGs
DEGDir <- paste0(Outdir,"DEGs/")
# directory for Progeny
ProgenyDir <- paste0(Outdir,"Progeny/")
# directory for GO enrichments
GOenrichDir <- paste0(Outdir,"GOenrich/")
# directory for Transcript Factors
TFDir <-  paste0(Outdir,"TranscriptFactors/")
# directory for KEGG
KEGGDir <-  paste0(Outdir,"KEGG/")
# directory for CellCommunication
CellCommDir <-  paste0(Outdir,"CellCommunication/")
# directory for pyseudotime
PyseudotimeDir <-  paste0(Outdir,"PyseudotimeMonocle2_Dyno/")

# directory for biomarkers

BioMarkerDir <- paste0(Outdir,"Biomarkers/")

dir.create(DEGDir, showWarnings = FALSE)
dir.create(ProgenyDir, showWarnings = FALSE)
dir.create(GOenrichDir, showWarnings = FALSE)
dir.create(TFDir, showWarnings = FALSE)
dir.create(KEGGDir, showWarnings = FALSE)
dir.create(CellCommDir, showWarnings = FALSE)
dir.create(BioMarkerDir, showWarnings = FALSE)
dir.create(PyseudotimeDir, showWarnings = FALSE)

set.seed(10000)
```
# generate the health cluster of PT cells

```{r}

# Read the preclustered object

scrna <- readRDS(file = paste0(Indir,"Pyelonephritis_singlecell_harmony_withAnnotation_v0803.RDS"))

# When you analyse the details please ignore the following analyse with Pipeline:
PTcells<- readRDS(file = "PIPSeq_Annotated_OnlyPT_v0823.rds")

DimPlot(PTcells)

# We only collect the PT cells
#CellType[CellType$cluster %in% c("1_1","0", "9","12","18", "19","3"),2] <- "PT" # "Proximal tubule"

PTcells @meta.data

# PTcells <- subset(scrna, subset = Raw_cell_type == "PT")
# 
# PTcells@meta.data
# # We then recluster the PT cells
# # recluster the cells
# PTcells <- SCTransform(PTcells, verbose = TRUE)
# PTcells <- RunPCA(PTcells,npcs = 5, verbose = TRUE)
# 
# # Run harmony for the PT cells
# PTcells <- RunHarmony(PTcells, 
# 				group.by.vars = c("DataSet"), 
# 				reduction = "pca", assay.use = "SCT", reduction.save = "harmony")
# PTcells <- RunUMAP(PTcells, dims = 1:5, verbose = TRUE, reduction = "harmony")
# PTcells <- FindNeighbors(PTcells, dims = 1:5,  verbose = TRUE, reduction = "harmony")
# 
# PTcells<- FindClusters(PTcells,  verbose = TRUE, resolution = 0.3)
# 
# PTcellsDim<- DimPlot(PTcells,label = T)
# PTcellsDim
# ggsave(filename = "PTcellsDim.pdf", plot = PTcellsDim, width = 8, height = 6)
# 
# PTcellsDimSplit<-DimPlot(PTcells, split.by = "DataSet",label = T)
# PTcellsDimSplit
# ggsave(filename = "PTcellsDimSplit.pdf", plot = PTcellsDimSplit, width = 24, height = 6)
# change the levels

Idents(PTcells)<-
# reorder the levels
ordered(Idents(PTcells), levels=rev(levels(Idents(PTcells))))

```

# Generate the figure for the PT
```{r}
# we only select the control:
PTcellControls  <- subset(PTcells, subset= (DataSet == "0DPI"))
PTcellControls@meta.data %>% head()
DimPlot(PTcellControls)

S1_S3_Markers<- c("Lrp2","Aqp1","Slc34a1","Ggt1","Atp1a1","Atp1b1","Slc5a2","Slc5a12","Slc22a8","Slc4a4","Slc13a3","Slc13a2","Slc22a6","Slc13a1","Slc22a2","Cubn","Slc22a30","Slc7a12","Slc22a7","Slc7a13","Aqp7")
S1_S3MarkerDotPlot<-DotPlot(PTcells, features =S1_S3_Markers,dot.min = 0.005, col.min = -6,
        col.max = 2, cols="RdBu")  + guides(color = guide_colorbar(title = 'AverExp'),
         size = guide_legend(title = "PercExp")) +  RotatedAxis()  
S1_S3MarkerDotPlot
# 
ggsave("S1_S3MarkerDotPlot_v0226.pdf",plot = S1_S3MarkerDotPlot, height = 3.5, width = 7)

S1_S3MarkerDotPlotControl<-DotPlot(PTcellControls, features =S1_S3_Markers,dot.min = 0.005, 
     cols="RdBu")  + guides(color = guide_colorbar(title = 'AverExp'),
         size = guide_legend(title = "PercExp")) +  RotatedAxis()  
S1_S3MarkerDotPlotControl
# 
ggsave("S1_S3MarkerDotPlotControl_v0226.pdf",plot = S1_S3MarkerDotPlotControl, height = 3.5, width = 6)

# Final Injury markers:
FinalInjuryMarkers<- c("Havcr1","Lcn2","Clu","Vcam1","Slc5a2","Agt", "Spp1","Fth1", "Dab2","Fn1","C3","Hsp90aa1", "Hsp90ab1","Hif1a","Gpx3","Hmox1","Nfkbia")
FinalInjuryMarkersDotPlot <-DotPlot(PTcells, features =FinalInjuryMarkers,dot.min = 0.005, cols="RdBu")  + guides(color = guide_colorbar(title = 'AverExp'),
         size = guide_legend(title = "PercExp")) +  RotatedAxis()  

ggsave("FinalInjuryMarkersDotPlotControl_v0226.pdf",plot = FinalInjuryMarkersDotPlot, height = 4, width = 6)

# Reduced PT markers
c(Lrp2,Aqp1, Slc34a1 )

```


# check the markers
```{r}
# the healthy cells
HealthMarkers<- c("Slc22a30", "Atp11a", "Trim7", "Slc22a6", "Slc7a13", "Slc5a2", "Slc5a10", "Spp2",  "Gatm", "Dcxr")

DotPlot(PTcells, features =  HealthMarkers, col.min = -0.5) +RotatedAxis()

CyclingMarker <- "Mki67"
DotPlot(PTcells, features =  CyclingMarker, col.min = -0.5) +RotatedAxis()

EarlyInjuryMarkers <- c("Cdh13", "Cdh6", "Havcr1")
DotPlot(PTcells, features =  EarlyInjuryMarkers, col.min = -0.5) +RotatedAxis()
LateInjuryMarkers <-c( "Ccl2", "Gm17268", "Cald1")
DotPlot(PTcells, features =  LateInjuryMarkers, col.min = -0.5) +RotatedAxis()

ReparingMarkers <- "Top2a"

DotPlot(PTcells, features =  ReparingMarkers, col.min = -0.5) +RotatedAxis()

FailedFepairMarkers <- c("Ypel2","Kcnip4")

DotPlot(PTcells, features =  FailedFepairMarkers, col.min = -0.5) +RotatedAxis()
# 
MaladptiveMarkers <- c("Vcam1", "Havcr1",  "Icam1", "Prom1","Cxcl2","Ccl3","Myc")

DotPlot(PTcells, features =  MaladptiveMarkers, col.min = -0.5) +RotatedAxis()

# Combined 
#"Lrp2","Slc34a1","Slc13a1"
KidneyStatuesMarkers<- c( "Slc5a2","Slc34a1","Inmt","Apom", "Slc7a13","Slc22a6", "Prom1","Top2a","Cyp7b1","Mki67","Cdh13", "Cdh6", "Havcr1","Ccl2", "Gm17268", "Cald1","Vcam1", "Ypel2","Kcnip4","Hsp90aa1","Col27a1","Plin2")

 DotPlot(PTcells, features =KidneyStatuesMarkers,dot.min = 0.005, col.min = -6,
        col.max = 2, cols="RdBu")  + guides(color = guide_colorbar(title = 'AverExp'),
         size = guide_legend(title = "PercExp")) +  RotatedAxis()  

# "Il1b", "Mlkl", "Gsdmd","Ripk3","Fn1", "Col1a1", "Col3a1", "Tgfb1","Vcam1", 
KidneyStatuesTypeMarkers<- c("Havcr1", "Icam1", "Prom1", "Spp1", "Cst3","Gdf15", "Clu", "Igfbp7", "Apoe", "Aldob", "B2m", "S100a6","S100a8","S100a9","Pdgfb","Bak1","Bad","Casp8","Ptgs2","Casp1","Hmox1")
 DotPlot(PTcells, features =  KidneyStatuesTypeMarkers) +RotatedAxis()
 
 KidneyStatuesTypeMarkers2<- c("Il1b", "Mlkl", "Gsdmd","Ripk3","Fn1", "Col1a1", "Col3a1", "Tgfb1","Vcam1", "Havcr1", "Icam1", "Prom1", "S100a6","S100a8","S100a9","Pdgfb","Bak1","Bad","Casp8","Ptgs2","Casp1","Hmox1","Nfkbia")
  DotPlot(PTcells, features =  KidneyStatuesTypeMarkers2) +RotatedAxis()

 #, cluster 5: Fn1, Havcr1, Cst3, Igfbp7, S100a8, Il1b, Slc7a13 : Profibrotic PT
 # cluster 1: Spp1, Prom1, Kcnip4, Col27a1,Top2a,Nfkbia: Cycling PT
 # cluster 8: Cdh13, Hsp90aa1,Inmt, Mki67 : Proliferating PT
 
 # cluster 1: ASpp1, Prom1, Kcnipp4: Progenitor PT
 
 # cluster 0: Aldob
```


# For the injury markers 
```{r}
InjuryMarkers<- c("Havcr1","Lcn2","Dab2","Clu","Vcam1","Slc5a2","Agt", "Spp1","Fth1")


TubularInjuryMarkerDotPlot <-DotPlot(PTcells, features =InjuryMarkers,dot.min = 0.005, cols="RdBu")  + guides(color = guide_colorbar(title = 'AverExp'),
         size = guide_legend(title = "PercExp")) +  RotatedAxis()  

ggsave("TubularInjuryMarkerDotPlotControl_v0226.pdf",plot = TubularInjuryMarkerDotPlot, height = 3.5, width = 6)

Inflammation_ImmuneResponseMarker <- c("C3","Vcam1","Il18","Tnfa","Il6","Cxcl1","Ccl2","Nlrp3")
TubularInjuryMarkerDotPlot <-DotPlot(PTcells, features =Inflammation_ImmuneResponseMarker,dot.min = 0.005, cols="RdBu")  + guides(color = guide_colorbar(title = 'AverExp'),
         size = guide_legend(title = "PercExp")) +  RotatedAxis()  

ggsave("Inflammation_ImmuneResponseDotPlotControl_v0226.pdf",plot = TubularInjuryMarkerDotPlot, height = 3.5, width = 6)

Fibrosis_CKDMarkers<- c("Col1a","Col3a1","Fn1","Tgfb1","Pdgfb","Acta2","Dab2","Mmp2","Mmp9")


Fibrosis_CKDMarkersDotPlot <-DotPlot(PTcells, features =Fibrosis_CKDMarkers,dot.min = 0.005, cols="RdBu")  + guides(color = guide_colorbar(title = 'AverExp'),
         size = guide_legend(title = "PercExp")) +  RotatedAxis()  

ggsave("Fibrosis_CKDMarkersDotPlotControl_v0226.pdf",plot = Fibrosis_CKDMarkersDotPlot, height = 3.5, width = 6)






PrerepairMarkers<- c("Hsp90aa1", "Hsp90ab1",  "Dab2")
DotPlot(PTcells, features =PrerepairMarkers,dot.min = 0.005, col.min = -6,
        col.max = 2, cols="RdBu")  + guides(color = guide_colorbar(title = 'AverExp'),
         size = guide_legend(title = "PercExp")) +  RotatedAxis() 

CellProliferation<- c("Top2a", "Mki67", "Lmnb1")
DotPlot(PTcells, features =CellProliferation,dot.min = 0.005, col.min = -6,
        col.max = 2, cols="RdBu")  + guides(color = guide_colorbar(title = 'AverExp'),
         size = guide_legend(title = "PercExp")) +  RotatedAxis() 

```


# the biomarkers for the 
```{r}
CandidateMarkers <- c("Aldob","Slc34a1","Spp1","Spp2","Lcn2", "Slc22a8","Slc5a2","Inmt","Prom1","Defb1","Top2a","Nfkbia", "Kcnip4", "Slc22a6", "Apoe","Havcr1", "Fn1","C3", "Il1b","S100a8", "Igfbp7", "Cst3", "Col27a1", "Hsp90aa1","Cdh13","Slc4a11","Slc3a1","Ren1","Slc12a1","Umod","Klk1")

CandidateMarkersDotPlot<-  DotPlot(PTcells, features =  CandidateMarkers, col.min = -2) +RotatedAxis()

ggsave("PT_CandidateMarkersDotPlot.pdf", plot = CandidateMarkersDotPlot, height = 6, width = 12)


# 


```


# biomarker from other cell types
```{r}
# Cfh: Parietal epithelial cells
Finalcelltypemarkers<- c("Lrp2","Slc34a1","Slc13a3", # Proximal tubule cells (PT)
                      "Slc12a3","Pvalb","Wnk1",# Distal connecting tublue (DCT)

                     "Umod","Slc12a1","Cldn10", # Thick ascending limb cells (TAL)
                      "Klk1","Slc8a1","Calb1", # Connecting tubule (CNT)
                     "Slc14a2","Fst","Bst1", # descending thin  limb cells (DTL)
                     #"Akr1b1","Sh3gl3","Prox1", # ascending thin limb cells (ATL)
                     #"Enox1","Thsd4", #Macula densa
                      "Aqp2","Hsd11b2","Scnn1g", # principal cells (PC)
                    "Atp6v1g3","Aqp6","Slc26a7", # Intercalated cells
                    "Krt5","Upk1b","Krt14", # Urothelium
                    "Cdh5","Igfbp3","Pecam1", # Endothelium
                    "Col3a1","Fbn1","Lum","Col1a1", # Fibroblast
                    "Ireb2","Alas2","C78334", # Plasmacytoid dentritic cells
                    "S100a8","S100a9","Il1b", # Neutrophil
                    "C1qa","C1qb","Aif1", # Macrophage
                    "Cxcr6","Cd247","Cd3e", # T cells
                    "Igkc","Cd79a","Cd79b", # B cells
                    "Ccl5","Nkg7","Cd7" # NK/Cd8 cells
)

OtherMarkerGenesDotPlots<- DotPlot(PTcells, features =  Finalcelltypemarkers, col.min = -0.5) +RotatedAxis()
ggsave(filename = "OtherMarkerGenesExpression_ClusterAnnotate.pdf", plot = OtherMarkerGenesDotPlots, width = 15, height = 8)

```

# #####################################
## generate the Trajectories of PT cells using monocle2, instead of monocle3, here we use dyno 
The dyno package offers end-users a complete TI pipeline. It features:

a uniform interface to 57 TI methods,
an interactive guideline tool to help the user select the most appropriate method,
streamlined interpretation and visualisation of trajectories, including colouring by gene expression or clusters, and
downstream analyses such as the identification of potential marker genes.
For information on how to install dyno, check out the installation instructions below.


# #####################################
# 2.2 Prepared cds the object
```{r}
# transfer Seurat object to cellDataset object

# build cds object
# set up the pathway
setwd(PyseudotimeDir)

# the expression
PTcells_expr_matrix = LayerData(PTcells, layer  = "counts")

# check expression matrix:

# Cell meta annotation -- phenotype
p_data <- PTcells@meta.data 
head(p_data)
rownames(PTcells_expr_matrix)
# Gene meta annotation
f_data<- data.frame(gene_short_name = rownames(PTcells_expr_matrix), row.names = rownames(PTcells_expr_matrix))
f_data
# UMIs or read ounts are better modele with the negative biomial.
pd <- new("AnnotatedDataFrame", data = p_data)
fd <- new("AnnotatedDataFrame", data = f_data)
PTcells_expr_matrix

CDS <- monocle::newCellDataSet(as.matrix(PTcells_expr_matrix), 
                       phenoData = pd, 
                       featureData = fd,
                       lowerDetectionLimit = 0.1,
                       expressionFamily=negbinomial.size())
CDS
CDS <- estimateSizeFactors(CDS)
CDS <- estimateDispersions(CDS)

# quality control:
## 起初是 24582 features, 768 samples 
#---------首先是对基因的过滤-------------
CDS <- detectGenes(CDS, min_expr = 0.1)
print(head(CDS@featureData@data))
expressed_genes <- row.names(subset(CDS@featureData@data,
                                    num_cells_expressed >= 10))
length(expressed_genes)
colnames(phenoData(CDS)@data)


# step1：dispersionTable()

disp_table <- dispersionTable(CDS)
unsup_clustering_genes <- subset(disp_table, 
                                 mean_expression >= 0.1)
CDS <- setOrderingFilter(CDS, unsup_clustering_genes$gene_id)
CDS
plot_ordering_genes(CDS) 
# 图中黑色的点就是被标记出来一会要进行聚类的基因
CDS[expressed_genes,]
# step2 plot pc variance explained
plot_pc_variance_explained(CDS, return_all = F) # norm_method='log'



## Trajactory step 1: choose genes that define a cells' progress that compare the infection vs noninfection within the infected timepoints
# First, we must decide which genes we will use to define a cell's progress through myogenesis. We ultimately want a set of genes that increase (or decrease) in expression as a function of progress through the process we're studying.

#Ideally, we'd like to use as little prior knowledge of the biology of the system under study as possible. We'd like to discover the important ordering genes from the data, rather than relying on literature and textbooks, because that might introduce bias in the ordering. We'll start with one of the simpler ways to do this, but we generally recommend a somewhat more sophisticated approach called "dpFeature".

# One effective way to isolate a set of ordering genes is to simply compare the cells collected at the beginning of the process to those at the end and find the differentially expressed genes, as described above. The command below will find all genes that are differentially expressed in response to the switch from growth medium to differentiation medium:

# here we will detect the genes based on the series of timepoints 
diff_test_res <- differentialGeneTest(CDS[expressed_genes,],
              fullModelFormulaStr = "~DataSet")
diff_test_res%>% arrange(qval) %>% head()
nrow(diff_test_res)
diff_test_res 
# these will order the gene from our choose from control to infection
ordering_genes <- row.names (subset(diff_test_res, qval < 0.01))

length(ordering_genes)
# we also select the genes with high dispersion across cells
dis_table<- dispersionTable(CDS)

ordering_genes_2 <- subset(dis_table,
                  mean_expression >= 0.05 & dispersion_empirical >= 0.05 * dispersion_fit)$gene_id
length(ordering_genes_2)
Final_ordering_genes<- intersect(ordering_genes,ordering_genes_2 )
write.csv(Final_ordering_genes, file = "Final_ordering_genes.csv")

length(Final_ordering_genes)
CDS <- setOrderingFilter(CDS, Final_ordering_genes)
plot_ordering_genes(CDS)

# reduce data dimention using DDRTree 
CDS <- reduceDimension(CDS, max_components = 2,
    method = 'DDRTree')
# trajectory step3: order cells along the trajectory
CDS <- monocle::orderCells(CDS)


 
??GM_state
#install.packages("igraph")

plot_cell_trajectory(CDS, color_by = "SCT_snn_res.0.3", cell_size = 0.5)
plot_cell_trajectory(CDS, color_by = "DataSet",cell_size=0.5)
?plot_cell_trajectory
# order by 


# "State" is just Monocle's term for the segment of the tree. The function below is handy for identifying the State which contains most of the cells from time zero. We can then pass that to orderCells:
unique(CDS@phenoData@data$SCT_snn_res.0.3)
T0_counts <- table(CDS@phenoData@data$SCT_snn_res.0.3, CDS@phenoData@data$DataSet)[,"0DPI"]
as.numeric(names(T0_counts)[which
          (T0_counts == max(T0_counts))])
table(CDS@phenoData@data$SCT_snn_res.0.3, CDS@phenoData@data$DataSet)[,"0DPI"]
GM_state <- function(CDS){
  if (length(unique(CDS@phenoData@data$SCT_snn_res.0.3)) > 1){
    T0_counts <- table(CDS@phenoData@data$SCT_snn_res.0.3, CDS@phenoData@data$DataSet)[,"0DPI"]
    return(as.numeric(names(T0_counts)[which
          (T0_counts == max(T0_counts))]))
  } else {
    return (1)
  }
}
GM_state(CDS)
#CDS2 <- orderCells(CDS, root_state = 1)
plot_cell_trajectory(CDS, color_by = "Pseudotime")


# 
BEAM_res <- BEAM(CDS, branch_point = 1, cores = 1)
BEAM_res <- BEAM_res[order(BEAM_res$qval),]
BEAM_res <- BEAM_res[,c("gene_short_name", "pval", "qval")]







# normalize and preproces the data
cds <- preprocess_cds(pre_cds) #标准化+PCA降维
plot_pc_variance_explained(cds)
# remove batch effects
## Step 2: Remove batch effects with cell alignment
cds <- align_cds(cds, alignment_group = "DataSet")

```

# 2.2 Umap by Monocle3
```{r}
#  Step 3: Reduce the dimensions using UMAP
cds <- reduce_dimension(cds)
## Step 4: Cluster the cells
cds <- cluster_cells(cds)

#plot_cells(cds, label_groups_by_cluster=FALSE,  color_cells_by = "partition")

head(cds@int_colData$reducedDims$UMAP)

plot_cells(cds, reduction_method="UMAP", 
           show_trajectory_graph = FALSE,
           label_cell_groups = FALSE, 
           color_cells_by="SCT_snn_res.0.3")
dev.off()
# UMAP by Seurat
seurat_umap <- Embeddings(PTcells, reduction = "umap")[colnames(cds),]
cds@int_colData$reducedDims$UMAP <- seurat_umap
head(cds@int_colData$reducedDims$UMAP)
plot_cells(cds, reduction_method="UMAP", 
           show_trajectory_graph = FALSE,
           label_cell_groups = FALSE, 
           color_cells_by="SCT_snn_res.0.3")

```

# Draw the monocle state
```{r}
#分群(类似monocle的State)

#cds <- cluster_cells(cds)

#预测轨迹
cds <- learn_graph(cds)
#交互式确定root节点，可以选择多个。我这里选择了一个
cds <- order_cells(cds) 

PT_pseudotime <-plot_cells(cds, color_cells_by = "pseudotime", 
           label_cell_groups = FALSE, 
           label_leaves = TRUE,  
           label_branch_points = FALSE)
PT_pseudotime
ggsave(filename = "PT_pseudotime_MonocaCluster_initial_v02122025.pdf", plot = PT_pseudotime, width = 4, height = 3)


```
# without any choose root
```{r}
#分群(类似monocle的State)
### Step 5: Learn a graph
cds <- cluster_cells(cds)

#预测轨迹
cds <- learn_graph(cds)
#交互式确定root节点，可以选择多个。我这里选择了一个
#cds <- order_cells(cds) 
# a helper function to identify the root principal points:
get_earliest_principal_node <- function(cds, time_bin="0DPI"){
  cell_ids <- which(colData(cds)[, "DataSet"] == time_bin)
  closest_vertex <-
  cds@principal_graph_aux[["UMAP"]]$pr_graph_cell_proj_closest_vertex
  closest_vertex <- as.matrix(closest_vertex[colnames(cds), ])
  root_pr_nodes <-
  igraph::V(principal_graph(cds)[["UMAP"]])$name[as.numeric(names
  (which.max(table(closest_vertex[cell_ids,]))))]
  
  root_pr_nodes
}
cds <- order_cells(cds, root_pr_nodes=get_earliest_principal_node(cds))


PTCells_pseudotime <-plot_cells(cds, color_cells_by = "pseudotime", 
           label_cell_groups = FALSE, 
           label_leaves = TRUE,  
           label_branch_points = FALSE,graph_label_size=1.5)
PTCells_pseudotime
ggsave(filename = "PTCells_pseudotime_MonocaCluster_V02122025.pdf", plot = PTCells_pseudotime, width = 4, height = 3)



```


```{r}
# Determine the psedudotime related genes
Track_genes <- graph_test(cds, neighbor_graph="principal_graph")

Track_genes %>% head()

```

# plot the expression of interesting genes during the PT pseudotime
```{r}

# re-order the tracked genes by the q value and morans I
Track_genes <- Track_genes[,c(5,2,3,4,1,6)] %>% 
  dplyr::arrange(desc(morans_I),q_value)

# Plot some interesting genes for the PT pseudotime : Top 2 genes, Napsa, Kap, Ttr, Igfbp7, Havcr1 and Prom1
KapTrack <- plot_genes_in_pseudotime(cds[Track_genes$gene_short_name[1],] , 
                         min_expr=0.5)
ggsave(filename = "PT_pseudotime_KapTrack.pdf", plot = KapTrack, width = 5, height = 3)
NapsaTrack <- plot_genes_in_pseudotime(cds[Track_genes$gene_short_name[2],] , 
                         min_expr=0.5)
ggsave(filename = "PT_pseudotime_NapsaTrack.pdf", plot = NapsaTrack, width = 5, height = 3)

TtrTrack <- plot_genes_in_pseudotime(cds["Ttr",] , 
                         min_expr=0.5)
ggsave(filename = "PT_pseudotime_TtrTrack.pdf", plot = TtrTrack, width = 5, height = 3)

Igfbp7Track <- plot_genes_in_pseudotime(cds["Igfbp7",] , 
                         min_expr=0.5)
ggsave(filename = "PT_pseudotime_Igfbp7Track.pdf", plot = Igfbp7Track, width = 5, height = 3)

Track_genes$gene_short_name[1:5]


Track_genes$gene_short_name[2]
Track_genes$gene_short_name[1]
Havcr1Track <- plot_genes_in_pseudotime(cds["Havcr1",] , 
                         min_expr=0.5)

ggsave(filename = "PT_pseudotime_Havcr1Track.pdf", plot = Havcr1Track, width = 5, height = 3)

# For the 
Prom1Track <- plot_genes_in_pseudotime(cds["Prom1",],min_expr=0.5 )
ggsave(filename = "PT_pseudotime_Prom1Track.pdf", plot = Prom1Track, width = 5, height = 3)

FeaturePlot(PTcells, features = "Havcr1")

```

# choose the top signifcant for visualization
```{r}
diff_pseudo_gene <- Track_genes %>% 
  dplyr::arrange(q_value) %>%
  # remove the genes with mitochondrion
  filter(!grepl('^mt-|Hbb-', gene_short_name)) %>%
  rownames() %>% head(60)

# we ignore the mitochondria genes



diff_pseudo_gene

# generate the heatmap for those genes

diff_pseudo_gene.matrix <- exprs(cds)[match(diff_pseudo_gene,rownames(rowData(cds))),order(pseudotime(cds))]

diff_pseudo_gene.matrix <- t(apply(diff_pseudo_gene.matrix,1,function(x){smooth.spline(x,df=3)$y}))
diff_pseudo_gene.matrix <- t(apply(diff_pseudo_gene.matrix,1,function(x){(x-mean(x))/sd(x)}))
rownames(diff_pseudo_gene.matrix) <- diff_pseudo_gene;

#pheatmap(diff_pseudo_gene.matrix)

# because the plot_pseudotime_heatmap no longer have, we used the complexheatmap to draw the figure
# Reference: https://github.com/cole-trapnell-lab/monocle-release/issues/295
pdf("PT_pseudotime_psudotimeGenes.pdf", height = 6, width = 8)
htkm <- Heatmap(
  diff_pseudo_gene.matrix,
  name    = "z-score",
  col     = colorRamp2(seq(from=-2,to=2,length=11),rev(brewer.pal(11, "Spectral"))),
  show_row_names  = TRUE,
  show_column_names = FALSE,
  row_names_gp  = gpar(fontsize = 6),
  km = 8,
  row_title_rot = 0,
  use_raster=TRUE,
  cluster_rows  = TRUE,
  cluster_row_slices = FALSE,
  cluster_columns   = FALSE)

print (htkm)
dev.off()
#ggsave(filename = "PT_pseudotime_psudotimeGenes.pdf", plot = htkm, width = 8, height = 6)


#Ward.D2 Hierarchical Clustering
pdf("PT_pseudotime_psudotimeGenes_Hierarchicalcluster.pdf", height = 7, width = 6)
hthc <- Heatmap(
  diff_pseudo_gene.matrix,
  name                         = "z-score",
  col                          = colorRamp2(seq(from=-2,to=2,length=11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 6),
  clustering_method_rows = "ward.D2",
  clustering_method_columns = "ward.D2",
  row_title_rot                = 0,
  use_raster= TRUE,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE)

print(hthc)
dev.off()

cds["Havcr1",] %>% head()
# Generate the Dotplot for these genes
plot_genes_in_pseudotime(cds[c("Havcr1"),] , 
                         min_expr=0.5)

DotPlot(PTcells, features =diff_pseudo_gene ) +RotatedAxis()

DotPlot(PTcells, features =c("Spp1","Havcr1","Itgb1","Top2a",
                             "Prom1","Pdgfb","Defb1","Fn1","Col6a1"), col.min = -0.5 ) +RotatedAxis()

FeaturePlot(PTcells, features = c("Lcn2","Defb1","Egf","Havcr1"))
# the following no longer have 

Track_genes
# p = plot_pseudotime_heatmap(cds[diff_pseudo_gene,], 
#                         num_clusters = 3,  # default 6
#                         return_heatmap=T)
# 

```




# generate the subclusters of PT, DCT, IC, PC
```{r}
# dividing into cycling, transitioning, adap- tive (successful or maladaptive repair) and degenerative (damaged or stressed). 
# cycling: Mki67
# Early Injury: Cdh13, Cdh6, Havcr1
# Late Injury: Ccl2, Gm17268, Cald1
# conserved: Slc13a1, Slc5a2,  Slc7a13, Cyp7b1



# Panoptosis
# Il1b, MLKL, Gsdmd,Ripk3

# Fibrosis 
# Fn1, Col1a1, Col3a1, Tgfb1
# adaptive
# VCAM1, DCDC2 and HAVCR1, C3, ICAM1
# VCAM1 PROM1 ITGB8
# Vcam1, Havcr1, Dcdc2, Icam1, Prom1, ITgb8

#degenerative
# SPP1, CST3, CLU and IGFBP7, APOE ALDOB
# B2M S100A6
# Spp1, Cst3, GDF15,Clu, Igfbp7, Apoe, Aldob, B2m, S100a6
# agionegious PEC,
# EGFR, VEGF, Cdh5

# Egfr, Vegf, Cdh5
KidneyStatuesMarkers<- c("Slc13a1", "Slc5a2", "Slc7a13", "Cyp7b1","Mki67","Cdh13", "Cdh6", "Havcr1","Ccl2", "Gm17268", "Cald1")

KidneyStatuesTypeMarkers<- c("Il1b", "Mlkl", "Gsdmd","Ripk3","Fn1", "Col1a1", "Col3a1", "Tgfb1","Vcam1", "Havcr1", "Icam1", "Prom1", "Spp1", "Cst3","Gdf15", "Clu", "Igfbp7", "Apoe", "Aldob", "B2m", "S100a6")

DotPlot(scrna, features =  KidneyStatuesMarkers, col.min = -0.5) +RotatedAxis()
DotPlot(scrna, features =  KidneyStatuesTypeMarkers, col.min = -0.5) +RotatedAxis()


for (i in KidneyStatuesMarkers){
# 
  FeaturePlots<-FeaturePlot(scrna, features =  i, split.by = "DataSet" ) +RotatedAxis()
  ggsave(paste0(Outdir,"/KidneyStatuesMarkers/", i, "_Feature_GeneMarkers_Annoatation.pdf"), plot =FeaturePlots, height =6, width = 14)
}

for (i in KidneyStatuesTypeMarkers){
# 
  FeaturePlots<-FeaturePlot(scrna, features =  i, split.by = "DataSet" ) +RotatedAxis()
  ggsave(paste0(Outdir,"/KidneyStatuesTypeMarkers/", i, "_Feature_GeneMarkers_Annoatation.pdf"), plot =FeaturePlots, height =6, width = 14)
}


# with the inflammation markers

# with the fibrosis markers

# with the apoptosis markers

# with the necrosis markers

# with the endothelial markers


```


