---
title: "SpatialDEGFigure_RegionalKidney"
author: "Xin Wang"
date: "2023-07-20"
output: html_document
---
---
title: "Differentially expressed genes detection for comparisions within timepoints to control in spatial transcriptomic datasets using all the niches"
output: html_document
Author: Xin Wang
Email: xin.wang@nationwidechildrens.org 
Copyright (c) 2023 Kidney and Urology Tract Center
Nationwide Children's Hospital
date: "2023-05-18"
Description: 
#   Detection of differentially expressed genes for different infection regions
#    - Detection of DEGs comparing the clusters
#    - Detection of DEGs under several different conditions
#    - Functional enrichments for the DEGs
#    - Several of associated figures to show the differences
#    - KEGG pathway enrichments
#    - Heatmap, spatialfeature plots, pathview of interesting KEGG pathways.

note: the genes including the list of
       cell type: Proximal tubule, connecting tubule, IC, PC, Loop of henle, Endothelium, NK, CD45, T cells, B Cells, Macrophage, Neutorphil

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE, cache.lazy = TRUE)
```


## loading the libraries

```{r libraries, cache=FALSE, warning=FALSE, error=FALSE, message=FALSE,}

# loading libraries required for seurat
library(Seurat)
#library(SeuratData)
library(ggplot2)
library(patchwork)
library(dplyr)
library(Matrix)
library(tidyverse)
library(SeuratObject)
#install.packages("harmony")
library(harmony)
library(SPOTlight)
# library(scater)
# library(scran)
library(SingleCellExperiment)

#install.packages("SeuratDisk")
#library(SeuratDisk)

# loading for the table bubble updates
library(L1pack)
library(reshape)
library(tidyverse)
library(stringr)
library(cowplot)

# loading the libraries for plot and colors
library(ggplot2)
library(reshape2)
library(Hmisc)
library(stats)
library(corrplot)
library(RColorBrewer)
## Deconvolve
#install.packages("L1pack")
#install.packages("reshape")

# loading libraries required for the DEG and functionally enrichment
#install.packages("msigdbr")
library(msigdbr)
library(fgsea)
library(data.table)
library(RColorBrewer)
library(ggrepel)
library(DESeq2)
library(pheatmap)
#BiocManager::install("MAST")
library(MAST)
library(edgeR)
library(clusterProfiler)
library(org.Mm.eg.db)
library(enrichplot)
#keytypes(org.Mm.eg.db) 

#BiocManager::install("ReactomePA")
library(ReactomePA)

# install thje KEGG packages
#MusKEGG<- search_kegg_organism('Mus musculus', by='scientific_name')
#dim(MusKEGG)
#head(MusKEGG)
# loading for KEGG analyes
library("pathview")


library(RColorBrewer)
myPalette <- colorRampPalette(rev(brewer.pal(11, "Spectral")))
#devtools::install_github("jinworks/CellChat")
library(CellChat)
```


# The main analyses section:
# setting up the environments and randomization
```{r}
## set directory pathway:

#rm(list=ls())
setwd("/Users/XXW004/Documents/Projects/RuizRosado/SpatialTranscriptomic/Results/NichesDEGs_V04222025/KidneyInfectedVsOther/")
Indir = "/Users/XXW004/Documents/Projects/RuizRosado/SpatialTranscriptomic/Results/Deconvolution_Six_V09062023/"
# Here is the Kidney different regions
Outdir =c("/Users/XXW004/Documents/Projects/RuizRosado/SpatialTranscriptomic/Results/NichesDEGs_V04222025/KidneyInfectedVsOther/")

# Generate the Different folders to store the results
DEGDir <- paste0(Outdir,"DEGs/")
# directory for Progeny
ProgenyDir <- paste0(Outdir,"Progeny/")
# directory for GO enrichments
GOenrichDir <- paste0(Outdir,"GOenrich/")
# directory for Transcript Factors
TFDir <-  paste0(Outdir,"TranscriptFactors/")
# directory for KEGG
KEGGDir <-  paste0(Outdir,"KEGG/")
# directory for CellCommunication
CellCommDir <-  paste0(Outdir,"CellCommunication/")
# directory for pyseudotime
PyseudotimeDir <-  paste0(Outdir,"Pyseudotime/")

# directory for biomarkers

BioMarkerDir <- paste0(Outdir,"Biomarkers/")

dir.create(DEGDir, showWarnings = FALSE)
dir.create(ProgenyDir, showWarnings = FALSE)
dir.create(GOenrichDir, showWarnings = FALSE)
dir.create(TFDir, showWarnings = FALSE)
dir.create(KEGGDir, showWarnings = FALSE)
dir.create(CellCommDir, showWarnings = FALSE)
dir.create(BioMarkerDir, showWarnings = FALSE)
dir.create(PyseudotimeDir, showWarnings = FALSE)
# setting up the seed 
set.seed(1000001)
```

# read the input object that have been well convoluted and annotated
```{r}
Pyelonephritis_Infection <- readRDS(file = paste0(Indir,"Pyelonephritis_Infection_Deconvoluted.RDS"))

rownames(Pyelonephritis_Infection)
colnames(Pyelonephritis_Infection)
# define the selcted slices
selectedslices<- c("slice1","slice3","slice4","slice7","slice8","slice9")

Pyelonephritis_Infection @meta.data $orig.ident
# define infected niches
Infectedniches <- c("N8","N10","N11","N12","N13","N14","N15","N16","N17")

DimPlot(Pyelonephritis_Infection)

Pyelonephritis_Infection@meta.data %>% pull(orig.ident) %>% table()
# Infected days and non-infected data
InfectedDate <- c("PyeloD1_2","PyeloD3_1","PyeloD5_2","PyeloD7_1", "PyeloD28_1")

```


# Detection all the DEGs for all the regions and only show the top 10 for N8, N10-N17
```{r}
setwd(BioMarkerDir)
DefaultAssay(Pyelonephritis_Infection) <- "SCT"
InfectedDate
#
Pyelonephritis_Infection@meta.data %>%  pull(group) %>% table()
# create an meta data
Pyelonephritis_Infection@meta.data<-
  Pyelonephritis_Infection@meta.data %>% mutate(
  InfectedNICH =
    case_when(Annotation %in% Infectedniches ~ "InfectedNiches",
              TRUE ~ "NonInfectedNiches"),
  InfectedDate =
    case_when(group %in% InfectedDate ~ "InfectionDays",
              TRUE ~ "NonInfectionDays"),
)

# we identify the DEG comparing between Infected and Noninfected niches, and group by the infected date and non infected date
Pyelonephritis_Infection @meta.data$InfectedNICH %>% table()
Pyelonephritis_Infection @meta.data$InfectedDate %>% table()
Idents(Pyelonephritis_Infection) <- "InfectedNICH"

# check the markers that differentially expressed between infected niches and non-infected niches
AllCategories_Clustermarkers_infectedvsnoninfected <- FindAllMarkers(Pyelonephritis_Infection,
            min.pct = 0.25,test.use= "MAST") 
# save the niches markers into a file
write.csv(AllCategories_Clustermarkers_infectedvsnoninfected, "AllCategories_Clustermarkers_MAST_infectedvsnoninfected.csv")
# AllCategories_Clustermarkers_infectedvsnoninfected <- read.csv("AllCategories_Clustermarkers_MAST_infectedvsnoninfected.csv")
# draw the figure for the top 40 DEGs
Top30MarkerGeneInfected<-AllCategories_Clustermarkers_infectedvsnoninfected %>% as.data.frame() %>% group_by(cluster) %>% slice_min(order_by = p_val_adj, n=100) %>% slice_max(order_by = avg_log2FC, n=20)

# get the DEGs 
DEGsInfectedRegions<- AllCategories_Clustermarkers_infectedvsnoninfected %>% as.data.frame() %>% group_by(cluster) %>% filter(cluster == "InfectedNiches" & p_val_adj<=0.05 & (avg_log2FC >1|avg_log2FC < -1) )

# InfectedhighConservedGenes <- FindConservedMarkers(Pyelonephritis_Infection, ident.1 = "InfectedNiches", grouping.var ="InfectedNiches")
nrow(DEGsInfectedRegions)
# we identify the highly conserved genes 
Pyelonephritis_Infection@meta.data
Idents(Pyelonephritis_Infection) <- "Annotation"
DefaultAssay(Pyelonephritis_Infection)

Top_InfectedrmarkersComparingAllGenesheatmap<- DoHeatmap(Pyelonephritis_Infection, features = Top30MarkerGeneInfected$gene, size = 2.5)  + scale_fill_gradientn(colors = c("lightblue", "white", "FireBrick")) +  theme(legend.position="bottom") 

Top_InfectedrmarkersComparingAllGenesheatmap
ggsave("Top_InfectedrmarkersComparingAllGenesheatmap.pdf", plot =Top_InfectedrmarkersComparingAllGenesheatmap, height = 8, width = 7 )

# we then find DEGs for each cluster comparing infected and non infected date
Idents(Pyelonephritis_Infection) <- Pyelonephritis_Infection@meta.data$InfectedDate

# Identify the markers that increased after the infection
AllCategories_Clustermarkers_infectedvsnoninfectedDate <- FindAllMarkers(Pyelonephritis_Infection,
            min.pct = 0.25,test.use= "MAST") 
# save the date markers into a file
write.csv(AllCategories_Clustermarkers_infectedvsnoninfectedDate, "AllCategories_Clustermarkers_infectedvsnoninfectedDate.csv")
AllCategories_Clustermarkers_infectedvsnoninfectedDate <- read.csv("AllCategories_Clustermarkers_infectedvsnoninfectedDate.csv")
# the DEGs comparing between infected dates and day 0
DEGsInfectedDays <- AllCategories_Clustermarkers_infectedvsnoninfectedDate %>% filter(cluster== "InfectionDays" & p_val_adj<=0.05 & (avg_log2FC >1|avg_log2FC < -1) )


nrow(DEGsInfectedDays)
# Check the overlapping gene that differentiated after the infection, as well as enriched in the infection regions
# draw the venn figure

#install.packages("ggVennDiagram")
library(ggVennDiagram)
library(ggvenn)

Deggene_list <- list(
  spatialregionaldegs = DEGsInfectedRegions$gene,
  spatialdatedegs = DEGsInfectedDays$gene
)
ggvenn(
  Deggene_list,
  fill_color = c("skyblue", "orange"),
  stroke_size = 0.5,
  set_name_size = 5
)
#ggVennDiagram(Deggene_list, label_alpha = 0)

# Generate the overlapping genes
FinalOverlappingGenes<- DEGsInfectedRegions %>% filter(gene %in% DEGsInfectedDays$gene)

# Generate overlapping activated genes

FinalUpregulatedGenes<- FinalOverlappingGenes %>% filter(avg_log2FC>1)
FinalDowmregulatedGenes<- FinalOverlappingGenes %>% filter(avg_log2FC< -1)
nrow(FinalUpregulatedGenes)
nrow(FinalDowmregulatedGenes)
# Draw the supplementary fig for the conserved 
Tope15FinalUpregulatedGenes <- FinalUpregulatedGenes%>% as.data.frame() %>% group_by(cluster) %>% slice_min(order_by = p_val_adj, n=100) %>% slice_max(order_by = avg_log2FC, n=15)
Tope15FinalUpregulatedGenes %>% head()
Tope15FinalDownregulatedGenes <- FinalDowmregulatedGenes%>% as.data.frame() %>% group_by(cluster) %>% slice_min(order_by = p_val_adj, n=100) %>% slice_min(order_by = avg_log2FC, n=15)
Top30FinalDEGs<- c(Tope15FinalUpregulatedGenes$gene, Tope15FinalDownregulatedGenes$gene)
Idents(Pyelonephritis_Infection) <- "Annotation"

Top30FinalDEGsGenesheatmap<- DoHeatmap(Pyelonephritis_Infection, features = Top30FinalDEGs, size = 2.5)  + scale_fill_gradientn(colors = c("lightblue", "white", "FireBrick")) +  theme(legend.position="bottom")   

Top30FinalDEGsGenesheatmap
ggsave("Top30FinalDEGsGenesheatmap.pdf", plot =Top30FinalDEGsGenesheatmap, height = 8, width = 7 )
```

# Generate the GO enrichment of these conserved DEGs
```{r}
FinalUpregulatedGenesGO<- enrichGO(FinalUpregulatedGenes$gene, OrgDb = "org.Mm.eg.db", ont='BP',pAdjustMethod = 'BH',pvalueCutoff = 1, qvalueCutoff = 1,keyType = 'SYMBOL')
#FinalUpregulatedGenesGO %>% head()
FinalDownregulatedGenesGO<- enrichGO(FinalDowmregulatedGenes$gene, OrgDb = "org.Mm.eg.db", ont='BP',pAdjustMethod = 'BH',pvalueCutoff = 1, qvalueCutoff = 1,keyType = 'SYMBOL')

FinalUpregulatedGenesGOSimplified<- clusterProfiler::simplify(FinalUpregulatedGenesGO)
FinalDownregulatedGenesGOSimplified<- clusterProfiler::simplify(FinalDownregulatedGenesGO)

# save into file
write.csv(FinalUpregulatedGenesGO,"FinalUpregulatedGenesGO.csv")
write.csv(FinalDownregulatedGenesGO,"FinalDownregulatedGenesGO.csv")
write.csv(FinalUpregulatedGenesGOSimplified,"FinalUpregulatedGenesGOSimplified.csv")
write.csv(FinalDownregulatedGenesGOSimplified,"FinalDownregulatedGenesGOSimplified.csv")

FinalUpregulatedGenesGOSimplified
# draw the top barplot
barplot(FinalUpregulatedGenesGOSimplified) 
?
barplot(FinalDownregulatedGenesGOSimplified)
```

# read the single cell RNAseq and bulk micro-array RNA
```{r}
Top30FinalDEGs
# microarray
MicroDir = "/Users/XXW004/Documents/Projects/BrianBecknell/GrantApplication/materials"
MouseMicroArrayPyelD7vsD0 <- read.csv("/Users/XXW004/Documents/Projects/BrianBecknell/GrantApplication/materials/HeOuJ 7 dpi vs 0 dpi.csv")
MouseMicroArrayDEGs<- MouseMicroArrayPyelD7vsD0 %>%  filter(adj.P.Val <=0.05 & (logFC >1 |logFC < -1)) 
MouseMicroArrayDEGs %>% nrow()
MouseMicroArrayDEGs %>% filter(mgi_symbol %in% (c("Vcam1","Ccl7","Mmp8","Arg1","Mmp13","Pf4","Ccl3","Ccl4","Ccl5","Cxcl3","Xdh","Il6","Il1b","Il23","Cxcl1","Il1rn","Tnf","Ctss")))
# single cell directory:
Scindir = "/Users/XXW004/Documents/Projects/RuizRosado/PIPseq/Results/Evaluation_V022723/CellEvaluation_Subclustering_V0803/FinalSingleCellObj_V08012024/"
# for the total cells
SinglecellObj <- readRDS(paste0(Scindir,"PIPSeq_Final_Annotated_v08012024.rds"))

dev.off()
FeaturePlot(SinglecellObj, features = c("Lcn2","Vcam1","Clu"), split.by = "DataSet")
# Identify the DEGs comparing the infected day 7 vs uninfected date
SinglecellObj@meta.data %>% pull(Condition) %>% table()
SinglecellObj %>% head()
 
SinglecellObjSub <- subset (SinglecellObj,subset= (DataSet %in% c("0DPI", "7DPI")))
SinglecellObjSub2 <- subset (SinglecellObj,subset= (DataSet %in% c("3DPI", "7DPI")))

# we used the overall DEGs between infection and noninfection. 
SingleCellOverallDEGs2 <- FindMarkers(SinglecellObjSub2, ident.1 = "7DPI", ident.2 = "3DPI", group.by = "DataSet", test.use= "MAST",min.pct =0)
# please check the DEGs of "Napsa,"Apom", F13a1, Lmnb1, Col5a1, Vcam1, Clu, Lcn2, 
write.csv(SingleCellOverallDEGs2, file = "SingleCellOverallDEGs_7vs3day.csv")
saveRDS(SingleCellOverallDEGs2, file = "SingleCellOverallDEGs_7vs3day.rds")
write.csv(SingleCellOverallDEGs, file = "SingleCellOverallDEGs.csv")

SingleCellOverallDEGs2[c("Vcam1","Ccl7","Mmp8","Arg1","Mmp13","Lcn2","Clu"),]
# to ensure the data have well normalized, I performed the normalization and scale again
SinglecellObjSub <- NormalizeData(SinglecellObjSub, normalization.method = "LogNormalize", scale.factor = 10000)
SinglecellObjSub <- FindVariableFeatures(SinglecellObjSub, selection.method = "vst", nfeatures = 2000)
SinglecellObjSub <- ScaleData(SinglecellObjSub, features = VariableFeatures(SinglecellObjSub))


CellTypes <- levels(as.factor(SinglecellObjSub$FinalAnnotation))
Idents(SinglecellObjSub)
#SinglecellObjSub %>% head()
SingleCellDEGsFinal<- list()
for (i in CellTypes) {
  SingleCellDEGs <-
    FindMarkers(
      SinglecellObjSub,
      ident.1 = "Infect",
      ident.2 = "Noninfect",
      subset.ident = i,
      group.by = "Condition",
      test.use = "MAST",
      min.pct = 0
    )
  ? FindMarkers
  SingleCellDEGsFinal[[i]] <-
    SingleCellDEGs %>% filter(p_val_adj <= 0.05 &
                                (avg_log2FC > 1 | avg_log2FC < -1))
  write.csv(SingleCellDEGsFinal[[i]], file = paste0(i, "singlecell.DEGs.csv"),)
}

SingleCellDEGsFinal
# save them into the list
saveRDS(SingleCellDEGsFinal, file = "SingleCellDEGsFinal.rds")

# we used the overall DEGs between infection and noninfection. 
SingleCellOverallDEGs <- FindMarkers(SinglecellObjSub, ident.1 = "Infect", ident.2 = "Noninfect", group.by = "Condition", test.use= "MAST",min.pct =0)
write.csv(SingleCellOverallDEGs, file = "SingleCellOverallDEGs.csv")

saveRDS(SingleCellOverallDEGs, "SingleCellOverallDEGs.RDS")
SingleCellOverallDEGs<-readRDS("SingleCellOverallDEGs.RDS")
SingleCellOverallFinalDEGs <- SingleCellOverallDEGs %>%  filter(p_val_adj <= 0.05 &(avg_log2FC > 1 | avg_log2FC < -1))
nrow(SingleCellOverallFinalDEGs)
 #nrow(SingleCellDEGsFinal)
SingleCellOverallDEGs[c("Vcam1","Ccl7","Mmp8","Arg1","Mmp13"),]

FeaturePlot(SinglecellObjSub, features = "Vcam1", split.by = "Condition")
SingleCellOverallFinalDEGs
# we initially considered to include those immune related degs
# # we also used the degs from immune cell differentially expressed genes
# # read the immune cells objects
# ImmuneScindir = "/Users/XXW004/Documents/Projects/RuizRosado/SpatialTranscriptomic/Results/Deconvolution_Six_Immune_V12092024/"
# 
# Immune_reference <- readRDS(file = paste0(ImmuneScindir, "CD45_ImmuneCell_withSubCluster.rds"))
# 
# # we only selected day 7 vs controls
# Immune_referenceSub <- subset (Immune_reference,subset= (orig.ident %in% c("0DPI", "7DPI")))
# 
# # to ensure the data have well normalized, I performed the normalization and scale again
# Immune_referenceSub <- NormalizeData(Immune_referenceSub, normalization.method = "LogNormalize", scale.factor = 10000)
# Immune_referenceSub <- FindVariableFeatures(Immune_referenceSub, selection.method = "vst", nfeatures = 2000)
# Immune_referenceSub <- ScaleData(Immune_referenceSub, features = VariableFeatures(Immune_referenceSub))
# # Identify DEGs
# # we used the overall DEGs between infection and noninfection. 
# Immune_referenceSubDEGs <- FindMarkers(Immune_referenceSub, ident.1 = "7DPI", ident.2 = "0DPI", group.by = "orig.ident", test.use= "MAST",min.pct =0)
# write.csv(Immune_referenceSubDEGs, file = "Immune_referenceSubDEGs.csv")
# 
# saveRDS(Immune_referenceSubDEGs, "Immune_referenceSubDEGs.RDS")
# Immune_referenceSubDEGs<-readRDS("Immune_referenceSubDEGs.RDS")
# Immune_referenceSubDEGsFinalDEGs <- Immune_referenceSubDEGs %>%  filter(p_val_adj <= 0.05 &(avg_log2FC > 1 | avg_log2FC < -1))
# #nrow(Immune_referenceSubDEGsFinalDEGs)
# Immune_referenceSubDEGs[c("Vcam1","Ccl7","Mmp8","Arg1","Mmp13"),]
# # For single cell, we obtain the degs in the immune cells.
# 
# # 

library(ggvenn)
# generate the Venn figure for DEGs among three datasets
Deggene_list <- list(
  spatialfinaldegs =FinalOverlappingGenes$gene,
  singlecellfinaldegs =rownames(SingleCellOverallFinalDEGs),
  microfinaldegs =MouseMicroArrayDEGs$mgi_symbol
)

ThreeDatasetsOveralapingVenn<- ggvenn(
  Deggene_list,
  fill_color = c("skyblue", "orange","gray"),
  stroke_size = 0.5,
  set_name_size = 5,
  show_percentage=F
)
?ggvenn
ggsave("ThreeDatasetsOveralapingVenn.pdf",plot = ThreeDatasetsOveralapingVenn, height = 3, width =3)
# the overlapping elements
FinalOverallapping<-  Reduce(intersect, Deggene_list)
length(FinalOverallapping)
write.csv(FinalOverallapping, file = "ThreeDatasetFinalOverlappingGenes.csv")
MouseMicroArrayDEGs %>% filter(mgi_symbol %in% c("Arg1"))
```

```{r}
# read the secreted proteins in mouse
# The publication was in SPD databse :
# Chen Y, Zhang Y, Yin Y, Gao G, Li S, Jiang Y, Gu X, Luo J. SPD--a web-based secreted protein database. Nucleic Acids Res. 2005 Jan 1;33(Database issue):D169-73. doi: 10.1093/nar/gki093. PMID: 15608170; PMCID: PMC540047.
SecretedProteinDir<- c("/Users/XXW004/Documents/Projects/Databases/SecretedProteins/")

SecretedMouseProteins<- read.csv(paste0(SecretedProteinDir,"Mouse.csv"))
SecretedMousefinalGenes<- SecretedMouseProteins$Gene %>% unique()
length(SecretedMousefinalGenes)
SecretedMousefinalGenes[SecretedMousefinalGenes %in% c("Vcam1","Ccl7","Mmp8","Arg1","Mmp13")]
#SecretedMousefinalGenes["Arg1"]
# read the urinary proteins collected from health, AIK, transplant, bladder, Prostate cancer, Non-specific proteinuria patients (NS)
# human urine:
# Swensen, A.C., He, J., Fang, A.C., Ye, Y., Nicora, C.D., Shi, T., Liu, A.Y., Sigdel, T.K., Sarwal, M.M. and Qian, W.J., 2021. A comprehensive urine proteome database generated from patients with various renal conditions and prostate cancer. Frontiers in Medicine, 8, p.548212.

HumanUrineProteinDir<- c("/Users/XXW004/Documents/Projects/Databases/UrinaryProteins/")

HumanUrineProteins<- read.csv(paste0(HumanUrineProteinDir,"UrinaryProteins.csv"))
HumanUrineProteins %>% nrow()
# transfer the human urinary proteins into Mouse homologous genes
library(biomaRt)

# connect to the ensmebl database
humanensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")

# 
listAttributes(humanensembl)[grep("uniprot", listAttributes(humanensembl)$name), ]
# the list of uniprotIDs
HumanUrineProteins %>% filter(grepl("_HUMAN",Protein) ) %>% pull(Accession) %>% unique()
HumanUrineProteinsUniprotID<- HumanUrineProteins %>% filter(grepl("_HUMAN",Protein) ) %>% pull(Accession) %>% unique()
length(HumanUrineProteinsUniprotID)
# Retrieve gene symbols for the UniProt IDs
HumanUrinegene_mapping <- getBM(
  attributes = c("uniprotswissprot", "hgnc_symbol"),
  filters = "uniprotswissprot",
  values = HumanUrineProteinsUniprotID,
  mart = humanensembl
)

HumanUrinegene_mapping  %>% nrow()
#HumanUrineProteins %>% nrow()
HumanUrinegene_mapping$hgnc_symbol %>% length()
# the mouse and human homologous genes

MouseHumanHomDir<- c("/Users/XXW004/Documents/Projects/Databases/HumanMouseTransfer/")
# read the mouse and human homologous genes
mh_data<- read.csv(paste0(MouseHumanHomDir,"HOM_MouseHuman_Symbol.csv"))


MouseHomo_HumanUrinegene <- mh_data %>% filter(Symbol.y %in% HumanUrinegene_mapping$hgnc_symbol) %>% pull(Symbol.x) %>% unique()
length(MouseHomo_HumanUrinegene)
SecretedMousefinalGenes
## check the overlapping with mousehuman urine homologous, secreted genes
CandiateDeggene_list <- list(
  Spatialdegs =FinalOverlappingGenes$gene,
  #spatialdatedegs = DEGsInfectedDays$gene,
  Singlecelldegs =rownames(SingleCellOverallFinalDEGs),
  Microarraydegs =MouseMicroArrayDEGs$mgi_symbol,
  Secretedgenes = SecretedMousefinalGenes,
  Urinarygenes= MouseHomo_HumanUrinegene
)

length(CandiateDeggene_list$Spatialdegs)
length(CandiateDeggene_list$Singlecelldegs)
length(CandiateDeggene_list$Microarraydegs)
length(CandiateDeggene_list$Secretedgenes)
length(CandiateDeggene_list$Urinarygenes)
# FiveDatasetsOveralapingVenn<- ggvenn(
#   CandiateDeggene_list,
#   #fill_color = c("skyblue", "orange","gray"),
#   fill_color =c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#D55E00"),
#   stroke_size = 0.5,
#   set_name_size = 5,
#   show_percentage=F
# )
#BiocManager::install("UpSetR")
#install.packages("UpSetR")
library(UpSetR)
# Convert to fromList format
input_for_upset <- fromList(CandiateDeggene_list)
names(CandiateDeggene_list)
sapply(CandiateDeggene_list, length)
pdf("CandidateBioMarkers_AcrossDifferentDatasets.pdf", height = 5, width = 10)
upset(
  input_for_upset,
  set_size.show = TRUE,
  sets = names(CandiateDeggene_list),
  order.by = "freq",
  keep.order = TRUE
)
dev.off()
FinalOverallapping<-  Reduce(intersect, CandiateDeggene_list)
FinalOverallapping

```


# Generate the biomarkers distribution of fold change heatmap
```{r}
# the fold change and significance across datasets
MouseMicroArrayPyelD7vsD0 %>% head()
## Spatial niches comparison
SingleCellOverallDEGs %>% head()
SpatialNichesComparison <- AllCategories_Clustermarkers_infectedvsnoninfected %>% dplyr::select(gene ,avg_log2FC,p_val_adj) %>%mutate(Group = "SpatialNiches") %>% rename("Gene"="gene","log2FoldChange"="avg_log2FC","Padj"="p_val_adj") 
SpatialNichesComparison %>% head()
## Spatial date comparision
SpatialDateComparison <- AllCategories_Clustermarkers_infectedvsnoninfectedDate %>% dplyr::select(gene ,avg_log2FC,p_val_adj) %>%mutate(Group = "SpatialDates") %>% rename("Gene"="gene","log2FoldChange"="avg_log2FC","Padj"="p_val_adj")
## Single cell comparision
SingleCellComparison <- SingleCellOverallDEGs %>% rownames_to_column("gene") %>% dplyr::select(gene ,avg_log2FC,p_val_adj) %>%mutate(Group = "SingleCells") %>% rename("Gene"="gene","log2FoldChange"="avg_log2FC","Padj"="p_val_adj")
## Microarray comparison
MicroArrayComparison <- MouseMicroArrayPyelD7vsD0 %>% dplyr::select(mgi_symbol ,logFC,adj.P.Val) %>%mutate(Group = "MicroArray") %>% rename("Gene"="mgi_symbol","log2FoldChange"="logFC","Padj"="adj.P.Val")
#MicroArrayComparison %>% head()

FinalExpDEGs<- rbind(SpatialNichesComparison,SpatialDateComparison, SingleCellComparison,MicroArrayComparison)
FinalExpDEGs
table(FinalExpDEGs$Group)


# we only overlapping genes : FinalOverallapping
# "F13a1"  "Timp1"  "Apom"   "Ccl6"   "Lyz1"   "Col5a1" "Napsa"  "Fgl2"   "Vcam1"  "Ctss"   "Lcn2"   "Clu"    "B2m"    "Lmnb1" 
PathwayName <- "Conserved Biomarkers"

FinalExpDEGs<- FinalExpDEGs %>% filter(Gene %in% FinalOverallapping)
# we then generate a heatmap for the 

# assign signaficance labels
FinalExpDEGs$significance <- ifelse(FinalExpDEGs$Padj <= 0.001, "***",
                   ifelse(FinalExpDEGs$Padj <= 0.01, "**",
                   ifelse(FinalExpDEGs$Padj <= 0.05, "*", "")))
#reorder the group
levels(as.factor(FinalExpDEGs$Group))
FinalExpDEGs$Group <- factor(FinalExpDEGs$Group, levels = c("SpatialNiches", "SpatialDates", "SingleCells","MicroArray"))  
levels(FinalExpDEGs$Group)

FinalOverallappingOrdered <- c("Lcn2","Clu","Vcam1","Lyz1","Timp1","Ccl6","Fgl2","Ctss","B2m","Col5a1","Lmnb1","F13a1","Apom","Napsa") ## we can customize the genes
FinalExpDEGs$Gene <- factor(FinalExpDEGs$Gene, levels = FinalOverallappingOrdered)
levels(FinalExpDEGs$Gene)

# # only ICA
# FinalExpDEGs$Group <- factor(FinalExpDEGs$Group, levels = c("HFD_Mouse_IC", "T2DM_Mouse_ICA", "T2DM_Human_ICA"))  
# Create heatmap using ggplot2
RequestedGenesHeatmap<-ggplot(FinalExpDEGs, aes(x = Group, y = Gene, fill = log2FoldChange)) +
  geom_tile(color = "white") +  # Heatmap tiles with borders
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0,limits = c(-8, 8))+
                       #limits = c(-2, 2)) +  # Color scale
  geom_text(aes(label = significance), color = "black", size = 2) +  # Add p-value stars
  theme_minimal() +  # Clean theme
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +  # Rotate x-axis labels
  labs(title = PathwayName, fill = "logFC") +
  theme(legend.key.size = unit(0.4, "cm"), legend.title = element_text(size = 8))
RequestedGenesHeatmap

ggsave(paste0(PathwayName, "_FC_HeatMap_inCollecDuct.pdf"),plot =RequestedGenesHeatmap, height = 5, width =3 )
RequestedGenesHeatmap

RequestedGenesDotPlot<-ggplot(FinalExpDEGs, aes(x = Group, y = Gene)) +
  geom_point(aes(size = -log10(Padj), color = log2FoldChange)) +  # Dot size based on logFC magnitude
  scale_color_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0,limits = c(-8, 8)) +  
  #geom_text(aes(label = significance), vjust = -1, size = 6) +  # Add significance stars
  theme_classic() +  # Clean theme
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +  # Rotate x-axis labels
  labs(title = PathwayName, size = "-log10(padj)", color = "LogFC")
RequestedGenesDotPlot
ggsave(paste0(PathwayName, "_FC_DotPlot_inCollecDuct_DKD.pdf"),plot =RequestedGenesDotPlot, height = 9, width = 4 )
RequestedGenesDotPlot
FinalExpDEGs$Gene
FinalExpDEGs$Gene <- factor(FinalExpDEGs$Gene, levels = c("Atp6v1g3","Atp6v1c2", "Atp6v1b1", "Atp6v0d2","Atp6v0a4"))  

ggplot(FinalExpDEGs, aes(x = log2FoldChange, y = Gene, fill = Group)) +
  geom_bar(stat = "identity", position = "dodge",color = "black") +  # "dodge" separates bars for each category
  labs(x = "Log2Fold Change",
       y = "Genes") +
  theme_classic() + 
  scale_fill_grey(start = 0.3, end = 0.9) +  # Apply gray color scale
  theme(legend.position = "top") 


```


```{r}
FindMarkers(object = Pyelonephritis_Infection,
                        ident.1 = "InfectionDays",
                        ident.2 = "NonInfectionDays",
                        group.by = "InfectedDate",
                        subset.ident = "N1",
                        min.pct=0.1, 
                        logfc.threshold= 0.25, 
                        test.use= "MAST")

for (i in Infectedniches){
  markers<- FindMarkers(object = Pyelonephritis_Infection,
                        ident.1 = "InfectionDays",
                        ident.2 = "NonInfectionDays",
                        group.by = "InfectedDate",
                        subset.ident = i,
                        min.pct=0.1, 
                        logfc.threshold= 0.25, 
                        test.use= "MAST")
  # create folder for each niches
  Nichesmarkers<- paste0(BioMarkerDir,i,"_Niches/")
  
  write.csv(markers, file = paste0(Nichesmarkers,i,"_markers.csv"))
}
```


```{r}
# we also restricted our analyes 
# check the specific marker by the MAST
#AllCategories_Clustermarkers <- FindAllMarkers(Pyelonephritis_Infection,
#            min.pct = 0.25,test.use= "MAST") # only markers that are in 25% of cells

# save all biomarkers as a dataframe
#write.csv(AllCategories_Clustermarkers, "AllCategories_Clustermarkers_MAST.csv")
# We have already generated the data in 
NicheseMarkerDir <-c("/Users/XXW004/Documents/Projects/RuizRosado/SpatialTranscriptomic/Results/NichesDEGs_V09062023/KidneyInfectedVsOther/Biomarkers/")

# here we draw the specific biomarkers for the infected niches
AllCategories_Clustermarkers <- read.csv(paste0(NicheseMarkerDir,"AllCategories_Clustermarkers_MAST.csv"), header = T)
AllCategories_Clustermarkers %>% head()
 
Top_InfectedrmarkersComparingAll<-AllCategories_Clustermarkers %>%
                group_by(cluster) %>% # group by cluster (cell type)
                slice_max(avg_log2FC, n = 10) %>% filter(cluster %in% Infectedniches)



Top25_InfectedrmarkersComparingAll<-AllCategories_Clustermarkers %>%
                group_by(cluster) %>% # group by cluster (cell type)
                slice_max(avg_log2FC, n = 25) %>% filter(cluster %in% Infectedniches)


Top_InfectedrmarkersComparingAllGene<- Top_InfectedrmarkersComparingAll %>% pull(gene) %>% unique()

Top25_InfectedrmarkersComparingAllGene<- Top25_InfectedrmarkersComparingAll %>% pull(gene) %>% unique()

  # generate the top 10 genes for each biomarker
# generate the heatmap
Top_InfectedrmarkersComparingAllGenesheatmap<- DoHeatmap(Pyelonephritis_Infection, features = Top_InfectedrmarkersComparingAllGene, size = 2.5)  + scale_fill_gradientn(colors = c("lightblue", "white", "FireBrick")) +  theme(legend.position="bottom") 
?DoHeatmap
Top_InfectedrmarkersComparingAllGenesheatmap
ggsave("Top_InfectedrmarkersComparingAllGenesheatmap.pdf", plot =Top_InfectedrmarkersComparingAllGenesheatmap, height = 10, width = 9 )

Top_InfectedrmarkersComparingAllGeneDotPlot<- 
   DotPlot(Pyelonephritis_Infection, features = Top_InfectedrmarkersComparingAllGene) +coord_flip() +
  theme(legend.text = element_text(size = 12), legend.title = element_text(size = 8))+RotatedAxis() 
  #DotPlot(Pyelonephritis_Infection, features = top_InfectedrmarkersGenes)+RotatedAxis() 

ggsave("Top_InfectedrmarkersComparingAllGeneDotPlot.pdf", plot =Top_InfectedrmarkersComparingAllGeneDotPlot, height = 9, width = 9 )


Top25_InfectedrmarkersComparingAllGeneDotPlot<- 
   DotPlot(Pyelonephritis_Infection, features = Top25_InfectedrmarkersComparingAllGene) +coord_flip() +
  theme(legend.text = element_text(size = 12), legend.title = element_text(size = 8))+RotatedAxis() 
  #DotPlot(Pyelonephritis_Infection, features = top_InfectedrmarkersGenes)+RotatedAxis() 
ggsave("Top25_InfectedrmarkersComparingAllGeneDotPlot.pdf", plot =Top25_InfectedrmarkersComparingAllGeneDotPlot, height = 18, width = 9 )

```


```{r}
setwd(BioMarkerDir)

# generate the top 30 genes for each biomarker
top_Infectedrmarkers <- as.data.frame(AllCategories_Clustermarkers %>%
                group_by(cluster) %>% # group by cluster (cell type)
                slice_max(avg_log2FC, n = 30))
# generate the featureplot for each biomarker
top_Infectedrmarkers

top_InfectedrmarkersGenes<- top_Infectedrmarkers %>% pull(gene) %>% unique()

for (t in top_InfectedrmarkersGenes) {
  # generate the figures for the top 10 biomarkers for each section
  RegionalSpecificSpatilaFeatureVariouCategoriesCut5 <-
    SpatialFeaturePlot(Pyelonephritis_Infection,
                       features = t,
                       images = selectedslices) &
    ggplot2::scale_fill_gradientn(colours = myPalette(4), limits =
                                    c(-1, 5))
  # save into a file
  ggsave(
    filename = paste0(BioMarkerDir, t, "_SpatialFeatureplot_Cut5.pdf"),
    plot = RegionalSpecificSpatilaFeatureVariouCategoriesCut5,
    height = 8,
    width = 15
  )
  
  RegionalSpecificSpatilaFeatureVariouCategoriesCut8 <-
    SpatialFeaturePlot(Pyelonephritis_Infection,
                       features = t,
                       images = selectedslices) &
    ggplot2::scale_fill_gradientn(colours = myPalette(4), limits =
                                    c(-1, 8))
  # save into a file
  ggsave(
    filename = paste0(BioMarkerDir, t, "_SpatialFeatureplot_Cut8.pdf"),
    plot = RegionalSpecificSpatilaFeatureVariouCategoriesCut8,
    height = 8,
    width = 15
  )
  
  RegionalSpecificSpatilaFeatureVariouCategoriesCut10 <-
    SpatialFeaturePlot(Pyelonephritis_Infection,
                       features = t,
                       images = selectedslices) &
    ggplot2::scale_fill_gradientn(colours = myPalette(4), limits =
                                    c(-1, 10))
  # save into a file
  ggsave(
    filename = paste0(BioMarkerDir, t, "_SpatialFeatureplot_Cut10.pdf"),
    plot = RegionalSpecificSpatilaFeatureVariouCategoriesCut10,
    height = 8,
    width = 15
  )
  
  RegionalSpecificSpatilaFeatureVariouCategoriesCut20 <-
    SpatialFeaturePlot(
      Pyelonephritis_Infection,
      features = t,
      pt.size.factor = 1800,
      alpha = c(0.8, 1),
      images = selectedslices
    ) &
    ggplot2::scale_fill_gradientn(colours = myPalette(4), limits =
                                    c(-1, 20))
  # save into a file
  ggsave(
    filename = paste0(BioMarkerDir, t, "_SpatialFeatureplot_Cut20.pdf"),
    plot = RegionalSpecificSpatilaFeatureVariouCategoriesCut20,
    height = 8,
    width = 15
  )
  
  
}
```

# generate the dotplot and heatmap
```{r}
Idents(Pyelonephritis_Infection)
levels(Idents(Pyelonephritis_Infection))

# only for the top 5
# generate the top 5 genes for each biomarker
top5_Infectedrmarkers_genes <- as.data.frame(InfectedCategories_Clustermarkers %>%
                group_by(cluster) %>% # group by cluster (cell type)
                slice_max(avg_log2FC, n = 5)) %>% pull(gene) %>% unique()
# generate the heatmap
top_InfectedrmarkersGenesheatmap<- DoHeatmap(Pyelonephritis_Infection, features = top5_Infectedrmarkers_genes, size = 2.5)  + scale_fill_gradientn(colors = c("lightblue", "white", "FireBrick")) +  theme(legend.position="bottom")
?DoHeatmap
top_InfectedrmarkersGenesheatmap
ggsave("top_InfectedrmarkersGenesheatmap.pdf", plot =top_InfectedrmarkersGenesheatmap, height = 10, width = 9 )

top_InfectedrmarkersGeneDotPlot<- 
   DotPlot(Pyelonephritis_Infection, features = top5_Infectedrmarkers_genes) +coord_flip() +
  theme(legend.text = element_text(size = 12), legend.title = element_text(size = 8))
  #DotPlot(Pyelonephritis_Infection, features = top_InfectedrmarkersGenes)+RotatedAxis() 

ggsave("top_InfectedrmarkersGeneDotPlot.pdf", plot =top_InfectedrmarkersGeneDotPlot, height = 8, width = 6 )

```

# Generate the dotplot for different CKDs biomarkers
```{r}
# For Glomerular: 
GlomerularMarker <- paste0(BioMarkerDir,"/Glomerular/")
dir.create(GlomerularMarker)
setwd(GlomerularMarker)
GlomerularCKD_markers<- c("Alb", "Ddn", "Nphs1", "Trf", "Col4a1","Col4a2","Col4a3", "Col4a4","Col4a5", "Fn1", "Pdpn", "Synpo", "Cp", "Ptgds", "Ighm", "Des", "Smad1", "Actn1", "Adam10", "Chi3l1", "Podxl", "Vegfa", "Ptpro", "Myb", "Nphs2", "Eno3", "B2m")

# generate the heatmap
GlomerularCKD_markersheatmap<- DoHeatmap(Pyelonephritis_Infection, features = GlomerularCKD_markers, size = 2.5)  + scale_fill_gradientn(colors = c("lightblue", "white", "FireBrick")) +  theme(legend.position="bottom")
?DoHeatmap
ggsave("GlomerularCKD_markersGenesheatmap.pdf", plot =GlomerularCKD_markersheatmap, height = 10, width = 10 )

GlomerularCKD_markersDotPlot<- 
   DotPlot(Pyelonephritis_Infection, features = GlomerularCKD_markers) +coord_flip() +
  theme(legend.text = element_text(size = 12), legend.title = element_text(size = 8))
  #DotPlot(Pyelonephritis_Infection, features = top_InfectedrmarkersGenes)+RotatedAxis() 

ggsave("GlomerularCKD_markersDotPlot.pdf", plot =GlomerularCKD_markersDotPlot, height = 8, width = 10 )

# make a directory for the spatial features
GlomerularMarkerSpFeature <- paste0(GlomerularMarker,"/SpatialFeature/")
dir.create(GlomerularMarkerSpFeature)
setwd(GlomerularMarkerSpFeature)

for (t in GlomerularCKD_markers){
  
  # generate the figures for the top 10 biomarkers for each section
  RegionalSpecificSpatilaFeatureVariouCategoriesCut5 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(-1,5))
     # save into a file
    ggsave(filename = paste0(GlomerularMarkerSpFeature,t, "_SpatialFeatureplot_Cut5.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut5, height = 8, width = 15 )
    
    RegionalSpecificSpatilaFeatureVariouCategoriesCut8 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(0,8))
     # save into a file
    ggsave(filename = paste0(GlomerularMarkerSpFeature,t, "_SpatialFeatureplot_Cut8.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut8, height = 8, width = 15 )
    
     RegionalSpecificSpatilaFeatureVariouCategoriesCut10 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(0,10))
     # save into a file
    ggsave(filename = paste0(GlomerularMarkerSpFeature,t, "_SpatialFeatureplot_Cut10.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut10, height = 8, width = 15 )
    
      RegionalSpecificSpatilaFeatureVariouCategoriesCut20 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(0,20))
     # save into a file
    ggsave(filename = paste0(GlomerularMarkerSpFeature,t, "_SpatialFeatureplot_Cut20.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut20, height = 8, width = 15 )

 
}

TubulointerstitialCKD_markers<- c("Cst3", "Havcr1", "Lcn2", "Oga", "Fabp3", "Ccn2", "Myb", "Umod", "Il18", "Lgals3", "Vnn1", "Nes", "Ambp", "Timp1", "Fabp1")
####
#### For Tubulointerstitial: 
TubulointerstitialMarker <- paste0(BioMarkerDir,"Tubulointerstitial/")
dir.create(TubulointerstitialMarker)
setwd(TubulointerstitialMarker)

# generate the heatmap
TubulointerstitialCKD_markersheatmap<- DoHeatmap(Pyelonephritis_Infection, features = TubulointerstitialCKD_markers, size = 2.5)  + scale_fill_gradientn(colors = c("lightblue", "white", "FireBrick")) +  theme(legend.position="bottom")
?DoHeatmap
ggsave("TubulointerstitialCKD_markersGenesheatmap.pdf", plot =TubulointerstitialCKD_markersheatmap, height = 10, width = 10 )

TubulointerstitialCKD_markersDotPlot<- 
   DotPlot(Pyelonephritis_Infection, features = TubulointerstitialCKD_markers) +coord_flip() +
  theme(legend.text = element_text(size = 12), legend.title = element_text(size = 8))
  #DotPlot(Pyelonephritis_Infection, features = top_InfectedrmarkersGenes)+RotatedAxis() 

ggsave("TubulointerstitialCKD_markersDotPlot.pdf", plot =TubulointerstitialCKD_markersDotPlot, height = 8, width = 10 )

# make a directory for the spatial features
TubulointerstitialMarkerSpFeature <- paste0(TubulointerstitialMarker,"SpatialFeature/")
dir.create(TubulointerstitialMarkerSpFeature)
setwd(TubulointerstitialMarkerSpFeature)

for (t in TubulointerstitialCKD_markers){
  
  # generate the figures for the top 10 biomarkers for each section
  RegionalSpecificSpatilaFeatureVariouCategoriesCut5 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(-1,5))
     # save into a file
    ggsave(filename = paste0(TubulointerstitialMarkerSpFeature,t, "_SpatialFeatureplot_Cut5.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut5, height = 8, width = 15 )
    
    RegionalSpecificSpatilaFeatureVariouCategoriesCut8 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(0,8))
     # save into a file
    ggsave(filename = paste0(TubulointerstitialMarkerSpFeature,t, "_SpatialFeatureplot_Cut8.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut8, height = 8, width = 15 )
    
     RegionalSpecificSpatilaFeatureVariouCategoriesCut10 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(0,10))
     # save into a file
    ggsave(filename = paste0(TubulointerstitialMarkerSpFeature,t, "_SpatialFeatureplot_Cut10.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut10, height = 8, width = 15 )
    
      RegionalSpecificSpatilaFeatureVariouCategoriesCut20 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(0,20))
     # save into a file
    ggsave(filename = paste0(TubulointerstitialMarkerSpFeature,t, "_SpatialFeatureplot_Cut20.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut20, height = 8, width = 15 )
 
}


# Ahsg	Ccl2	Il18	Cd40lg	Chit1	
####
#### For Endothelial dysfunction

EndothelialDysCKD_markers <- c("Ahsg", "Ccl2", "Cxcl15", "Cd40lg", "Chit1")
EndothelialDysMarker <- paste0(BioMarkerDir,"EndothelialDys/")
dir.create(EndothelialDysMarker)
setwd(EndothelialDysMarker)

# generate the heatmap
EndothelialDysCKD_markersheatmap<- DoHeatmap(Pyelonephritis_Infection, features = EndothelialDysCKD_markers, size = 2.5)  + scale_fill_gradientn(colors = c("lightblue", "white", "FireBrick")) +  theme(legend.position="bottom")
?DoHeatmap
ggsave("EndothelialDysCKD_markersGenesheatmap.pdf", plot =EndothelialDysCKD_markersheatmap, height = 10, width = 10 )

EndothelialDysCKD_markersDotPlot<- 
   DotPlot(Pyelonephritis_Infection, features = EndothelialDysCKD_markers) +coord_flip() +
  theme(legend.text = element_text(size = 12), legend.title = element_text(size = 8))
  #DotPlot(Pyelonephritis_Infection, features = top_InfectedrmarkersGenes)+RotatedAxis() 

ggsave("EndothelialDysCKD_markersDotPlot.pdf", plot =EndothelialDysCKD_markersDotPlot, height = 8, width = 10 )

# make a directory for the spatial features
EndothelialDysMarkerSpFeature <- paste0(EndothelialDysMarker,"SpatialFeature/")
dir.create(EndothelialDysMarkerSpFeature)
setwd(EndothelialDysMarkerSpFeature)

for (t in EndothelialDysCKD_markers){
  
  # generate the figures for the top 10 biomarkers for each section
  RegionalSpecificSpatilaFeatureVariouCategoriesCut5 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(-1,5))
     # save into a file
    ggsave(filename = paste0(EndothelialDysMarkerSpFeature,t, "_SpatialFeatureplot_Cut5.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut5, height = 8, width = 15 )
    
    RegionalSpecificSpatilaFeatureVariouCategoriesCut8 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(0,8))
     # save into a file
    ggsave(filename = paste0(EndothelialDysMarkerSpFeature,t, "_SpatialFeatureplot_Cut8.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut8, height = 8, width = 15 )
    
     RegionalSpecificSpatilaFeatureVariouCategoriesCut10 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(0,10))
     # save into a file
    ggsave(filename = paste0(EndothelialDysMarkerSpFeature,t, "_SpatialFeatureplot_Cut10.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut10, height = 8, width = 15 )
    
      RegionalSpecificSpatilaFeatureVariouCategoriesCut20 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(0,20))
     # save into a file
    ggsave(filename = paste0(EndothelialDysMarkerSpFeature,t, "_SpatialFeatureplot_Cut20.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut20, height = 8, width = 15 )
 
}


# Inflammation

#####
#### For Inflammation dysfunction

InflammationCKD_markers <- c("Crp", "Ptx3","Tnfrsf1a", "Tnfrsf1b", "Il18", "Il6", "Il10","Gdf15","Ccl6")

# Final used
FinalInflammationCKD_markers <- c("Tnfrsf1a", "Tnfrsf1b", "Il18", "Il6", "Ccl6")

InflammationMarker <- paste0(BioMarkerDir,"Inflammation/")
dir.create(InflammationMarker)
setwd(InflammationMarker)

# generate the heatmap
InflammationCKD_markersheatmap<- DoHeatmap(Pyelonephritis_Infection, features = InflammationCKD_markers, size = 2.5)  + scale_fill_gradientn(colors = c("lightblue", "white", "FireBrick")) +  theme(legend.position="bottom")
?DoHeatmap
ggsave("InflammationCKD_markersGenesheatmap.pdf", plot =InflammationCKD_markersheatmap, height = 10, width = 10 )

InflammationCKD_markersDotPlot<- 
   DotPlot(Pyelonephritis_Infection, features = InflammationCKD_markers) +coord_flip() +
  theme(legend.text = element_text(size = 12), legend.title = element_text(size = 8))
  #DotPlot(Pyelonephritis_Infection, features = top_InfectedrmarkersGenes)+RotatedAxis() 

ggsave("InflammationCKD_markersDotPlot.pdf", plot =InflammationCKD_markersDotPlot, height = 8, width = 10 )

FinalInflammationCKD_markersDotPlot<- 
   DotPlot(Pyelonephritis_Infection, features = FinalInflammationCKD_markers) +coord_flip() +
  theme(legend.text = element_text(size = 12), legend.title = element_text(size = 8))
  #DotPlot(Pyelonephritis_Infection, features = top_InfectedrmarkersGenes)+RotatedAxis() 

ggsave("FinalLimitedInflammationCKD_markersDotPlot.pdf", plot =FinalInflammationCKD_markersDotPlot, height = 4, width = 9 )


# make a directory for the spatial features
InflammationMarkerSpFeature <- paste0(InflammationMarker,"SpatialFeature/")
dir.create(InflammationMarkerSpFeature)
setwd(InflammationMarkerSpFeature)

for (t in InflammationCKD_markers){
  
  # generate the figures for the top 10 biomarkers for each section
  RegionalSpecificSpatilaFeatureVariouCategoriesCut5 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(-1,5))
     # save into a file
    ggsave(filename = paste0(InflammationMarkerSpFeature,t, "_SpatialFeatureplot_Cut5.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut5, height = 8, width = 15 )
    
    RegionalSpecificSpatilaFeatureVariouCategoriesCut8 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(0,8))
     # save into a file
    ggsave(filename = paste0(InflammationMarkerSpFeature,t, "_SpatialFeatureplot_Cut8.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut8, height = 8, width = 15 )
    
     RegionalSpecificSpatilaFeatureVariouCategoriesCut10 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(0,10))
     # save into a file
    ggsave(filename = paste0(InflammationMarkerSpFeature,t, "_SpatialFeatureplot_Cut10.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut10, height = 8, width = 15 )
    
      RegionalSpecificSpatilaFeatureVariouCategoriesCut20 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(0,20))
     # save into a file
    ggsave(filename = paste0(InflammationMarkerSpFeature,t, "_SpatialFeatureplot_Cut20.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut20, height = 8, width = 15 )
 
}




#### For Fibrosis dysfunction

FibrosisCKD_markers <- c("Tgfb1", "Cst3", "Tgfa", "Egf", "Il1rl1",  "Postn", "Pth", "Inhba", "Acp5", "Fgf23", "Kl", "Mmp9", "Mmp2", "Mmp7", "Dkk3")

FibrosisMarker <- paste0(BioMarkerDir,"Fibrosis/")
dir.create(FibrosisMarker)
setwd(FibrosisMarker)

# generate the heatmap
FibrosisCKD_markersheatmap<- DoHeatmap(Pyelonephritis_Infection, features = FibrosisCKD_markers, size = 2.5)  + scale_fill_gradientn(colors = c("lightblue", "white", "FireBrick")) +  theme(legend.position="bottom")
?DoHeatmap
ggsave("FibrosisCKD_markersGenesheatmap.pdf", plot =FibrosisCKD_markersheatmap, height = 10, width = 10 )

FibrosisCKD_markersDotPlot<- 
   DotPlot(Pyelonephritis_Infection, features = FibrosisCKD_markers) +coord_flip() +
  theme(legend.text = element_text(size = 12), legend.title = element_text(size = 8))
  #DotPlot(Pyelonephritis_Infection, features = top_InfectedrmarkersGenes)+RotatedAxis() 

ggsave("FibrosisCKD_markersDotPlot.pdf", plot =FibrosisCKD_markersDotPlot, height = 8, width = 10 )

# make a directory for the spatial features
FibrosisMarkerSpFeature <- paste0(FibrosisMarker,"SpatialFeature/")
dir.create(FibrosisMarkerSpFeature)
setwd(FibrosisMarkerSpFeature)

for (t in FibrosisCKD_markers){
  
  # generate the figures for the top 10 biomarkers for each section
  RegionalSpecificSpatilaFeatureVariouCategoriesCut5 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(-1,5))
     # save into a file
    ggsave(filename = paste0(FibrosisMarkerSpFeature,t, "_SpatialFeatureplot_Cut5.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut5, height = 8, width = 15 )
    
    RegionalSpecificSpatilaFeatureVariouCategoriesCut8 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(0,8))
     # save into a file
    ggsave(filename = paste0(FibrosisMarkerSpFeature,t, "_SpatialFeatureplot_Cut8.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut8, height = 8, width = 15 )
    
     RegionalSpecificSpatilaFeatureVariouCategoriesCut10 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(0,10))
     # save into a file
    ggsave(filename = paste0(FibrosisMarkerSpFeature,t, "_SpatialFeatureplot_Cut10.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut10, height = 8, width = 15 )
    
      RegionalSpecificSpatilaFeatureVariouCategoriesCut20 <- 
          SpatialFeaturePlot(Pyelonephritis_Infection,
                             features = t, 
                             images = selectedslices) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(0,20))
     # save into a file
    ggsave(filename = paste0(FibrosisMarkerSpFeature,t, "_SpatialFeatureplot_Cut20.pdf"), 
       plot =RegionalSpecificSpatilaFeatureVariouCategoriesCut20, height = 8, width = 15 )
 
}

# for AKI
AKI<- c("S100a8", "S100a9", "Lcn2", "Il18",  "Havcr1", "Igfbp7", "Fabp1", "Cst3","Umod") 
AKIMarker <- paste0(BioMarkerDir,"AKI/")
dir.create(AKIMarker)
setwd(AKIMarker)
# generate the heatmap
AKIMarker_markersheatmap<- DoHeatmap(Pyelonephritis_Infection, features = AKI, size = 2.5)  + scale_fill_gradientn(colors = c("lightblue", "white", "FireBrick")) +  theme(legend.position="bottom")
?DoHeatmap
ggsave("AKIMarker_markersGenesheatmap.pdf", plot =AKIMarker_markersheatmap, height = 10, width = 10 )

AKIMarker_markersDotPlot<- 
   DotPlot(Pyelonephritis_Infection, features = rev(AKI)) +coord_flip() +RotatedAxis() +
  theme(legend.text = element_text(size = 12), legend.title = element_text(size = 8))
  #DotPlot(Pyelonephritis_Infection, features = top_InfectedrmarkersGenes)+RotatedAxis() 

ggsave("AKIMarker_markersDotPlot.pdf", plot =AKIMarker_markersDotPlot, height = 5, width = 8 )

# for final figure, including the tubular function, tubular injury, Inflammation, Fibrosis
RenalInjuryMarkers<-c ("Havcr1","Vcam1","C3","Lcn2","Vim","Cst3","Umod","Igfbp7","Il18", "S100a8","S100a9")
FinalAKICKD_Inflammation_Fibrosis_Markers<- c("Nfkb1","Rela", "Tnf","Tnfrsf1a","Tnfrsf1b", "Il6","Il1b", "Nlrp3","Ccl6","Ccl2","Cxcl10","Cxcl12","Tgfb1","Smad3","Ccn2","Col1a1","Col3a1","Col4a1","Fn1","Acta2","Timp1","Timp2","Mmp2","Mmp7","Mmp9","Cybb","Hmox1","Nfe2l2","Cat","Nox4","Prdx1","Sod1","Sod3","Hmgb1","Ripk3","Ripk1","Mlkl","Ldhb")

# 
FinalAKICKD_Markers <- c("Lcn2","Havcr1","S100a8","S100a9", "Clu","Spp1", "Vcam1","C3","Igfbp7","Il18","Il1b","Ccl2","Tnf","Akt","Vim","Cst3","Lrp2","Umod","Col1a1","Agt","Fn1","Acta2","Tgfb1","Serpinf1","Mmp9")


# AKI CKD markers 

# generate the heatmap
FinalAKICKD_markersheatmap<- DoHeatmap(Pyelonephritis_Infection, features = FinalAKICKD_Inflammation_Fibrosis_Markers, size = 2.5)  + scale_fill_gradientn(colors = c("lightblue", "white", "FireBrick")) +  theme(legend.position="bottom")
?DoHeatmap
ggsave("FinalAKICKD_markersGenesheatmap.pdf", plot =FinalAKICKD_markersheatmap, height = 10, width = 10 )

DefaultAssay(Pyelonephritis_Infection) <-"SCT"
DoHeatmap(Pyelonephritis_Infection, features = FinalAKICKD_Markers, size = 2.5)  + scale_fill_gradientn(colors = c("lightblue", "white", "FireBrick")) +  theme(legend.position="bottom")
FinalAKICKD_markersDotPlot<-DotPlot(
    Pyelonephritis_Infection,
    features = FinalAKICKD_Markers,
    cols = "RdBu"
  ) + guides(color = guide_colorbar(title = 'AverExp'),
             size = guide_legend(title = "PercExp"))  + RotatedAxis()
FinalAKICKD_markersDotPlot
ggsave("FinalOnlyAKICKD_markersDotPlot.pdf", plot =FinalAKICKD_markersDotPlot, height = 5, width = 7)
FinalAKICKD_markersDotPlot <-
  DotPlot(
    Pyelonephritis_Infection,
    features = FinalAKICKD_Inflammation_Fibrosis_Markers,
    cols = "RdBu"
  ) + guides(color = guide_colorbar(title = 'AverExp'),
             size = guide_legend(title = "PercExp"))  + RotatedAxis() 
  # theme(legend.text = element_text(size = 12),
  #       legend.title = element_text(size = 8))
  #DotPlot(Pyelonephritis_Infection, features = top_InfectedrmarkersGenes)+RotatedAxis() 
FinalAKICKD_markersDotPlot
ggsave("FinalAKICKD_markersDotPlot.pdf", plot =FinalAKICKD_markersDotPlot, height = 5, width = 15 )

#NovelBiomarkers<- c("Ccl7","Arg1","Mmp8")
# Generate the feature Plots
for (j in FinalAKICKD_Inflammation_Fibrosis_Markers) {
#for (j in NovelBiomarkers) {  
    # Generate DimPlot or SpatialDimPlot based on user's choice
    # Function to generate DimPlot plot
    #  j="Calr"
    
    # we use this details to escape the figures that did not show up in spatial
    skip_to_next <- FALSE
    
    # Note that print(b) fails since b doesn't exist
    
    tryCatch(
      print(Pyelonephritis_Infection@assays$SCT@data[j, ]),
      error = function(e) {
        skip_to_next <<- TRUE
      }
    )
    
    if (skip_to_next) {
      next
    }
    # get the maxium and minvalue of the gene
    Maxvalue <- ceiling(max(Pyelonephritis_Infection@assays$SCT@data[j, ]))
    Minvalue <- floor(min(Pyelonephritis_Infection@assays$SCT@data[j, ]))
    GenesInGOtermMouseSpatialFeature <-
      SpatialFeaturePlot(
        Pyelonephritis_Infection,
        features = j,
        pt.size.factor = 1600,
        crop = TRUE
      ) &
      ggplot2::scale_fill_gradientn(colours = myPalette(4),
                                    limits = c(Minvalue, Maxvalue))
    
    # save the figure into the folder
    ggsave(
      filename = paste0(AKIMarker, "/", "Spatial_", j, "_In_AKI_CKD_Inf_Fib_ROS.pdf"),
      plot = GenesInGOtermMouseSpatialFeature,
      height = 3,
      width = 15
    )
  
  }



```


# add precise pathway scoring using GSVA
```{r}

setwd(KEGGDir)
# BiocManager::install("GSVA")
# BiocManager::install("GSEABase")
library(GSVA)
library(GSEABase)
# Using msigdbr to pull hallmark angiogenesis genes
library(msigdbr)
library(dplyr)

msig <- msigdbr(species = "Mus musculus", category = "H")  # H = Hallmark
msig %>% pull(gs_name) %>% table()
hallmark_angiogenesis <- msig %>% filter(gs_name == "HALLMARK_ANGIOGENESIS") %>% pull(gene_symbol)
hallmark_tgfb <- msig %>% filter(gs_name == "HALLMARK_TGF_BETA_SIGNALING") %>% pull(gene_symbol)
hallmark_em <- msig %>% filter(gs_name == "HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION") %>% pull(gene_symbol)
hallmark_myogenesis <- msig %>% filter(gs_name == "HALLMARK_MYOGENESIS") %>% pull(gene_symbol)
# Examples:
hallmark_inflam <- msig %>% filter(gs_name == "HALLMARK_INFLAMMATORY_RESPONSE") %>% pull(gene_symbol)
hallmark_ifn_gamma <- msig %>% filter(gs_name == "HALLMARK_INTERFERON_GAMMA_RESPONSE") %>% pull(gene_symbol)
hallmark_tnfa <- msig %>% filter(gs_name == "HALLMARK_TNFA_SIGNALING_VIA_NFKB") %>% pull(gene_symbol)
hallmark_ros <- msig %>% filter(gs_name == "HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY") %>% pull(gene_symbol)
hallmark_hypoxia <- msig %>% filter(gs_name == "HALLMARK_HYPOXIA") %>% pull(gene_symbol)
hallmark_Apoptosis <- msig %>% filter(gs_name == "HALLMARK_APOPTOSIS") %>% pull(gene_symbol)

ifn_response <- c("Irf7", "Stat1", "Ifit1", "Ifit2", "Ifit3", "Isg15", "Oas1a", "Mx1") # Type I Interferon Response Response to viral and some bacterial PAMPs
nfkb_pathway <- c("Nfkb1", "Nfkb2", "Rel", "Relb", "Ikbkb", "Ikbkg", "Chuk", "Traf6", "Myd88", "Il1b", "Tnfa") # Master regulator of cytokine production and immune activation
amp_genes <- c("Lcn2", "Defb1", "Defb4", "Camp", "S100a8", "S100a9", "Slpi", "Reg3g", "Nos2")
neutrophil_recruitment <- c("Cxcl1", "Cxcl2", "Cxcr2", "Il1b", "Tnf", "S100a8", "S100a9", "Nlrp3", "Ptx3")
complement_pathway <- c("C1qa", "C1qb", "C1qc", "C3", "C5", "Cfb", "Cfh", "Cfd", "Cd55", "Cd59a")
# Query mouse necroptosis pathway
necroptosispathway_id <- "mmu04217"
library(KEGGREST)
Necropathway_info <- keggGet(necroptosispathway_id)
# Extract gene symbols
necroptosisgene_entries <- Necropathway_info[[1]]$GENE
necroptosismouse_genes <- necroptosisgene_entries[seq(2, length(necroptosisgene_entries), 2)]  # Every 2nd entry is the gene symbol
necroptosismouse_genes
necroptosismouse_genessymbol <- unique(sub("; .*", "", necroptosismouse_genes))  # Clean extra annotation
# 
tlr_id <- "mmu04620"
tlrpathway_info <- keggGet(tlr_id)
# Extract gene symbols
tlrgene_entries <- tlrpathway_info[[1]]$GENE
tlrmouse_genes <- tlrgene_entries[seq(2, length(tlrgene_entries), 2)]  # Every 2nd entry is the gene symbol
tlrmouse_genes
tlr_pathway <- unique(sub("; .*", "", tlrmouse_genes))  # 
tlr_pathway

cytokine_id <- "mmu04060"
cytokinepathway_info <- keggGet(cytokine_id)
# Extract gene symbols
cytokinegene_entries <-cytokinepathway_info[[1]]$GENE
cytokinemouse_genes <- cytokinegene_entries[seq(2, length(cytokinegene_entries), 2)]  # Every 2nd entry is the gene symbol
cytokinemouse_genes
cytokine_pathway <- unique(sub("; .*", "", cytokinemouse_genes))  # 
cytokine_pathway
#cytokine_pathway <- c("Cxcl1", "Cxcl2", "Cxcl10", "Il6", "Il1b", "Tnf", "Ccl2", "Ccl5", "Il10ra", "Il18r1") ## KEGG pathway: mmu04060



DefaultAssay(Pyelonephritis_Infection) <-"SCT"

Pyelonephritis_Infection <- AddModuleScore(
  Pyelonephritis_Infection,
  features = list(
    amp_genes,
    tlr_pathway,
    nfkb_pathway,
    neutrophil_recruitment,
    cytokine_pathway,
    complement_pathway,
    necroptosismouse_genessymbol,
    hallmark_tnfa,
    hallmark_ifn_gamma,
    hallmark_inflam,
    hallmark_angiogenesis,
    hallmark_tgfb,
    hallmark_em,
    hallmark_myogenesis,
    hallmark_ros,
    hallmark_hypoxia,
    hallmark_Apoptosis
  ),
  name = c(
    "Amp",
    "TLRSignaling",
    "NFKB",
    "Cytokine",
    "NeutrophilRecruitment",
    "ComplementPathway",
    "Necroptosis",
    "TNFASignaling",
    "InterferonGammaResponse",
    "InflammatoryResponse",
    "angiogenesis",
    "tgfbsignal",
    "epithelialmesenchymaltransition",
    "myogenesis",
    "ROS",
    "Hypoxia",
    "Apoptosis"
  )
)
ColumnOrders<- colnames(Pyelonephritis_Infection@meta.data)[20:36]
ColumnOrders
# create a data frame for dotplot from metadata
df<- Pyelonephritis_Infection@meta.data %>% 
  dplyr::select(Annotation, ColumnOrders) %>% pivot_longer(cols = starts_with("Amp"):starts_with("Apopto"), 
               names_to = "Pathway", values_to = "Score") %>% 
  group_by(Annotation, Pathway) %>%
  dplyr::summarize(
    avg_score = mean(Score),
    pct_expr = sum(Score > 0) / n() * 100
  )

# plot 
df$avg_score <- as.numeric(df$avg_score)
df$pct_expr <- as.numeric(df$pct_expr)
dev.off()
FinalPathwayAvergeDotPlot<-ggplot(df, aes(x = Pathway, y = Annotation)) +
  geom_point(aes(size = pct_expr, color = avg_score)) +
 scale_color_viridis_c(option = "rocket") +
  geom_tile() +
  #scale_fill_distiller(palette = "Spectral", direction = -1) +
  theme_minimal() +
  labs(title = "Pathway Scores", x = "Pathway", y = "Cluster") +RotatedAxis() +coord_flip()
ggsave(
      filename = "FinalPathwayAvergeDotplots.pdf",
      plot = FinalPathwayAvergeDotPlot,
      height = 4,
      width = 6
    )
# using the pheatmap
FinalPathwayAvergeExp<-df %>% dplyr::select(Annotation, Pathway, avg_score)%>% pivot_wider(names_from = Pathway, values_from = avg_score) 
FinalPathwayAvergeExpTable<- FinalPathwayAvergeExp[,-1] %>% as.data.frame()
rownames(FinalPathwayAvergeExpTable) <-FinalPathwayAvergeExp$Annotation
FinalPathwayAvergeExpTable
myColor = colorRampPalette(c("royalblue", "white","lightcoral"))(paletteLength)


FinalPathwayAvergeExpTable <- FinalPathwayAvergeExpTable[,ColumnOrders]
FinalPathwayAvergeExpTable
FinalPathwayAvergepheatmap<-pheatmap(t(FinalPathwayAvergeExpTable),fontsize=12, 
                        fontsize_row = 10, scale="row",cluster_rows = F, cluster_cols = F,
                        color=myColor,  
                        #main = "PROGENy signaling", 
                        #angle_col = 45,
                        treeheight_col = 0,  border_color = NA, treeheight_row=0) 

ggsave(
      filename = "FinalPathwayAvergepheatmap.pdf",
      plot = FinalPathwayAvergepheatmap,
      height = 4,
      width = 6
    )


# then generate the pathway featuremap
# Generate the feature Plots
for (m in ColumnOrders) {
    # Generate DimPlot or SpatialDimPlot based on user's choice
    # Function to generate DimPlot plot
    #  j="Calr"
   # ColumnOrders
  #i=  "Apoptosis17"
  # we use this details to escape the figures that did not show up in spatial

    # get the maxium and minvalue of the pathway
   Maxvalue <- max(Pyelonephritis_Infection[[m]])
    Minvalue <- min(Pyelonephritis_Infection[[m]])

       PathwayMouseSpatialFeature <-
      SpatialFeaturePlot(
        Pyelonephritis_Infection,
        features = m,
        pt.size.factor = 1600,
        crop = TRUE
      ) &
      ggplot2::scale_fill_gradientn(colours = myPalette(4),
                                    limits = c(Minvalue, Maxvalue))
    GenesInGOtermMouseSpatialFeature
    # save the figure into the folder
    ggsave(
      filename = paste0("Spatial",m,"PathwayscoreFeaturePlot.pdf"),
      plot = PathwayMouseSpatialFeature,
      height = 3,
      width = 15
    )
  
}


```


# generate the Dotplot for the functional enrichment of each classification:
```{r}
# change the folder back
setwd(file.path(GOenrichDir))
AllCategories_Clustermarkers %>% head()
SectionClusters <- levels(Idents(Pyelonephritis_Infection))
SectionClusters

# create a gene list for differnt regions
SectionDEGlistWithInfection <- list()
SectionClusters

# generate the DEGs for the different regions :"Cortex","Medulla","Pelvis","Vascular","Processing Infection"
for (m in SectionClusters){
  # put he DEG into the DEGlist for comparision
  # we only consider the Cortex, Medulla, Pelvis 
  if( m == "N12" | m == "N13" | m == "N14" | m== "N15" | m== "N16"| m== "N17"){
     SectionDEGlistWithInfection[[m]] <- AllCategories_Clustermarkers %>% filter(cluster %in% m & avg_log2FC >= 0.25) %>% pull(gene)
  }
 }

# Comparing the Cluster 
SectionDEGlistGOPlotWithInfection <- compareCluster(geneClusters = SectionDEGlistWithInfection, fun = "enrichGO", OrgDb = "org.Mm.eg.db",ont="BP",keyType = 'SYMBOL')

# simplify the GO enrichment
SectionDEGlistGOPlotWithInfectionSimplifed <- clusterProfiler::simplify(SectionDEGlistGOPlotWithInfection)

SectionDEGlistGOPlotWithInfectionSimplifed
SectionDEGlistGOPlotWithInfectionDotPlot<- dotplot(SectionDEGlistGOPlotWithInfection,font.size = 12,label_format = 80,showCategory = 6) +RotatedAxis()
?dotplot
#low = "LightBlue", midpoint=0, mid="white", high = "FireBrick",
ggsave(paste0("InfectedSectionDEGlistGOPlotDotPlot_withinfection.pdf"), plot = SectionDEGlistGOPlotWithInfectionDotPlot, height = 10, width = 10 )
write.csv(SectionDEGlistGOPlotWithInfectionSimplifed, "InfectedSectionDEGlistWithInfectionGOPlotSimplifed.csv")

# considering the GO enrichment took a long time, we read the saved files

SectionDEGlistGOPlotWithInfectionSimplifed <- read.csv("InfectedSectionDEGlistWithInfectionGOPlotSimplifed.csv", header = T)


```


# generate the KEGG pathway oever-representation analysis

```{r}

setwd(KEGGDir)
library(org.Mm.eg.db)
KEGGpathways <-list()
KEGGpathwaysForView<-list()
## put the list of UNIPROT into the list
# we generate this for the kegg pathview
SectionDEGTransferUnirotID <-list()
SectionClusters
#SectionDEGlistWithInfection
for (m in SectionClusters){
  # put he DEG into the DEGlist for comparision
  # we only consider the infected niches 
  if( m == "N12" | m == "N13" | m == "N14" | m== "N15" | m== "N16"| m== "N17"){
    
     #m= "Processing Infection"
     SectionDEGlistWithInfection[[m]] <- InfectedCategories_Clustermarkers %>% filter(cluster %in% m & avg_log2FC >= 0.25) %>% pull(gene)
     
     # transfer biolgical ID to KEGG ID
     TransferIDs<- bitr(SectionDEGlistWithInfection[[m]] , fromType = "SYMBOL", toType = c("UNIPROT","ENSEMBL"), OrgDb=org.Mm.eg.db)
     TransferIDs %>% head()
      
     SectionDEGTransferUnirotID[[m]] <- TransferIDs$UNIPROT
     
     egID = bitr(SectionDEGlistWithInfection[[m]] , fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Mm.eg.db")

     KEGGpathwaysForView[[m]] <-enrichKEGG(gene = egID$ENTREZID,
                                           organism = 'mmu',
                                           pvalueCutoff = 0.05)
     KEGGpathways[[m]] <- enrichKEGG(gene =  TransferIDs$UNIPROT, organism = 'mmu', pvalueCutoff = 1,keyType = 'uniprot')
  
     write.csv(KEGGpathways[[m]], file = paste0(KEGGDir, m, "_KeggEnrichpath.csv"))
  }
 }
```


## compare the KEGG pathways among Cortex, medulla, Pelvis and Processing Infection
```{r}

#SectionDEGTransferUnirotID
length(SectionDEGTransferUnirotID)
 # compare four up-regulated gene list with GO enrichment, we did not set up the p value 
SectionDEGTransferKEGG <- compareCluster(geneCluster = SectionDEGTransferUnirotID, fun = enrichKEGG, organism="mmu", keyType = 'uniprot',pvalueCutoff=1)

# setting into readable ID
SectionDEGTransferKEGGReadable <- setReadable(SectionDEGTransferKEGG, OrgDb = "org.Mm.eg.db", keyType="UNIPROT")

# remove the string - Mus musculus (house mouse)
SectionDEGTransferKEGGReadable@compareClusterResult$Description <-

gsub(pattern = " - Mus musculus (house mouse)", replacement = "", SectionDEGTransferKEGGReadable@compareClusterResult$Description, fixed = T) 


#%>% mutate(Description = str_remove(Description, "- Mus musculus (house mouse)")) %>% head()
# 
# SectionDEGTransferKEGGReadable$Description <- gsub(pattern = " - Mus musculus (house mouse)", replacement = "", SectionDEGTransferKEGGReadable$Description, fixed = T)

# Show all the top15 pathways
Top15Comparion<-  
  dotplot(SectionDEGTransferKEGGReadable,font.size = 12,label_format = 80,showCategory = 15) 

ggsave(paste0(KEGGDir,"Top15Comparion_Sections_KEGG_comp.pdf"), plot = Top15Comparion, width = 15, height = 15)
# save into a enriched pathway file
write.csv(SectionDEGTransferKEGGReadable, file = paste0(KEGGDir,"AllSections_KeggEnrichpath.csv") )


# save the top ID

# show the candidate pathway in the infected regions
selectedPathway<- c("Carbon metabolism","Glyoxylate and dicarboxylate metabolism","Galactose metabolism","Cell adhesion molecules","ECM-receptor interaction","Focal adhesion","Tight junction","Necroptosis","Complement and coagulation cascades","Apoptosis","Endocytosis","Lysosome","Hippo signaling pathway","Phagosome","Neutrophil extracellular trap formation","Chemokine signaling pathway","Hematopoietic cell lineage","Cytokine-cytokine receptor interaction","NOD-like receptor signaling pathway","Osteoclast differentiation","NF-kappa B signaling pathway","Th17 cell differentiation","Antigen processing and presentation","Inflammatory bowel disease","Toll-like receptor signaling pathway","Human cytomegalovirus infection","TNF signaling pathway","C-type lectin receptor signaling pathway","Th1 and Th2 cell differentiation","IL-17 signaling pathway","Fc gamma R-mediated phagocytosis","Cytosolic DNA-sensing pathway","Natural killer cell mediated cytotoxicity","B cell receptor signaling pathway","JAK-STAT signaling pathway","Intestinal immune network for IgA production","Leukocyte transendothelial migration","T cell receptor signaling pathway","Rap1 signaling pathway","Bacterial invasion of epithelial cells","Fc epsilon RI signaling pathway","p53 signaling pathway")

# select these interesting pathway
SelectedPathwayForfigure<-SectionDEGTransferKEGGReadable %>% as.data.frame() %>% filter(Description %in% selectedPathway) %>% arrange(p.adjust)

# setting up the minium p adjust value for size visulization 
maximum<- max(c(-log10(SelectedPathwayForfigure$p.adjust)))

## set the levels of Description and reorder them
SelectedPathwayForfigure$Description<- ordered(SelectedPathwayForfigure$Description, levels=selectedPathway)
#reorder(variable, value)
p<-ggplot(SelectedPathwayForfigure, aes(x=SelectedPathwayForfigure$Cluster, y=SelectedPathwayForfigure$Description))

InterestingComparionsDotplot<- p+ geom_point(aes(size=-log10(p.adjust)),
                                             colour= ifelse(SelectedPathwayForfigure$p.adjust <=0.05,"red",
                                                            ifelse(SelectedPathwayForfigure$p.adjust==1,"white","gray")), 
                                             fill= ifelse(SelectedPathwayForfigure$p.adjust<=0.05,"red",
                                                          ifelse(SelectedPathwayForfigure$p.adjust==1,"white","gray"))) +
  labs(title = "KEGG Pathway Enrichment", x = "Spatial region", y = "KEGG pathway enrichment") +
  theme_bw()+
  theme(plot.background = element_blank(),
        text = element_text(size=10),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.text.x = element_text(angle = 30, hjust = 1)
  ) + scale_size_area(limits=c(0,maximum),breaks=c(0,4,6,8,12))


InterestingComparions <- dotplot(SectionDEGTransferKEGGReadable,font.size = 12,label_format = 80,showCategory = selectedPathway)  +RotatedAxis()

ggsave(paste0(KEGGDir, "InterestingPathwayComparion_Sections_KEGG_comp.pdf"), plot = InterestingComparions, width = 8, height = 10)
ggsave(paste0(KEGGDir, "InterestingPathwayComparion_Sections_KEGG_comp_format.pdf"), plot = InterestingComparionsDotplot, width = 8, height = 10)
```

# Using Progeny for main signals
```{r}
setwd(ProgenyDir)
## we set up with a high memory

options(future.globals.maxSize = 89128960000000) 
## We compute the Progeny activity scores and add them to our Seurat object
## as a new assay called Progeny. 
Idents(Pyelonephritis_Infection)
# create cluster data frame with the cells, Celltypes
CorMedCluster<- data.frame(Cell = names(Idents(Pyelonephritis_Infection)),   
                              CellType = as.character(Idents(Pyelonephritis_Infection)),
                              stringsAsFactors = FALSE)

?progeny
library(progeny)
Pyelonephritis_Infection <- progeny(Pyelonephritis_Infection, scale=FALSE, organism="Mouse", top=500, perm=1, return_assay = TRUE,assay_name = "SCT")
Pyelonephritis_Infection
# setup the default assay as progeny
DefaultAssay(object = Pyelonephritis_Infection) <- "progeny"

Pyelonephritis_Infection <- Seurat::ScaleData(Pyelonephritis_Infection, assay = "progeny") 
## We can now directly apply Seurat functions in our Progeny scores. 
## For instance, we scale the pathway activity scores. 
progeny_scores_df <- 
  as.data.frame(t(GetAssayData(Pyelonephritis_Infection, slot = "scale.data", 
                               assay = "progeny"))) %>%
  rownames_to_column("Cell") %>%
  gather(Pathway, Activity, -Cell) 

progeny_scores_df
#progeny_scores_df %>% filter(Cell =="0DPI_AAAAAAAACCACCAGA")
## We match Progeny scores with the cell clusters.
progeny_scores_df <- inner_join(progeny_scores_df, CorMedCluster)

progeny_scores_df %>% head()
## We summarize the Progeny scores by cellpopulation
summarized_progeny_scores <- progeny_scores_df %>% 
  group_by(Pathway, CellType) %>%
  summarise(avg = mean(Activity), std = sd(Activity))
summarized_progeny_scores


## We prepare the data for the plot
summarized_progeny_scores_df <- summarized_progeny_scores %>%
  dplyr::select(-std) %>%   
  spread(Pathway, avg) %>%
  data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 

paletteLength = 80
myColor = colorRampPalette(c("royalblue", "white","firebrick3"))(paletteLength)

progenyBreaks = c(seq(min(summarized_progeny_scores_df), 0, 
                      length.out=ceiling(paletteLength/2) + 1),
                  seq(max(summarized_progeny_scores_df)/paletteLength, 
                      max(summarized_progeny_scores_df), 
                      length.out=floor(paletteLength/2)))
CorMedForHeatmap<-as.data.frame(t(summarized_progeny_scores_df[,-1]))
CorMedForHeatmap
NichesOrder<- c(paste0("N",1:17))
PathwayOrder<- c("TNFa","NFkB","JAK-STAT","p53","Hypoxia","Trail","PI3K","EGFR","MAPK","WNT","TGFb","VEGF","Estrogen")

# change the order of 

CorMedForHeatmap_Ordered<- CorMedForHeatmap[PathwayOrder,NichesOrder]

CorMedForHeatmap_Ordered
# generate the heatmap for PT
pdf(file="CorMedForHeatmap_Progeny_ForHeatmap.pdf",height = 3, width = 3.5)
progeny_hmap = pheatmap(CorMedForHeatmap_Ordered,fontsize=8, 
                        fontsize_row = 8,  show_rownames=T, cluster_cols=F, cluster_rows=F, scale="row",
                        cex=1, clustering_distance_rows="euclidean", cex=1,
                        # clustering_distance_cols="euclidean", clustering_method="complete", 
                        color=myColor, breaks = progenyBreaks, 
                        main = "PROGENy (500)", angle_col = 45,
                        border_color = NA,
                        treeheight_col = 0,   treeheight_row=0)

progeny_hmap
dev.off()
test<- FeaturePlot(Pyelonephritis_Infection, features = "NFkB")
ggsave("test_NFKB.pdf", plot = test, width = 5, height = 5)
# generate the featureplot 
ProgenyFeatures <- summarized_progeny_scores$Pathway %>% unique()

SpatialColors <- colorRampPalette(colors = rev(x = brewer.pal(n = 8, name = "Spectral")))


for (i in ProgenyFeatures) {
  p2 <- (
    FeaturePlot(Pyelonephritis_Infection, features = i) &
      scale_colour_gradient2(
        low = 'blue',
        mid = 'white',
        high = 'red'
      )
  ) &
    ggtitle(paste0(i, " activity"))
  
  ggsave(
    paste0(i, "_CorMedInfected_cluster_featurePlot.pdf"),
    plot = p2,
    width = 6,
    height = 6
  )
  Maxvalue <-
    ceiling(max(Pyelonephritis_Infection@assays$progeny@data[i, ]))
  Minvalue <-
    floor(min(Pyelonephritis_Infection@assays$progeny@data[i, ]))
  p3 <-
    SpatialFeaturePlot(
      Pyelonephritis_Infection,
      features = i,
      pt.size.factor = 1800,
      alpha = c(0.8, 1),
      crop = TRUE
    ) &
    ggplot2::scale_fill_gradientn(colours =  myPalette(4),
                                  limits = c(Minvalue, Maxvalue))
  ggsave(
    paste0(i, "_CorMedInfected_cluster_SpatialfeaturePlot.pdf"),
    plot = p3,
    width = 15,
    height = 3
  )
  
}

# generate the P53
RequestedPathFeaturePlot<- SpatialFeaturePlot(
      Pyelonephritis_Infection,
      features = "p53",
      pt.size.factor = 1800,
      alpha = c(0.8, 1),
      crop = TRUE
    ) &
    ggplot2::scale_fill_gradientn(colours =  myPalette(4),
                                  limits = c(-100, 600))
  ggsave(
    paste0("P53", "_CorMedInfected_cluster_SpatialfeaturePlot.pdf"),
    plot = RequestedPathFeaturePlot,
    width = 15,
    height = 3
  )
  

```


# Using PROGENy model from DecoupleR
```{r}
library(decoupleR)
# set the identity as orginal
Idents(Pyelonephritis_Infection) <- Pyelonephritis_Infection @meta.data$Annotation
#devtools::install_github("saezlab/progeny")
library(decoupleR)
setwd(ProgenyDir)
net <- decoupleR::get_progeny(organism = 'mouse', top = 500)

Pyelonephritis_Infection@assays$SCT@data %>% head()
# Activity inference with Multivariate Linear Model (MLM)
# Extract the normalized log-transformed counts
mat <- as.matrix(Pyelonephritis_Infection@assays$SCT@data)

# Run mlm
acts <- run_mlm(mat=mat, net=net, .source='source', .target='target',
                .mor='weight', minsize = 5)

acts 
# Visualization:
# Extract mlm and store it in pathwaysmlm in data
Pyelonephritis_Infection[['pathwaysmlm']] <- acts %>%
  pivot_wider(id_cols = 'source', names_from = 'condition',
              values_from = 'score') %>%
  column_to_rownames('source') %>%
  Seurat::CreateAssayObject(.)


# Change assay
DefaultAssay(object = Pyelonephritis_Infection) <- "pathwaysmlm"

# Scale the data
Pyelonephritis_Infection <- ScaleData(Pyelonephritis_Infection)
Pyelonephritis_Infection@assays$pathwaysmlm@data <- Pyelonephritis_Infection@assays$pathwaysmlm@scale.data

# run the source of pathways
ProgenyFeatures <- levels(factor(acts$source))
ProgenyFeatures
for (i in ProgenyFeatures ){
  p2 <-SpatialFeaturePlot(Pyelonephritis_Infection, features = i) +
  ggtitle(paste0(i," activity")) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(-1,5))
  
  ggsave(paste0(i, "_Pyelonephritis_Infection_cluster_featurePlot_DecoupleR.pdf"), plot = p2, width = 20, height = 4)
}

#(c("royalblue", "white","lightcoral"))
Pyelonephritis_Infection@meta.data$orig.ident
SpatialFeaturePlot(Pyelonephritis_Infection, features = "Hypoxia", min.cutoff = -4, max.cutoff = 6) & 
   ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(-1,5))& 
  ggtitle(paste0("Hypoxia"," activity"))

# SpatialFeaturePlot(Pyelonephritis_Infection, features = "NFkB", min.cutoff = -4, max.cutoff = 6, split.by = "orig.ident") & 
#     ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(-1,5)) & 
#   ggtitle(paste0("NFkB"," activity"))
# 
# 
# SpatialFeaturePlot(Pyelonephritis_Infection, features = "TNFa", min.cutoff = -4, max.cutoff = 6) & 
#    ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(-1,5))& 
#   ggtitle(paste0("TNFa"," activity"))
# 
# p1 <- DimPlot(Pyelonephritis_Infection, reduction = "umap", label = TRUE, pt.size = 0.5) + 
#   NoLegend() + ggtitle('Cell types')
# p2 <- (SpatialFeaturePlot(Pyelonephritis_Infection, features = c("Trail") & 
#      ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(-1,5))&
#   ggtitle('Trail activity')



# Expore the mean activiey per group across pathways:
# Extract activities from object as a long dataframe
df <- t(as.matrix(Pyelonephritis_Infection@assays$pathwaysmlm@data)) %>%
  as.data.frame() %>%
  mutate(cluster = Idents(Pyelonephritis_Infection)) %>%
  pivot_longer(cols = -cluster, names_to = "source", values_to = "score") %>%
  group_by(cluster, source) %>%
  summarise(mean = mean(score))

# Transform to wide matrix
top_acts_mat <- df %>%
  pivot_wider(id_cols = 'cluster', names_from = 'source',
              values_from = 'mean') %>%
  column_to_rownames('cluster') %>%
  as.matrix()

# Choose color palette
palette_length = 100
#my_color = colorRampPalette(c("dodgerblue2", "white","firebrick1"))(palette_length)
my_color = colorRampPalette(c("Darkblue", "white","red"))(palette_length)


my_breaks <- c(seq(-2, 0, length.out=ceiling(palette_length/2) + 1),
               seq(0.05, 2, length.out=floor(palette_length/2)))

# Plot
Pyelonephritis_InfectionHeatmap<- pheatmap(t(top_acts_mat), border_color = NA, cluster_cols = F, color=my_color, breaks = my_breaks, angle_col = 45) 
ggsave("Pyelonephritis_InfectioncellsHeatmap_Progeny_Decouple.pdf", plot = Pyelonephritis_InfectionHeatmap, width = 6, height = 4)

```

# Transcription factor activity inference from Immune scRNA-seq
```{r}
# CollecTRI network
# set up the directory
library(decoupleR)
setwd(TFDir)
netTF <-  readRDS("/Users/XXW004/Documents/Projects/RuizRosado/PIPseq/Datasets/get_collectri_human_mouse20230716/get_collectri_mouse_process_table.rds")

mat <- as.matrix(Pyelonephritis_Infection@assays$SCT@data)
# Run ulm
#acts <- run_ulm(mat=mat, net=netTF, .source='source', .target='target',
 #               .mor='mor', minsize = 5)

#saveRDS(acts, file="Pyelonephritis_InfectionNiches_Transcriptfactor.RDS")
# this took a long time,
#acts <- readRDS(file = "Pyelonephritis_Infection_Transcriptfactor.RDS")

# Activity inference with Univariate Linear Model (ULM)
# Extract ulm and store it in tfsulm in pbmc
Pyelonephritis_Infection[['tfsulm']] <- acts %>%
  pivot_wider(id_cols = 'source', names_from = 'condition',
              values_from = 'score') %>%
  column_to_rownames('source') %>%
  Seurat::CreateAssayObject(.)

# Change assay
DefaultAssay(object = Pyelonephritis_Infection) <- "tfsulm"

# Scale the data
Pyelonephritis_Infection <- ScaleData(Pyelonephritis_Infection)
Pyelonephritis_Infection@assays$tfsulm@data <- Pyelonephritis_Infection@assays$tfsulm@scale.data

# Explore the top 20 more varaible TFs
n_tfs <- 20

t(as.matrix(Pyelonephritis_Infection@assays$tfsulm@data))
Idents(Pyelonephritis_Infection)
Idents(Pyelonephritis_Infection)

#t(as.matrix(Pyelonephritis_Infection@assays$tfsulm@data)) 
# Extract activities from object as a long dataframe
df <- t(as.matrix(Pyelonephritis_Infection@assays$tfsulm@data)) %>%
  as.data.frame() %>%
  mutate(cluster = Idents(Pyelonephritis_Infection)) %>%
  pivot_longer(cols = -cluster, names_to = "source", values_to = "score") %>%
  group_by(cluster, source) %>%
  summarise(mean = mean(score))


# # get the top 10 biomarkers 
# TopTF<- df %>% arrange(mean,cluster)%>% group_by(cluster) %>% slice_max(mean,n=10)
# nrow(Top)
# Get top tfs with more variable means across clusters
tfs <- df %>%
  group_by(source) %>%
  summarise(std = sd(mean)) %>%
  arrange(-abs(std)) %>%
  head(n_tfs) %>%
  pull(source)



# Transform to wide matrix for topTF
top_acts_mat <- df %>%
  pivot_wider(id_cols = 'cluster', names_from = 'source',
              values_from = 'mean') %>%
  column_to_rownames('cluster') %>%
  as.matrix()

top_acts_mat
# Choose color palette
palette_length = 100
my_color = colorRampPalette(c("Darkblue", "white","red"))(palette_length)

my_breaks <- c(seq(-2, 0, length.out=ceiling(palette_length/2) + 1),
               seq(0.05, 2, length.out=floor(palette_length/2)))

# Plot
pheatmap(top_acts_mat, border_color = NA, color=my_color, breaks = my_breaks) 
# drop "Znf335" due to not exist in the expression
#tfs <- tfs[!tfs == "Znf335"]

#tfs[!tfs == "Znf335"]
# Subset long data frame to top tfs and transform to wide matrix
top_acts_mat <- df %>%
  filter(source %in% tfs) %>%
  pivot_wider(id_cols = 'cluster', names_from = 'source',
              values_from = 'mean') %>%
  column_to_rownames('cluster') %>%
  as.matrix()

# Choose color palette
palette_length = 100
my_color = colorRampPalette(c("Darkblue", "white","red"))(palette_length)

my_breaks <- c(seq(-3, 0, length.out=ceiling(palette_length/2) + 1),
               seq(0.05, 3, length.out=floor(palette_length/2)))

# Plot
IMMTFheatmap<- pheatmap(t(top_acts_mat),border_color = NA, cluster_cols = F, 
                        color=my_color,fontsize=12, fontsize_row = 10, 
                        breaks = my_breaks, 
                        #main = "Predicted TF activities", 
                        angle_col = 45,
                        treeheight_col = 0, treeheight_row=0)


ggsave("Pyelonephritis_InfectedNichesHeatmap_Transcriptfactor_Progeny_Decouple.pdf", plot = IMMTFheatmap, width = 5, height = 5)

# draw the transcriptor factor activity and their expression
# generate the figure of transcript factor activity:
for (m in tfs){
  p2 <- (SpatialFeaturePlot(Pyelonephritis_Infection, features = m))  +
  ggtitle(paste0(m," activity")) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(-1,5)) 
  
  ggsave(paste0("TFActivity_",m,"_Pyelonephritis_Infection_FeaturePlot.pdf"), plot = p2, width = 16, height = 4)
}

p3 <- (SpatialFeaturePlot(Pyelonephritis_Infection, features = "Hbp1"))  +
  ggtitle(paste0("Hbp1"," activity")) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(-2,5)) 
p3  
  ggsave(paste0("TFActivity_","Hbp1","_Pyelonephritis_Infection_FeaturePlot.pdf"), plot = p3, width = 16, height = 4)

  p4 <- (SpatialFeaturePlot(Pyelonephritis_Infection, features = "Nfkb1"))  +
  ggtitle(paste0("Nfkb1"," activity")) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(-2,6)) 
p4  
  ggsave(paste0("TFActivity_","Nfkb1","_Pyelonephritis_Infection_FeaturePlot.pdf"), plot = p4, width = 16, height = 4)

Pyelonephritis_Infection
DefaultAssay(object = Pyelonephritis_Infection) <- "SCT"

for (n in tfs){
  p3 <- (SpatialFeaturePlot(Pyelonephritis_Infection, features = n) +
  ggtitle(paste0(m," activity")) &
          ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(-1,5)))
  
  ggsave(paste0("TFExpression_",n,"_Pyelonephritis_Infection_FeaturePlot.pdf"), plot = p3, width = 16, height = 4)

}

# add their activity in Dotplot
Idents(Pyelonephritis_Infection)
# get the order of row name
GeneOrderByPheatmap <- rownames(t(top_acts_mat))[IMMTFheatmap$tree_row$order]
#IMMTFheatmap$tree_row$order

?DotPlot
TFDotPlot <- DotPlot(Pyelonephritis_Infection, features = rev(GeneOrderByPheatmap)) +
  RotatedAxis() + coord_flip()+ 
  #ggtitle("Predicted TF expression")+
   theme(legend.position = "right")
ggsave("Spatial_TFDotPlot.pdf", plot = TFDotPlot, width = 10, height = 6)

```


# MarkerGenes in IC PCs
```{r}

ICMarkerDir <- c("/Users/XXW004/Documents/Projects/RuizRosado/SpatialTranscriptomic/Results/NichesDEGs_V04222025/KidneyInfectedVsOther/Biomarkers/ICPCMarkers/")
setwd(ICMarkerDir)
IC_PCMarkers<- c("Atp6v1g3","Aqp6","Slc4a1","Slc26a7","Kit","Slc26a4","Spink8","Hmx2","Aqp2","Hsd11b2","Scnn1g","Syt7")

DefaultAssay(Pyelonephritis_Infection) <-"SCT"
DoHeatmap(Pyelonephritis_Infection, features = IC_PCMarkers, size = 2.5)  + scale_fill_gradientn(colors = c("lightblue", "white", "FireBrick")) +  theme(legend.position="bottom")
FinalICPC_markersDotPlot<-DotPlot(
    Pyelonephritis_Infection,
    features = IC_PCMarkers,
    cols = "RdBu"
  ) + guides(color = guide_colorbar(title = 'AverExp'),
             size = guide_legend(title = "PercExp"))  + RotatedAxis()
FinalAKICKD_markersDotPlot
ggsave("FinalICPC_markersDotPlotDotPlot.pdf", plot =FinalICPC_markersDotPlot, height = 5, width = 6)

#NovelBiomarkers<- c("Ccl7","Arg1","Mmp8")
# Generate the feature Plots
for (j in IC_PCMarkers) {
#for (j in NovelBiomarkers) {  
    # Generate DimPlot or SpatialDimPlot based on user's choice
    # Function to generate DimPlot plot
    #  j="Calr"
    
    # we use this details to escape the figures that did not show up in spatial
    skip_to_next <- FALSE
    
    # Note that print(b) fails since b doesn't exist
    
    tryCatch(
      print(Pyelonephritis_Infection@assays$SCT@data[j, ]),
      error = function(e) {
        skip_to_next <<- TRUE
      }
    )
    
    if (skip_to_next) {
      next
    }
    # get the maxium and minvalue of the gene
    Maxvalue <- ceiling(max(Pyelonephritis_Infection@assays$SCT@data[j, ]))
    Minvalue <- floor(min(Pyelonephritis_Infection@assays$SCT@data[j, ]))
    GenesInGOtermMouseSpatialFeature <-
      SpatialFeaturePlot(
        Pyelonephritis_Infection,
        features = j,
        pt.size.factor = 1600,
        crop = TRUE
      ) &
      ggplot2::scale_fill_gradientn(colours = myPalette(4),
                                    limits = c(Minvalue, Maxvalue))
    
    # save the figure into the folder
    ggsave(
      filename = paste0(ICMarkerDir, "/", "Spatial_", j, "_IC_PCMarkers.pdf"),
      plot = GenesInGOtermMouseSpatialFeature,
      height = 3,
      width = 15
    )
  
  }


```

# check the cell type distribution

```{r}
SpatialCelltypeDir<- c("/Users/XXW004/Documents/Projects/RuizRosado/SpatialTranscriptomic/Results/NichesDEGs_V04222025/KidneyInfectedVsOther/CellTypes/")
setwd(SpatialCelltypeDir)
Celltypes<- rownames(Pyelonephritis_Infection[["predictions"]])
max(Pyelonephritis_Infection[["predictions"]]@data["IC",])


for (i in Celltypes ){
  maxvalue<- max(Pyelonephritis_Infection[["predictions"]]@data[i,])
  CelltypeSpatialFeature<- SpatialFeaturePlot(
        Pyelonephritis_Infection,
        features = i,
        pt.size.factor = 2000,
        crop = TRUE
      ) &
      ggplot2::scale_fill_gradientn(colours = myPalette(4),
                                    limits = c(0, maxvalue-0.05))
  ggsave(
      filename = paste0(SpatialCelltypeDir, "/", "Spatial_", i, "_CellTypeFeature.pdf"),
      plot = CelltypeSpatialFeature,
      height = 3,
      width = 15
    )
  
}


```

