---
title: "Spatial_Integration_Cluster"
author: "Xin Wang"
date: "2023-08-30"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# load the library
```{r}
library(Seurat)
#devtools::install_github('satijalab/seurat-data')
library(SeuratData)
library(ggplot2)
library(patchwork)
library(dplyr)
library(Matrix)
library(tidyverse)
library(SeuratObject)
#install.packages("harmony")
library(harmony)
library(SPOTlight)

#BiocManager::install("scater")
library(scater)

#BiocManager::install("scran")
library(scran)
library(SingleCellExperiment)

#remotes::install_github("mojaveazure/seurat-disk")
library(SeuratDisk)
library(tidyverse)
library(cowplot)

library(ggplot2)
library(reshape2)
library(Hmisc)
library(stats)
library(corrplot)
library(RColorBrewer)
## Deconvolve
# install.packages("L1pack")
# install.packages("reshape")
library(L1pack)
library(reshape)
if (!("xbioc" %in% rownames(inst))) {
  remotes::install_github("renozao/xbioc", dependencies = FALSE)
}
if (!("SCDC" %in% rownames(inst))) {
  remotes::install_github("meichendong/SCDC", dependencies = FALSE)
}

suppressPackageStartupMessages(require(SCDC))
suppressPackageStartupMessages(require(Biobase))

```

# set up the input and output directories
```{r}
outdir = "/Users/XXW004/Documents/Projects/RuizRosado/SpatialTranscriptomic/Results/Integration_v04142023/SpatialClusters_V0425/"
dir.create(outdir, showWarnings = F)
setwd("/Users/XXW004/Documents/Projects/RuizRosado/SpatialTranscriptomic/Results/Integration_v04142023/SpatialClusters_V0425/")
```

# load 10X Genomics Visium dataset
```{r}

P0<-Load10X_Spatial(data.dir="../Datasets_V04282022/ControlSample/", 
                    filename="filtered_feature_bc_matrix.h5", assay="Spatial", 
                    slice="slice1",  filter.matrix=TRUE, to.upper=FALSE, image=NULL)
P0[['orig.ident']] <-"PyeloD0_1"

P31<-Load10X_Spatial(data.dir="../WithSlideID_V0111/Pyelonephritis_31/", 
                            filename="filtered_feature_bc_matrix.h5", assay="Spatial", 
                            slice="slice2",  filter.matrix=TRUE, to.upper=FALSE, image=NULL)

P31[['orig.ident']] <-"PyeloD1_1"

#bb<-Read10X_Image(image.dir="acute_chronic_pyelo/seurat/",
#                 image.name = "tissue_lowres_image.png", filter.matrix=TRUE)
P32<-Load10X_Spatial(data.dir="../WithSlideID_V0111/Pyelonephritis_32", 
                                     filename="filtered_feature_bc_matrix.h5", assay="Spatial", 
                                     slice="slice3", filter.matrix=TRUE, to.upper=FALSE, image=NULL)
P32[['orig.ident']] <-"PyeloD1_2"
#dd<-Read10X_Image(image.dir="chronic/seurat/",
 #                 image.name = "tissue_lowres_image.png", filter.matrix=TRUE)
P33<-Load10X_Spatial(data.dir="../WithSlideID_V0111/Pyelonephritis_33", 
                               filename="filtered_feature_bc_matrix.h5", assay="Spatial", 
                               slice="slice4", filter.matrix=TRUE, to.upper=FALSE, image=NULL)
P33[['orig.ident']] <-"PyeloD3_1"
#cc<-Read10X_Image(image.dir="control/seurat/",
 #                 image.name = "tissue_lowres_image.png", filter.matrix=TRUE)
P34<-Load10X_Spatial(data.dir="../WithSlideID_V0111/Pyelonephritis_34", 
                               filename="filtered_feature_bc_matrix.h5", assay="Spatial", 
                               slice="slice5", filter.matrix=TRUE, to.upper=FALSE, image=NULL)
P34[['orig.ident']] <-"PyeloD3_2"
#aa<-Read10X_Image(image.dir="Pyelonephritis_31/spatial/", 
#                 image.name = "", filter.matrix=TRUE)
P41<-Load10X_Spatial(data.dir="../WithSlideID_V0111/Pyelonephritis_41/", 
                     filename="filtered_feature_bc_matrix.h5", assay="Spatial", 
                     slice="slice6", filter.matrix=TRUE, to.upper=FALSE, image=NULL)
P41[['orig.ident']] <-"PyeloD5_1"
#bb<-Read10X_Image(image.dir="acute_chronic_pyelo/seurat/",
#                 image.name = "tissue_lowres_image.png", filter.matrix=TRUE)
P42<-Load10X_Spatial(data.dir="../WithSlideID_V0111/Pyelonephritis_42", 
                     filename="filtered_feature_bc_matrix.h5", assay="Spatial", 
                     slice="slice7", filter.matrix=TRUE, to.upper=FALSE, image=NULL)
P42[['orig.ident']] <-"PyeloD5_2"


P11<-Load10X_Spatial(data.dir="../Datasets_V04282022/AcutePyelo/", 
                    filename="filtered_feature_bc_matrix.h5", assay="Spatial", 
                    slice="slice8",  filter.matrix=TRUE, to.upper=FALSE, image=NULL)
P11[['orig.ident']] <-"PyeloD7_1"

P12<-Load10X_Spatial(data.dir="../Datasets_V04282022/AcutePyeloneph/", 
                    filename="filtered_feature_bc_matrix.h5", assay="Spatial", 
                    slice="slice9",  filter.matrix=TRUE, to.upper=FALSE, image=NULL)
P12[['orig.ident']] <-"PyeloD7_2"

P28<-Load10X_Spatial(data.dir="../Datasets_V04282022/Chronicpyelo/", 
                     filename="filtered_feature_bc_matrix.h5", assay="Spatial", 
                     slice="slice10",  filter.matrix=TRUE, to.upper=FALSE, image=NULL)
P28[['orig.ident']] <-"PyeloD28_1"

#dd<-Read10X_Image(image.dir="chronic/seurat/",
#                 image.name = "tissue_lowres_image.png", filter.matrix=TRUE)
P43<-Load10X_Spatial(data.dir="../WithSlideID_V0111/Pyelonephritis_43", 
                     filename="filtered_feature_bc_matrix.h5", assay="Spatial", 
                     slice="slice11", filter.matrix=TRUE, to.upper=FALSE, image=NULL)
P43[['orig.ident']] <-"PyeloD56_1"

```

# Spatial normalization
```{r}
P0 <- SCTransform(P0, assay = "Spatial", verbose = FALSE)

P31 <- SCTransform(P31, assay = "Spatial", verbose = FALSE)
#P31 <- NormalizeData(P31, verbose = FALSE, assay = "Spatial")

P32 <- SCTransform(P32, assay = "Spatial", verbose = FALSE)
#P32 <- NormalizeData(P32, verbose = FALSE, assay = "Spatial")

P33 <- SCTransform(P33, assay = "Spatial", verbose = FALSE)
#P33 <- NormalizeData(P33, verbose = FALSE, assay = "Spatial")

P34 <- SCTransform(P34, assay = "Spatial", verbose = FALSE)
#P34 <- NormalizeData(P34, verbose = FALSE, assay = "Spatial")

P41 <- SCTransform(P41, assay = "Spatial", verbose = FALSE)
#P41 <- NormalizeData(P41, verbose = FALSE, assay = "Spatial")

P42 <- SCTransform(P42, assay = "Spatial", verbose = FALSE)
#P42 <- NormalizeData(P42, verbose = FALSE, assay = "Spatial")

P11 <- SCTransform(P11, assay = "Spatial", verbose = FALSE)

P12 <- SCTransform(P12, assay = "Spatial", verbose = FALSE)

P28 <- SCTransform(P28, assay = "Spatial", verbose = FALSE)

P43 <- SCTransform(P43, assay = "Spatial", verbose = FALSE)

```

### due to the fact that there are stong batch effects betweeen different ST sections, so it may be a good idea to integrate the data across sections
```{r}
st.list = list(D0 = P0, D1_1=P31, D1_2=P32, D3_1=P33, D3_2=P34, D5_1=P41,D5_2=P42,D7_1=P11, D7_2=P12, D28=P28, D56=P43 )

# run SCT on both datasets
st.list = lapply(st.list, SCTransform, assay = "Spatial", method = "poisson")
```

# Perform the actual integration
```{r}
# need to set maxSize for PrepSCTIntegration to work
options(future.globals.maxSize = 2000000 * 1024^2)  # set allowed size to 2K MiB

st.features = SelectIntegrationFeatures(st.list, nfeatures = 3000, verbose = FALSE)
st.list <- PrepSCTIntegration(object.list = st.list, anchor.features = st.features,
                              verbose = FALSE)

int.anchors <- FindIntegrationAnchors(object.list = st.list, normalization.method = "SCT",
                                      verbose = FALSE, anchor.features = st.features)
Pyelonephritis.integrated <- IntegrateData(anchorset = int.anchors, normalization.method = "SCT",verbose = FALSE)

#rm(int.anchors, st.list)
#gc()
saveRDS(Pyelonephritis.integrated, file =sprintf("%s/PreliminaryIntegratedSptatil_Pyelonephritis.RDS",outdir))
### due to the integration take a long time, we ignore the previous and rea accordingly
#Pyelonephritis.integrated<-readRDS(file = "PreliminaryIntegratedSptatil_Pyelonephritis.RDS")

```

# Run harmony and clustering with standard pipeline 
```{r}
## change the order:
DefaultAssay(Pyelonephritis.integrated) <- "SCT"
levels(factor(Pyelonephritis.integrated$orig.ident))

Pyelonephritis.integrated$orig.ident<-ordered(factor(Pyelonephritis.integrated$orig.ident), levels=c("PyeloD0_1","PyeloD1_1","PyeloD1_2","PyeloD3_1","PyeloD3_2","PyeloD5_1","PyeloD5_2","PyeloD7_1","PyeloD28_1","PyeloD28_2","PyeloD56_1"))
levels(factor(Pyelonephritis.integrated$orig.ident))

Pyelonephritis.integrated$orig.ident <- as.factor(Pyelonephritis.integrated$orig.ident)
#VariableFeatures(brain.merge) <- c(VariableFeatures(brain), VariableFeatures(brain2))
#install.packages("harmony")

#### Run harmony to remove the batch effects
Pyelonephritis.integrated_harmony <- RunPCA(Pyelonephritis.integrated, verbose = FALSE, npcs=20)
Pyelonephritis.integrated_harmony<- RunHarmony(Pyelonephritis.integrated_harmony, group.by.vars = "orig.ident")
Pyelonephritis.integrated_harmony <- RunUMAP(Pyelonephritis.integrated_harmony,reduction = "harmony", dims = 1:15)
Pyelonephritis.integrated_harmony <- FindNeighbors(Pyelonephritis.integrated_harmony,   dims = 1:15)
Pyelonephritis.integrated_harmony <- FindClusters(Pyelonephritis.integrated_harmony, algorithm = 1,verbose = FALSE, resolution = 0.3)

```

```{r}
Pyelonephritis.integrated_harmony<- readRDS("Pyelonephritis.integrated_harmony.RDS")
selectedslices<- c("slice1","slice3","slice4","slice7","slice8","slice9")

myPalette <- colorRampPalette(rev(brewer.pal(11, "Spectral")))
```

# Draw the dimplot, spatialdimplots
```{r}
# create a fold to save these figures
SpatialClusterFigures <- paste0(outdir, "/SpatialClusterFigures/")
dir.create(SpatialClusterFigures,showWarnings = FALSE)
setwd(SpatialClusterFigures)

pdf("SpatialIntegrated_Pyeloneprhitis_harmony_Dimplot_Seperate.pdf", height = 4, width = 30)
DimPlot(Pyelonephritis.integrated_harmony, reduction = "umap", split.by ="orig.ident",label = TRUE )
dev.off()
#
pdf("SpatialIntegrated_Pyeloneprhitis_harmony_SpatialDimplot_withlabels.pdf", height = 8, width = 40)
SpatialDimPlot(Pyelonephritis.integrated_harmony, label = TRUE,label.size=2.5)
dev.off()
Pyelonephritis.integrated_harmony_highresolution <- FindClusters(Pyelonephritis.integrated_harmony, algorithm = 1,verbose = FALSE, resolution = 1)
pdf("SpatialIntegrated_Pyeloneprhitis_harmony_Dimplot_Seperate_highresolution.pdf", height = 4, width = 30)
DimPlot(Pyelonephritis.integrated_harmony_highresolution, reduction = "umap", split.by ="orig.ident",label = TRUE )
dev.off()

pdf("SpatialIntegrated_Pyeloneprhitis_harmony_SpatialDimplot_withlabels_highresolution.pdf", height = 8, width = 40)
SpatialDimPlot(Pyelonephritis.integrated_harmony_highresolution, label = TRUE,label.size=2.5)
dev.off()

Pyelonephritis.integrated_highersolution.Infection<- subset(Pyelonephritis.integrated_harmony_highresolution, idents = c(25,26))

```

# check the biomarker of Renal pelvis, renal pelvis and medulla, renal medulla and cortex, renal cortex
```{r}
# # create a fold to save these figures
RegionalBioMarkerFigures <- paste0(outdir, "/RegionalBioMarkerFigures")
dir.create(RegionalBioMarkerFigures,showWarnings = FALSE)
setwd(RegionalBioMarkerFigures)

RenalBiomarkers <- c("Upk3a","Psca","Slc4a11","Nfat5","Slc3a1","Prlr","Ren1", "Nphs1", "Havcr1","Il1b","Lcn2")
for (i in RenalBiomarkers){
  BiomarkerSpatialFeature<- SpatialFeaturePlot(Pyelonephritis.integrated_harmony, features = i, pt.size.factor = 1.6, crop = TRUE,  images = selectedslices) &
    ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(-2,8))
  ggsave(filename = paste0(i,"_SpatialFeature.pdf"), plot = BiomarkerSpatialFeature, height = 4, width = 16)
  
}
# change the levels of Subcluster:
DotPlot(Pyelonephritis.integrated_harmony,features = RenalBiomarkers) +RotatedAxis()
```

# we subset the infected regions, cluster 2, 5 and 10
```{r}
## find the graph names
names(Pyelonephritis.integrated_harmony@graphs)
### only select the infection region
Pyelonephritis.integrated.Infection<- subset(Pyelonephritis.integrated_harmony, idents = c(2,4,5,10))

pdf("SpatialIntegrated_Pyeloneprhitis_infected_area.pdf", height =8, width = 40)
SpatialDimPlot(Pyelonephritis.integrated.Infection, crop = TRUE, label = TRUE,images = selectedslices)

dev.off()
pdf("SpatialIntegrated_Pyeloneprhitis_infected_area_Seperated.pdf", height = 4, width = 30)
DimPlot(Pyelonephritis.integrated.Infection, reduction = "umap", split.by ="orig.ident",label = TRUE)
dev.off()
```

```{r}
## find subcluster for Infected region

Subcluster10<-FindSubCluster(Pyelonephritis.integrated_harmony,cluster=c("10"),graph.name = "integrated_snn", subcluster.name = "Sub10",
                             resolution = 0.1)

Subcluster10@meta.data

Idents(Subcluster10) <- Subcluster10@meta.data$Sub10

Subcluster5<-FindSubCluster(Subcluster10,cluster=c("5"),graph.name = "integrated_snn", subcluster.name = "Sub5",
                            resolution = 0.1)
Subcluster5@meta.data
Idents(Subcluster5) <- Subcluster5@meta.data$Sub5

Subcluster2<-FindSubCluster(Subcluster5,cluster=c("2"),graph.name = "integrated_snn", subcluster.name = "Sub2",
                            resolution = 0.05)
Subcluster2@meta.data$Sub2

Idents(Subcluster2) <- Subcluster2@meta.data$Sub2

```


# choose the subset of slides: from "PyeloD0_1", "PyeloD1_2", "PyeloD3_1", "PyeloD5_2", "PyeloD7_1", "PyeloD28_1"
```{r}
# Define the default assay as SCT 
DefaultAssay(Subcluster2) <- "SCT"

# select the subset of spatial transcriptomics
## we only choose the slides from PyeloD0_1, PyeloD1_2, PyeloD3_1, PyeloD5_1, PyeloD7_1, PyeloD28_1
#SlidesSelected_Condtion<-c("PyeloD0_1", "PyeloD1_1", "PyeloD3_1", "PyeloD5_2", "PyeloD7_1", "PyeloD28_1")

FinalAbcessClustersObjSubSlides<-subset(Subcluster2, subset= (orig.ident == "PyeloD0_1" |orig.ident == "PyeloD1_2"|orig.ident == "PyeloD3_1"|orig.ident == "PyeloD5_2"|orig.ident == "PyeloD7_1"|orig.ident == "PyeloD28_1"))

str(FinalAbcessClustersObjSubSlides)

table(FinalAbcessClustersObjSubSlides@meta.data$orig.ident)

```

# change the order of slides
```{r}
## plot the figure
FinalAbcessClustersObjSubSlides$orig.ident<-ordered(factor(FinalAbcessClustersObjSubSlides$orig.ident), levels=c("PyeloD0_1", "PyeloD1_2", "PyeloD3_1", "PyeloD5_2", "PyeloD7_1", "PyeloD28_1"))

FinalAbcessClustersObjSubSlides@meta.data
Idents(FinalAbcessClustersObjSubSlides)

## For all the regions:
Idents(FinalAbcessClustersObjSubSlides) <- FinalAbcessClustersObjSubSlides@meta.data$orig.ident

Idents(FinalAbcessClustersObjSubSlides)
DimPlot(FinalAbcessClustersObjSubSlides, reduction = "umap", group.by = "Sub2",split.by ="orig.ident",label = TRUE,label.size = 2)

```

# redefine the default assay by the SCT using the scale data
```{r}
FinalAbcessClustersObjSubSlides[["LOL"]] <- CreateAssayObject(counts = GetAssayData(FinalAbcessClustersObjSubSlides,
                                                                                    assay="SCT", slot="data")) #Copy "SCT data" to "LOL counts" and "LOL data"
FinalAbcessClustersObjSubSlides <- SetAssayData(FinalAbcessClustersObjSubSlides, assay = "LOL", 
                                                slot = "scale.data", 
                                                new.data = GetAssayData(FinalAbcessClustersObjSubSlides, assay="SCT",
                                                                        slot="scale.data")); gc() #"Copy "SCT scale.data" to "LOL scale.data"
DefaultAssay(FinalAbcessClustersObjSubSlides) <- "LOL"
FinalAbcessClustersObjSubSlides <- DietSeurat(FinalAbcessClustersObjSubSlides, assays = "LOL", counts=T, data=T, scale.data=T)
FinalAbcessClustersObjSubSlides <- RenameAssays(FinalAbcessClustersObjSubSlides, LOL = 'SCT')
## set up default of FinalAbcessClustersObjSubSlides

DefaultAssay(FinalAbcessClustersObjSubSlides) <- "SCT"

# set the identity as the cluster name
Idents(FinalAbcessClustersObjSubSlides) <-FinalAbcessClustersObjSubSlides@meta.data$Sub2
Idents(FinalAbcessClustersObjSubSlides)
```


# Draw the dimplot, spatialdimplots for the subclustering niches
```{r}
# 


Idents(FinalAbcessClustersObjSubSlides)

# change the name of sub2 
pdf("SpatialIntegrated_Pyeloneprhitis_harmony_Dimplot_Seperate.pdf", height = 4, width = 30)
DimPlot(FinalAbcessClustersObjSubSlides, reduction = "umap", split.by ="orig.ident",label = TRUE )
dev.off()
#
pdf("SpatialIntegrated_Pyeloneprhitis_harmony_SpatialDimplot_withlabels.pdf", height = 8, width = 40)
SpatialDimPlot(Subcluster2, label = TRUE,label.size=2.5)
dev.off()
```

# check the biomarker of Renal pelvis, renal pelvis and medulla, renal medulla and cortex, renal cortex
```{r}
# # create a fold to save these figures
RegionalBioMarkerFigures <- paste0(outdir, "/RegionalBioMarkerFigures")
dir.create(RegionalBioMarkerFigures,showWarnings = FALSE)
setwd(RegionalBioMarkerFigures)

RenalBiomarkers <- c("Upk3a","Psca","Slc4a11","Nfat5","Slc3a1","Prlr","Ren1", "Nphs1", "Havcr1","Il1b","Lcn2")
for (i in RenalBiomarkers){
  BiomarkerSpatialFeature<- SpatialFeaturePlot(Pyelonephritis.integrated_harmony, features = i, pt.size.factor = 1.6, crop = TRUE,  images = selectedslices) &
    ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(-2,8))
  ggsave(filename = paste0(i,"_SpatialFeature.pdf"), plot = BiomarkerSpatialFeature, height = 4, width = 16)
  
}
# change the levels of Subcluster:

DotPlot(Subcluster2,features = RenalBiomarkers) +RotatedAxis()
```

### due to the integration take a long time, we ignore the previous and rea accordingly
Pyelonephritis.integrated<-readRDS(file = "PreliminaryIntegratedSptatil_Pyelonephritis.RDS")
## save the integrated 
## change the order:

levels(factor(Pyelonephritis.integrated$orig.ident))

Pyelonephritis.integrated$orig.ident<-ordered(factor(Pyelonephritis.integrated$orig.ident), levels=c("PyeloD0_1","PyeloD1_1","PyeloD1_2","PyeloD3_1","PyeloD3_2","PyeloD5_1","PyeloD5_2","PyeloD7_1","PyeloD7_2","PyeloD28_1","PyeloD56_1"))
levels(factor(Pyelonephritis.integrated$orig.ident))

pdf("SpatialFeaturePlots_Pyeloneprhitis_integrated_NOFeature_v0323.pdf", height = 4, width = 40)
PSpatialFeature_NO <-SpatialFeaturePlot(Pyelonephritis.integrated, features = NULL) + theme(legend.position = "top")
PSpatialFeature_NO

dev.off()

pdf("SpatialFeaturePlots_Pyeloneprhitis_integrated_NOFeature_2_v0323.pdf", height = 4, width = 40)
PSpatialFeature_NO <-SpatialFeaturePlot(Pyelonephritis.integrated, features = NULL, stroke = 0) + theme(legend.position = "top")
PSpatialFeature_NO

dev.off()

#Pyelonephritis.integrated<- RunHarmony(Pyelonephritis.integrated, group.by.vars = "orig.ident")
Pyelonephritis.integrated <- RunPCA(Pyelonephritis.integrated, verbose = FALSE)
Pyelonephritis.integrated <- FindNeighbors(Pyelonephritis.integrated, dims = 1:15)
Pyelonephritis.integrated <- FindClusters(Pyelonephritis.integrated, verbose = FALSE, resolution = 0.2)
Pyelonephritis.integrated <- RunUMAP(Pyelonephritis.integrated, dims = 1:15)


#### Run harmony to remove the 
# Pyelonephritis.integrated_harmony<- RunHarmony(Pyelonephritis.integrated, group.by.vars = "orig.ident")
# Pyelonephritis.integrated_harmony <- RunPCA(Pyelonephritis.integrated_harmony, verbose = FALSE)
# Pyelonephritis.integrated_harmony <- FindNeighbors(Pyelonephritis.integrated_harmony, dims = 1:15)
# Pyelonephritis.integrated_harmony <- FindClusters(Pyelonephritis.integrated_harmony, verbose = FALSE, resolution = 0.4)
# Pyelonephritis.integrated_harmony <- RunUMAP(Pyelonephritis.integrated_harmony,reduction = "harmony", dims = 1:15)
# 

### hereis to have low number of clusters

saveRDS(Pyelonephritis.integrated, file =sprintf("%s/PreliminaryIntegratedSptatil_Pyelonephritis_afterUMAP.RDS",outdir))


pdf("SpatialIntegrated_Pyeloneprhitis_Dimplot.pdf", height = 4, width = 9)
DimPlot(Pyelonephritis.integrated, reduction = "umap", group.by = c("ident", "orig.ident"))
dev.off()

pdf("SpatialIntegrated_Pyeloneprhitis_Dimplot_Seperate.pdf", height = 4, width = 30)
DimPlot(Pyelonephritis.integrated, reduction = "umap", split.by ="orig.ident" )
dev.off()

pdf("SpatialIntegrated_Pyeloneprhitis_Dimplot_Seperate_rotated.pdf", height = 30, width = 4)
DimPlot(Pyelonephritis.integrated, reduction = "umap", split.by ="orig.ident",ncol = 1 )
dev.off()

?DimPlot
pdf("SpatialIntegrated_Pyeloneprhitis_SpatialDimplot.pdf", height = 8, width = 40)
SpatialDimPlot(Pyelonephritis.integrated)
dev.off()



pdf("SpatialIntegrated_Pyeloneprhitis_SpatialDimplot_withlabels.pdf", height = 8, width = 40)
SpatialDimPlot(Pyelonephritis.integrated, label = TRUE,label.size=2.5)
dev.off()

?SpatialDimPlot

pdf("SpatialIntegrated_Pyeloneprhitis_SpatialDimplot_withlabels_rotated.pdf", height = 40, width = 8)
SpatialDimPlot(Pyelonephritis.integrated, label = TRUE,label.size=2.5, ncol=1)
dev.off()

# 
# 
# #################################
# # then merge the objects
# #################################
# # merge did perform worse
# Pyelonephritis<-merge(P0,P31)
# Pyelonephritis<-merge(Pyelonephritis,P32)
# Pyelonephritis<-merge(Pyelonephritis,P33)
# Pyelonephritis<-merge(Pyelonephritis,P34)
# Pyelonephritis<-merge(Pyelonephritis,P41)
# Pyelonephritis<-merge(Pyelonephritis,P42)
# Pyelonephritis<-merge(Pyelonephritis,P11)
# Pyelonephritis<-merge(Pyelonephritis,P12)
# Pyelonephritis<-merge(Pyelonephritis,P28)
# Pyelonephritis<-merge(Pyelonephritis,P43)
# 
# ## save the combined prelimnary dataset in rds
# 
# saveRDS(Pyelonephritis, file =sprintf("%s/PreliminaryMergedSptatil_Pyelonephritis.RDS",outdir))
# 
# 
# ### loading the Pyelonephritis
# 
# Pyelonephritis<-readRDS(file = "PreliminaryMergedSptatil_Pyelonephritis.RDS")
# 

#######################################################
##  2. basical details
#######################################################
#2.1 dimension of the activae assay here is the number of features (genes) by samples (spots)
# dim(P31)
# dim(P32)
# dim(P33)
# dim(P34)
# dim(P41)
# dim(P42)
# dim(P43)
# dim(Pyelonephritis)
### 2.1 dimension of the active assay # the number of feature (genes) by spots (spots)
dim(x=Pyelonephritis.integrated)

nrow(x=Pyelonephritis.integrated) ## the number of features
ncol(x=Pyelonephritis.integrated) ### the number of samples

### 2.2 feature levels and sample name
## check how many spots each spatial
table(Pyelonephritis.integrated$orig.ident)
# calculate the feature (gene) and sample names
head(x=rownames(Pyelonephritis.integrated),n=5)
tail(x=rownames(Pyelonephritis.integrated),n=5)
## sample name 
tail(x=colnames(Pyelonephritis.integrated),n=5)

## 2.3. Sample level metadata
class(Pyelonephritis.integrated[[]])
colnames(Pyelonephritis.integrated[[]])
head(Pyelonephritis.integrated@meta.data)

## nFeature spatial; the number of unique gene in each sample

sum(Pyelonephritis.integrated$nFeature_Spatial == colSums(Pyelonephritis.integrated@assays$SCT@counts >0))


## consist of one or more assay objects as individual represetation of single-cell expression data,
## Each Assay stores multiple slots, including raw (counts), normalized (data) and scaled data (scaled.data) as well as a vector of variable features (var.features) and feature-level metadata (meta.features).
#Pyelonephritis.integrated@assays$Spatial@meta.features

## check the mitochondrial gene list
#grep (pattern = "^hb", x=(rownames(P31)), ignore.case=TRUE, value = TRUE)

# 2.3 sample level metadata
class(P31[[]])
colnames(P31[[]])
colnames(Pyelonephritis.integrated[[]])
tail(rownames(Pyelonephritis.integrated[[]]))
# 2.4 the number of unique genes in each sample
#colSums(P31@assays$Spatial@counts)
sum(P31$nFeature_Spatial == colSums(P31@assays$Spatial@counts>0))
# 2.5 the total numebr of detected molecules in each sample
sum(P31$nCount_Spatial == colSums(P31@assays$Spatial@counts))

## 2.4 objects (e.g. Assaty) together with feature level metadata
# a vector of names of assiacted objects 
names(x=P31)

# Spatial details
Pyelonephritis.integrated[['Spatial']] # equivalent to P31@assays$Spatial
# slice detais for image key
Pyelonephritis.integrated[['slice1']] # equivalent to P31@assays$slice1

Pyelonephritis.integrated[['Spatial']]

# each seurat object consists of one or more assay objects (basis unit of seurat ) as individual representation of single-cell expression data
# example of the assay objuect include RNA, ADT in CITE-seq, or Spatial. Each assay store multiple slots, including raw (counts), normalized (data) and scaled data as well as 
# a vector of varible features (var.features) and feature-level metadata
Pyelonephritis.integrated@assays$Spatial ## 
Pyelonephritis.integrated[['Spatial']]@counts[5:10,1:3]
Pyelonephritis.integrated[['Spatial']]@counts[5:10,1:3]
# Feature level metadata is associated with each individual assay



#####################################################################################################################
### 3. Generated a matrix for the control datasets to evaluate the number of spatial counts and number of features
#####################################################################################################################

#Then we run dimensionality reduction and clustering as before.
DefaultAssay(Pyelonephritis.integrated) <- "SCT"

### 
Pyelonephritis.integrated[[]]

# quality control

Pyelonephritis.integrated<-PercentageFeatureSet(Pyelonephritis.integrated, "^mt-", col.name = "percent_mito")
Pyelonephritis.integrated<- PercentageFeatureSet(Pyelonephritis.integrated, "^Hb.*-", col.name = "percent_hb")
Pyelonephritis.integrated<- PercentageFeatureSet(Pyelonephritis.integrated, "^Rp.*", col.name = "percent_rRNA")

# add the group details and change the order of the groups


Pyelonephritis.integrated[["group"]] = Pyelonephritis.integrated$orig.ident

levels(factor(Pyelonephritis.integrated[[]]$group))

## measure how many spots 
# create a list
TotalSpot<- data.frame(name=row.names(table(Pyelonephritis.integrated$group)), value=as.numeric(table(Pyelonephritis.integrated$group)))

sum(Pyelonephritis.integrated[[]]$nFeature_Spatial)/nrow(Pyelonephritis.integrated[[]])/ (sum(Pyelonephritis.integrated[[]]$nCount_Spatial)/sum(Pyelonephritis.integrated[[]]$nFeature_Spatial))

## chagne the order the rowname

TotalSpot$name<-ordered(factor(TotalSpot$name), levels=c("PyeloD0_1","PyeloD1_1","PyeloD1_2","PyeloD3_1","PyeloD3_2","PyeloD5_1","PyeloD5_2","PyeloD7_1","PyeloD7_2","PyeloD28_1","PyeloD56_1"))

TotalSpot
###use the default colors
library(scales)
###plot total number of spots
cols = gg_color_hue(11)
## plot the spots number distibution
pdf("SpotsPlots_Pyeloneprhitis_integrated_v0323.pdf", height = 3, width = 6)

#show_col(hue_pal()(16))

# These two are equivalent; by default scale_fill_hue() is used
# These two are equivalent; by default scale_colour_hue() is used
pspots<- ggplot(data=TotalSpot, aes(x=name, y=value, fill=name)) +geom_bar(stat="identity") +theme_classic()+theme(axis.text.x = element_text(angle = 45, hjust=1))+ scale_fill_hue() + ylim(0, 3000)

#+scale_color_brewer(palette="Spectral")
print (pspots)

dev.off()

### plot the number of counts, numbe of features in the spatial 

pdf("FeaturePlots_Pyeloneprhitis_integrated_v0323.pdf", height = 4, width = 20)
PVlnFeature<- VlnPlot(Pyelonephritis.integrated, features = c("nCount_Spatial", "nFeature_Spatial", "percent_rRNA",
                                                   "percent_hb"), group.by ="group", pt.size = 0, ncol = 4) + NoLegend()

PVlnFeature
dev.off()

pdf("FeaturePlots_Pyeloneprhitis_integrated_withSpots_v0323.pdf", height = 4, width = 20)
PVlnFeature<- VlnPlot(Pyelonephritis.integrated, features = c("nCount_Spatial", "nFeature_Spatial", "percent_rRNA",
                                                              "percent_hb"), group.by ="group", pt.size = 0, ncol = 4) + NoLegend()

PVlnFeature + pspots
dev.off()

pdf("SpatialFeaturePlots_Pyeloneprhitis_integrated_SpatialFeatureCounts_v0323.pdf", height = 12, width = 40)
PSpatialFeature <-SpatialFeaturePlot(Pyelonephritis.integrated, features = c("nCount_Spatial", "nFeature_Spatial", "percent_rRNA","percent_hb")) 
# ggplot2::scale_fill_continuous(limits = c(0.0,1.0), breaks = c(0.0, 1.0))
## the following can scale the legend size
#++ ggplot2::scale_fill_continuous(limits = c(0.0,1.0),, breaks = c(0.0, 0.5, 1.0))
PSpatialFeature

dev.off()

?SpatialFeaturePlot

################################################################
# Fig S. the cell-type cluster proportion across different time points
################################################################
Pyelonephritis.integrated@meta.data$seurat_clusters
### add the cluster proportion within all the datasets
table(Pyelonephritis.integrated@meta.data)
## 4.3 add the proportion of each cluster and show on the annotated clusters from all the datasets
length(Pyelonephritis.integrated@meta.data$seurat_clusters)
df<- data.frame(clu=names(table(Pyelonephritis.integrated@meta.data$seurat_clusters)), totalnumber=sprintf("%1.0f",table(Pyelonephritis.integrated@meta.data$seurat_clusters)), percentage= sprintf("%1.2f", 100*table(Pyelonephritis.integrated@meta.data$seurat_clusters)/length(Pyelonephritis.integrated@meta.data$seurat_clusters)))
names(table(Pyelonephritis.integrated@meta.data$seurat_clusters))
table(Pyelonephritis.integrated@meta.data$seurat_clusters)
df
## then add the proportion into the meta data
Pyelonephritis.integrated@meta.data$TotalclusterPercent <- df[match(Pyelonephritis.integrated@meta.data$seurat_clusters,df$clu),3]
Pyelonephritis.integrated@meta.data$TotalclusterNumber <-df[match(Pyelonephritis.integrated@meta.data$seurat_clusters,df$clu),2]

### add the cluster proportion within differnt time points
Pyelonephritis.integrated@meta.data$group
## add the proportion of each cluster at different time points
Pyelonephritis.integrated@meta.data$TimepointCluster<-paste0(Pyelonephritis.integrated@meta.data$group,":",Pyelonephritis.integrated@meta.data$seurat_clusters)
#length(Pyelonephritis.integrated@meta.data %>% filter (`group` == strsplit(Pyelonephritis.integrated@meta.data$TimepointCluster,":") [[1]][1]) %>% pull(group))

## make a data frame to store the different proportion
table(Pyelonephritis.integrated@meta.data$TimepointCluster)
TimeDf <- data.frame(Timecluster=names(table(Pyelonephritis.integrated@meta.data$TimepointCluster)), percentage= sprintf("%1.2f", 100*table(Pyelonephritis.integrated@meta.data$TimepointCluster)/length(Pyelonephritis.integrated@meta.data %>% filter (`group` == strsplit(Pyelonephritis.integrated@meta.data$TimepointCluster,":") [[1]][1]) %>% pull(group))))
TimeDf

## we then added the TimeClusterPercentage
Pyelonephritis.integrated@meta.data$TimeClusterPercentage <- TimeDf[match(Pyelonephritis.integrated@meta.data$TimepointCluster, TimeDf$Timecluster),2]
Pyelonephritis.integrated@meta.data

Idents(Pyelonephritis.integrated)

pdf("Frequency_split_cluster_V0322.pdf", width = 15, height = 4)
ggplot(Pyelonephritis.integrated@meta.data, aes(fill=Pyelonephritis.integrated@meta.data$group, y=as.numeric(Pyelonephritis.integrated@meta.data$TimeClusterPercentage), x=Idents(Pyelonephritis.integrated))) + 
  geom_bar(position="dodge", stat="identity")+ theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                                                  axis.text.x = element_text(vjust = 1, hjust=1),
                                                                  panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) + scale_y_continuous(limits=c(0,40)) 
dev.off()

pdf("FeaturePlots_Pyeloneprhitis_UmapDimPlot_Withlabels_V0322.pdf", width = 5, height = 4)
DimPlot(Pyelonephritis.integrated, label = TRUE,reduction = "umap",  pt.size=0.1)

dev.off()

#df[match(Pyelonephritis.integrated@meta.data$seurat_clusters,df$clu),3]
## set up the default to SCT

## enable joint dimensional reduction and clustering on the undelyiing RNA seq expression data
# DefaultAssay(Pyelonephritis.integrated) <- "SCT"
# 
# st.list = list(D0 = P0, D1_1=P31, D1_2=P32, D3_1=P33, D3_2=P34, D5_1=P41,D5_2=P42,D7_1=P11, D7_2=P12, D28=P28, D56=P43 )
# VariableFeatures(Pyelonephritis.integrated) <-c(VariableFeatures(P0), VariableFeatures(P31),VariableFeatures(P32),VariableFeatures(P33),VariableFeatures(P34),VariableFeatures(P41),VariableFeatures(P42), VariableFeatures(P43),VariableFeatures(P11),VariableFeatures(P12),VariableFeatures(P28))
# #VariableFeatures(Pyelonephritis.integrated) 
# 

#####################################################################################################################
### 4. Normalization the variance of molecular counts expresses, dimension reduction
#####################################################################################################################
### without the reference
## normalized bu SCTranform and check the structure
## SCTransform returns the Seurat object with a new assay called SCT, where its counts slot stores the corrected UMI counts, the data slot stores the log-normalized version of the corrected UMI counts, and scale.data slot stores the pearson residuals (normalized values) and is used as PCA input.

dim(Pyelonephritis.integrated@assays$SCT@counts)
dim(Pyelonephritis.integrated@assays$SCT@data) 
dim(Pyelonephritis.integrated@assays$SCT@scale.data) 
#

## Computes the correlation of the log normalized data and sctransform residuals with the
# # number of UMIs
# Pyelonephritis_norm <- GroupCorrelation(Pyelonephritis_norm, group.assay = "Spatial", assay = "Spatial", slot = "data", do.plot = FALSE)
# Pyelonephritis_norm <- GroupCorrelation(Pyelonephritis_norm, group.assay = "Spatial", assay = "SCT", slot = "scale.data", do.plot = FALSE)
# 
# p1 <- GroupCorrelationPlot(Pyelonephritis_norm, assay = "Spatial", cor = "nCount_Spatial_cor") + ggtitle("Log Normalization") +
#   theme(plot.title = element_text(hjust = 0.5))
# p2 <- GroupCorrelationPlot(Pyelonephritis_norm, assay = "SCT", cor = "nCount_Spatial_cor") + ggtitle("SCTransform Normalization") +
#   theme(plot.title = element_text(hjust = 0.5))
# p1 + p2

# ## Run PCA 
# 
# DefaultAssay(Pyelonephritis) <-"SCT"
# VariableFeatures(Pyelonephritis) <-c (VariableFeatures(P31),VariableFeatures(P32),VariableFeatures(P33),VariableFeatures(P34),VariableFeatures(P41),VariableFeatures(P42), VariableFeatures(P43))
# Pyelonephritis <- RunPCA(Pyelonephritis, assay = "SCT", verbose = FALSE)
# # compute K nearest neighbors (KNN)
# Pyelonephritis <- FindNeighbors(Pyelonephritis, reduction = "pca", dims = 1:20)
# # Leiden algorithm for community detection
# Pyelonephritis <- FindClusters(Pyelonephritis, verbose = FALSE)
# # PCA result is the default UMAP input, use dimensions 1:30 as input features
# Pyelonephritis <- RunUMAP(Pyelonephritis, reduction = "pca", dims = 1:20)
# 
# pdf(file = "Pyeonephritis_umap_clustered_Dimplot.pdf",  height = 12, width = 40)
# plot3 <- DimPlot(Pyelonephritis_norm, reduction = "umap", label = TRUE, split.by = "group") + NoLegend()
# plot3
# dev.off()
# pdf(file = "Pyeonephritis_Spactial_clustered_SpatialDimplot.pdf",  height = 12, width = 40)
# plot4 <- SpatialDimPlot(Pyelonephritis_norm, label = TRUE, label.size = 3) + NoLegend()
# plot4
# dev.off()

Idents(Pyelonephritis.integrated)

# save the final integrated into rds file
saveRDS(Pyelonephritis.integrated, file = "Pyelonephritis.integrated_Beforedeconvolution_V0325.rds")

# read the rds file

Pyelonephritis.integrated<- readRDS(file = "Pyelonephritis.integrated_Beforedeconvolution_V0325.rds")
