---
title: "Single Cell cell-cell communication at day 3"
output: html_document
Author: Xin Wang
Email: xin.wang@nationwidechildrens.org 
Copyright (c) 2024 Kidney and Urology Tract Center
Nationwide Children's Hospital
date: "2024-08-23"
Description: 
#   This script is to profile the cell-cell communications in the single cell resolution.

#     Note:
#    #     We will try to concentrate on the immune cells, especially the chemikine and cytokine, and 
#.   #     We will identify the other epithelial cells that result in this interactions.
#    #     We will trace the interaction within the different timepoints

#     Function:
#           1). Generate the ligand and receptor interaction in the 
#           2).  Generate the deconvoluted cell types of these abscess regions 
#   
#           8). 
#     Method:  
#         Using the cellchat
#     Reference: https://github.com/jinworks/CellChat
---

# load the library packages
```{r}
# loading the library packages
library(Seurat)
#library(SeuratData)
library(ggplot2)
library(patchwork)
library(dplyr)
library(Matrix)
library(tidyverse)
library(SeuratObject)
#install.packages("harmony")
library(harmony)
library(SPOTlight)
#library(scater)
#library(scran)
library(SingleCellExperiment)

#install.packages("SeuratDisk")
#library(SeuratDisk)
library(tidyverse)
library(cowplot)

library(ggplot2)
library(reshape2)
library(Hmisc)
library(stats)
library(corrplot)
library(RColorBrewer)
## Deconvolve
#install.packages("L1pack")
#install.packages("reshape")
#install.packages("local/path/to/Matrix_1.6-4.tar.gz", repos = NULL, type="source")
library("Matrix")
library(L1pack)
library(reshape)
library(reshape2)

library(SPOTlight)
library(tidyHeatmap)
library(ComplexHeatmap)
#library(ComplexHeatmap)

library(RColorBrewer)
library(circlize)
library(monocle3)
#install.packages("magick")
library(magick)
library(progeny)
library(pheatmap)
library(clusterProfiler)
library(decoupleR)
library(CellChat)
library(patchwork)
options(stringsAsFactors = FALSE)
options(future.globals.maxSize = 891289600000) 

```

# Set up the enviroment: 
```{r}
# clear all the objects in the environments
#rm(list=ls())
# set seed
set.seed(200000)

# single cell directory:
Scindir = "/Users/XXW004/Documents/Projects/RuizRosado/PIPseq/Results/Evaluation_V022723/CellEvaluation_Subclustering_V0803/FinalSingleCellObj_V08012024/"

# spatial transcript directory that contain the deconvoluted file 
# Spindir ="/Users/XXW004/Documents/Projects/RuizRosado/SpatialTranscriptomic/Results/Deconvolution_Six_V07292024/"

# The cell-cell communication output directory for day 7:
Outdir = "/Users/XXW004/Documents/Projects/RuizRosado/PIPseq/Results/Evaluation_V022723/CellEvaluation_Subclustering_V0803/Cell_CellCommunication/OverallDay28/"

data_path = "/Users/XXW004/Documents/Projects/LectureForSingleCellGroup/Datasets/NicheNet/"



getwd()
setwd(Outdir)
```

<!-- # Read the RDS objects from the spatial and single cells, briefly check the objects -->
<!-- ```{r} -->
<!-- # Reading the RDS file -->
<!-- Pyelonephritis_Infection<-readRDS(file = paste0(Spindir, "PreliminaryIntegratedSpatialHarmorny_Pyelonephritis_onlySlide_QC_v0906.RDS")) -->

<!-- #  -->
<!-- Pyelonephritis_Infection -->

<!-- SpatialDimPlot(Pyelonephritis_Infection) -->
<!-- # We have two assays: Spatial and SCT -->
<!-- Pyelonephritis_Infection[["Spatial"]][1:10,1:10] -->

<!-- Pyelonephritis_Infection[["SCT"]][1:10,1:10] -->

<!-- rownames(Pyelonephritis_Infection[["Spatial"]]) -->
<!-- colnames(Pyelonephritis_Infection[["Spatial"]]) -->
<!-- selectedslices<- c("slice1","slice3","slice4","slice7","slice8","slice9") -->
<!-- Idents(Pyelonephritis_Infection) -->

<!-- DimPlot(Pyelonephritis_Infection) -->
<!-- ``` -->

# read the final single cell object with detailed annotation
```{r}
#rm(list=ls())
#saveRDS(file = "Pyelonephritis_Infection_Deconvoluted.RDS", object = Pyelonephritis_Infection)
# read the single cell
SingleCellObj <- readRDS(file = paste0(Scindir,"PIPSeq_Final_Annotated_v08012024.rds"))
SingleCellObj@meta.data %>% head()
# check the dimplot
DimPlot(SingleCellObj, split.by = "DataSet")

```
# ############################## 
# We concentrate on date 28
# ############################## 
```{r}

# we  select the all dates
SingleCellObj$DataSet
Pyelonephritis_InfectionDay28 <- 
    subset (SingleCellObj, subset= (DataSet == "28DPI"))
Pyelonephritis_InfectionDay28
nrow(Pyelonephritis_InfectionDay28@meta.data)
Pyelonephritis_InfectionDay28 %>% head()
### create a cellchat object

cellchat <- createCellChat(object = Pyelonephritis_InfectionDay28, group.by = "FinalAnnotation", assay = "RNA")

cellchat
# set the database#
CellChatDB <- CellChatDB.mouse # use CellChatDB.mouse if running on mouse data
showDatabaseCategory(CellChatDB)

# subset the expression data of signaling genes for saving computation cost
cellchat@DB <- CellChatDB

#Preprocessing the expression data for cell-cell communication analysis
# subset the expression data of signaling genes for saving computation cost
#cellchat <- subsetData(cellchat) # This step is necessary even if using the whole 
cellchat <- subsetData(cellchat) # This step is necessary even if using the whole database

#future::plan("multisession", workers = 4) # do parallel

cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat)
#> The number of highly variable ligand-receptor pairs used for signaling inference is 692

#execution.time = Sys.time() - ptm
# print(as.numeric(execution.time, units = "secs"))

#### Part II: Inference of cell-cell communication network
ptm = Sys.time()
cellchat <- computeCommunProb(cellchat, type = "triMean")

# save the object of all the cells

saveRDS(cellchat, file = "SingleCell_Date28_Cellchat.rds")

# read the object
cellchat<- readRDS("SingleCell_Date28_Cellchat.rds")
# then we fileterd the cellgroup for cell cell communintion is 10
cellchat <- filterCommunication(cellchat, min.cells = 3)

# Then extracte the infered celllular communciation network as a dataframe
df.net <- subsetCommunication(cellchat)
#df.net <- subsetCommunication(cellchat, sources.use = c(1,2), targets.use = c(4,5))
#df.net <- subsetCommunication(cellchat, signaling = c("WNT", "TGFb"))
df.net


# Infer the cell-cell communication at a signaling pathway level
# CellChat computes the communication probability on signaling pathway level by summarizing the communication probabilities of all ligands-receptors interactions associated with each signaling pathway.

# NB: The inferred intercellular communication network of each ligand-receptor pair and each signaling pathway is stored in the slot ‘net’ and ‘netP’, respectively.
cellchat <- computeCommunProbPathway(cellchat)

# Calculate the aggregated cell-cell communication network
cellchat <- aggregateNet(cellchat)
ptm = Sys.time()
groupSize <- as.numeric(table(cellchat@idents))
pdf("SignalingSedingFromAllCellTypes_Day28.pdf", height = 6, width = 12)

par(mfrow = c(1,2), xpd=TRUE)
netVisual_circle(cellchat@net$count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Number of interactions")
netVisual_circle(cellchat@net$weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")
dev.off()
# Due to the complicated cell-cell communication network, we can examine the signaling sent from each cell group. Here we also control the parameter edge.weight.max so that we can compare edge weights between differet networks.
mat <- cellchat@net$weight
nrow(mat)
pdf("SignalingSedingFromEachCelltype_Day28.pdf", height = 40, width = 40)
par(mfrow = c(5,7), xpd=TRUE)
for (i in 1:nrow(mat)) {
  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
  mat2[i, ] <- mat[i, ]
  netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i])
}
dev.off()

# Part III: Visualization of cell-cell communication network
cellchat@netP$pathways

# 
pathways.show <- c("CXCL","CCL","TNF") 
# Hierarchy plot
# Here we define `vertex.receive` so that the left portion of the hierarchy plot shows signaling to fibroblast and the right portion shows signaling to immune cells 
vertex.receiver = seq(1,4) # a numeric vector. 

#cellchat@net$LRs
pdf("Pathwayshow_Day28.pdf", height = 6, width = 6)
netVisual_aggregate(cellchat, signaling = pathways.show,  vertex.receiver = vertex.receiver)
dev.off()
# Circle plot
pdf("Pathwayshow_Cxcl_CCL_Day28.pdf", height = 6, width = 6)
par(mfrow=c(1,1))
netVisual_aggregate(cellchat, signaling = pathways.show, layout = "circle")
dev.off()
# using the chord diagram
pdf("Pathwayshow_Cxcl_CCL_Day28_Chord.pdf", height = 5, width = 5)
netVisual_aggregate(cellchat, signaling = pathways.show, layout = "chord")
dev.off()

 
 # save the netAnalysis contritbution
Test<- netAnalysis_contribution(cellchat, signaling = pathways.show, return.data=T)

# save the netAnalysis contribution for Cytokine and Chemokine
write.csv(Test$LR.contribution, file = "Chemokine_Cytokine_Cell_Contribution_date28.csv")

# Compute the contribution of each ligand-receptor pair to the overall signaling pathway and visualize cell-cell communication mediated by a single ligand-receptor pair
# generate the ligand receptor pairs
pdf("Pathwayshow_Cxcl_CCL_Day28_netAnalysis_contribution.pdf", height = 5, width = 4)
netAnalysis_contribution(cellchat, signaling = pathways.show)
dev.off()
pairLR.CXCL <- extractEnrichedLR(cellchat, signaling = pathways.show, geneLR.return = FALSE)
pairLR.CXCL$interaction_name

# generate the potential interactions
for (i in pairLR.CXCL$interaction_name){
  pdf(paste0(i,"PairedPathwayshow_Chord_Day28.pdf"), height = 4, width = 4)
  netVisual_individual(cellchat, signaling = pathways.show, pairLR.use = i, layout = "chord")
  dev.off()
}

LR.show <- i # show one ligand-receptor pair
LR.show
# Hierarchy plot
vertex.receiver = seq(1,4) # a numeric vector
netVisual_individual(cellchat, signaling = pathways.show,  pairLR.use = LR.show, vertex.receiver = vertex.receiver)
#> [[1]]
# Circle plot

# gernerate the interaction of two pairs
#CXCL1_CXCR2

netVisual_individual(cellchat, signaling = pathways.show, pairLR.use = LR.show, layout = "circle")
pdf("PairedPathwayshow_CXCL1_CXCR2_Chord_Day28.pdf", height = 5, width = 5)
netVisual_individual(cellchat, signaling = pathways.show, pairLR.use = LR.show, layout = "chord")
dev.off()

pdf("PairedPathwayshow_CCL_CXCL_DotPlots_Day28.pdf", height = 5, width = 10)
netVisual_bubble(cellchat, signaling = c("CCL","CXCL"), remove.isolate = FALSE)
dev.off()
# visulaize cell-cell communication mediated by ligand and receptors

# save the aggregated cell-cell communication
saveRDS(cellchat, file = "cellchat_PN_D28_final.rds")
cellchat
```







# We generate the contribution of the ligand-receptor from CCL, CXCL and TNF, comparing the different dates

```{r}
# we need to perform the contribution of each L-R pair from day 0 and day 7
rm(list=ls())
# 
InfectedDates <- levels(SingleCellObj@meta.data$DataSet)

# we select each dates and perform the cellchat
seurat.obj.day0 <- subset (SingleCellObj, subset= (DataSet == "0DPI"))

cellChatDay0 <- createCellChat(object = seurat.obj.day0, group.by = "DataSet", assay = "RNA")


# set the database#
CellChatDB <- CellChatDB.mouse # use CellChatDB.mouse if running on mouse data
showDatabaseCategory(CellChatDB)

# subset the expression data of signaling genes for saving computation cost
cellChatDay0@DB <- CellChatDB

#Preprocessing the expression data for cell-cell communication analysis
# subset the expression data of signaling genes for saving computation cost
#cellchat <- subsetData(cellchat) # This step is necessary even if using the whole 
cellChatDay0 <- subsetData(cellChatDay0) # This step is necessary even if using the whole database

#future::plan("multisession", workers = 4) # do parallel

cellChatDay0 <- identifyOverExpressedGenes(cellChatDay0)
cellChatDay0 <- identifyOverExpressedInteractions(cellChatDay0)
#> The number of highly variable ligand-receptor pairs used for signaling inference is 692

#execution.time = Sys.time() - ptm
# print(as.numeric(execution.time, units = "secs"))

#### Part II: Inference of cell-cell communication network
ptm = Sys.time()
cellChatDay0 <- computeCommunProb(cellChatDay0, type = "triMean")


#### day 7
# we select each dates and perform the cellchat
seurat.obj.day7 <- subset (SingleCellObj, subset= (DataSet == "7DPI"))

cellChatDay7 <- createCellChat(object = seurat.obj.day7, group.by = "DataSet", assay = "RNA")


# set the database#
CellChatDB <- CellChatDB.mouse # use CellChatDB.mouse if running on mouse data
showDatabaseCategory(CellChatDB)

# subset the expression data of signaling genes for saving computation cost
cellChatDay7@DB <- CellChatDB

#Preprocessing the expression data for cell-cell communication analysis
# subset the expression data of signaling genes for saving computation cost
#cellchat <- subsetData(cellchat) # This step is necessary even if using the whole 
cellChatDay7 <- subsetData(cellChatDay7) # This step is necessary even if using the whole database

#future::plan("multisession", workers = 4) # do parallel

cellChatDay7 <- identifyOverExpressedGenes(cellChatDay7)
cellChatDay7 <- identifyOverExpressedInteractions(cellChatDay7)
#> The number of highly variable ligand-receptor pairs used for signaling inference is 692

#execution.time = Sys.time() - ptm
# print(as.numeric(execution.time, units = "secs"))

#### Part II: Inference of cell-cell communication network
ptm = Sys.time()
cellChatDay7 <- computeCommunProb(cellChatDay7, type = "triMean")






object.list <- list(NL = cellChatDay0, LS = cellChatDay7)
cellchat <- mergeCellChat(object.list, add.names = names(object.list))

```


















# ############################## 
# Generate the deconvoluted details
# ############################## 
```{r}
setwd(CDecodir)
# the prediction of the object
Pyelonephritis_InfectionDay28[["predictions"]]
InfectedMat_CorMed<- t(Pyelonephritis_InfectionDay28[["predictions"]]@data)
colnames(InfectedMat_CorMed) <- ordered (colnames(InfectedMat_CorMed), levels =CelltypeOrder)

# change the order
InfectedMat_CorMed<- InfectedMat_CorMed[ ,CelltypeOrder]
# for final celltype matrix: 
InfectedMat_CorMedCellType<- InfectedMat_CorMed[,!colnames(InfectedMat_CorMed) %in% c("max")]
InfectedMat_CorMedCellTypeNozero <- InfectedMat_CorMedCellType[,colSums(InfectedMat_CorMedCellType)>0]


# correlation
InfectedMat_CorMedCellTypeCorrelationMatrix<-plotCorrelationMatrix(InfectedMat_CorMedCellTypeNozero)
InfectedMat_CorMedCellTypeCorrelationMatrix
InfectedMat_CorMedCellTypeInteractions<- plotInteractions(InfectedMat_CorMedCellTypeNozero, "heatmap") + scale_fill_gradientn(colors = c("blue", "white", "red"))

ggsave("SpatialScatterPieInfectedCorMed_CorrelationMatrix.pdf", plot =InfectedMat_CorMedCellTypeCorrelationMatrix, height = 6, width = 6 )

ggsave("SpatialScatterPieInfectedCorMed_HeatmapColocalization.pdf", plot =InfectedMat_CorMedCellTypeInteractions, height = 6, width = 6 )
# scatterpie plot
# scatterpie plot
# 
ct <- colnames(InfectedMat_CorMedCellType)
ct
# we did not add the low proportion in the scatterpie, we here also consider the low 
#InfectedMat_CorMedCellType[InfectedMat_CorMedCellType < 0.1] <- 0

  paletteMartin <- c( 
    "#b6dbff", # PT-S1S2-N1
    "#99CCFF", # PT-S1S2-N2
    "#0099FF", # PT-S3-1
    "#6699FF", # PT-S3-2
    "#006ddb", # PT-S1S2-N1
    "#EE4E34", # PT-S1S2-N2
    "#66CC99", # PT-S1S2-N3
    "#F5DEB3", # InfPT
    "#A8554E", # ProfPT
    "#ff6db6", #PC1
    "#FF6666", #PC2
    "#CCF381", #ICA
    "#CCF101", #ICB
    "#BEE5BF", #DegCD
    "#33FFFF", # DTL
    "#b66dff", #TAL
    "#FF9900", #DCT-TAL
    "#FFD700", # CNT
    "#ADE700", # SMC
    "#C5DECD", #URO
    "#66Ff66", #DegURO
    "#9999CC", #FIB
    "#FF0099", #END
    "#8ad8e8", #PL
    "#F3ED87",#ResMAC
    "#F6874F",#MONa
    "#F3AB87",#MONb
    "#FFFF00",#MACa
    "#EFEB15",#MACb
    "black", #NEU
    "#ebece5",# T
   "#f47a22",# B 
   "#9b59b6"#NK 
  )

#DefaultAssay(Pyelonephritis_InfectionDay28) <- "Spatial"

# choose three color
# pal<-colorRampPalette(c("LightBlue","white","FireBrick"))(length(ct))
# choose the selected color
pal <- colorRampPalette(paletteMartin)(length(ct))

# choose the spectral range colors:
# https://earlglynn.github.io/RNotes/package/RColorBrewer/index.html
#pal <- (colorRampPalette(brewer.pal(11,"Paired"))(length(ct)))

names(pal) <- ct
#pal
rownames(Pyelonephritis_InfectionDay28)
colnames(Pyelonephritis_InfectionDay28)
str(Pyelonephritis_InfectionDay28)
#SPOTlight:::. extract_image(Pyelonephritis_Infection )
#rownames(Pyelonephritis_Infection)
SpatialScatterPieCorMed <-plotSpatialScatterpie(
  x= Pyelonephritis_InfectionDay28,
  y= InfectedMat_CorMedCellType,
  cell_types = colnames(InfectedMat_CorMedCellType),
    img = TRUE,
    scatterpie_alpha = 1,
    pie_scale = 0.8) +
    scale_fill_manual(
        values = pal,
        breaks = names(pal)) 
#+ coord_sf(xlim = c(130,260), ylim = c(330, 460))

SpatialScatterPieCorMed

ggsave("SpatialScatterPieCorMed.pdf",plot = SpatialScatterPieCorMed, height = 3,width = 4.5)
# colnames(DeconvolutedMatrix)
# ncol(Pyelonephritis_Infection)
# nrow(DeconvolutedMatrix)
InfectedMat_CorMedCellType
pdf(file = "SpatialScatterPieCorMed_CellTypeNetwork.pdf", height = 8,width = 8)
plotInteractions(InfectedMat_CorMedCellType, "network")
dev.off()

```


# ############################## 
#  Perform the cell-cell communication using CellChat
# ############################## 
```{r}

setwd(CNichComdir)

# we only concentrate on the cell-cell communication for Day 28 pelvic

Seurat::SpatialDimPlot(Pyelonephritis_InfectionDay28,pt.size.factor = 1500)
# Prepare input data for CelChat analysis
data.input = Seurat::GetAssayData(Pyelonephritis_InfectionDay28, slot = "data", assay = "SCT") # normalized data matrix
#> Warning: The `slot` argument of `GetAssayData()` is deprecated as of SeuratObject 5.0.0.
Seurat::Idents(Pyelonephritis_InfectionDay28)
# Define the meta data: 
# a column named `samples` should be provided for spatial transcriptomics analysis, which is useful for analyzing cell-cell communication by aggregating multiple samples/replicates. Of note, for comparison analysis across different conditions, users still need to create a CellChat object seperately for each condition.  
meta = data.frame(labels = Seurat::Idents(Pyelonephritis_InfectionDay28), samples = "sample1", row.names = names(Seurat::Idents(Pyelonephritis_InfectionDay28))) # manually create a dataframe consisting of the cell labels
meta

meta$samples <- factor(meta$samples)
unique(meta$labels) # check the cell labels
#> [1] L2/3 IT Astro   L6b     L5 IT   L6 IT   L6 CT   L4      Oligo  
#> Levels: Astro L2/3 IT L4 L5 IT L6 IT L6 CT L6b Oligo
unique(meta$samples) # check the sample labels
#> [1] sample1
#> Levels: sample1


# load spatial transcriptomics information
# Spatial locations of spots from full (NOT high/low) resolution images are required. For 10X Visium, this information is in `tissue_positions.csv`. 
spatial.locs = Seurat::GetTissueCoordinates(Pyelonephritis_InfectionDay28, scale = NULL, cols = c("imagerow", "imagecol")) 

# Spatial factors of spatial coordinates
# For 10X Visium, the conversion factor of converting spatial coordinates from Pixels to Micrometers can be computed as the ratio of the theoretical spot size (i.e., 65um) over the number of pixels that span the diameter of a theoretical spot size in the full-resolution image (i.e., 'spot_diameter_fullres' in pixels in the 'scalefactors_json.json' file). 
# Of note, the 'spot_diameter_fullres' factor is different from the `spot` in Seurat object and thus users still need to get the value from the original json file. 
scalefactors = jsonlite::fromJSON(txt = file.path("/Users/XXW004/Documents/Projects/RuizRosado/SpatialTranscriptomic/Datasets/VisiumRun1/acute_pyelo/seurat", 'scalefactors_json.json'))
spot.size = 65 # the theoretical spot size (um) in 10X Visium
conversion.factor = spot.size/scalefactors$spot_diameter_fullres
spatial.factors = data.frame(ratio = conversion.factor, tol = spot.size/2)



d.spatial <- computeCellDistance(coordinates = spatial.locs, ratio = spatial.factors$ratio, tol = spatial.factors$tol)
min(d.spatial[d.spatial!=0]) # this value should approximately equal 100um for 10X Visium data
#> [1] 99.52835
#> 
#> 

# create a cellchat object
cellchat <- createCellChat(object = data.input, meta = meta, group.by = "labels",
                           datatype = "spatial", coordinates = spatial.locs, spatial.factors = spatial.factors)
cellchat
#> CellChat analysis of spatial data! The input spatial locations are 

# Set the  ligand-receptor interaction database
CellChatDB <- CellChatDB.mouse # use CellChatDB.human if running on human data
showDatabaseCategory(CellChatDB)
# Show the structure of the database
dplyr::glimpse(CellChatDB$interaction)

# use a subset of CellChatDB for cell-cell communication analysis
CellChatDB.use <- subsetDB(CellChatDB, search = "Secreted Signaling", key = "annotation") # use Secreted Signaling

# set the cell cell communication
# set the used database in the object
cellchat@DB <- CellChatDB.use


# Preprocessing the expression data for cell-cell communication analysis
# subset the expression data of signaling genes for saving computation cost
cellchat <- subsetData(cellchat) # This step is necessary even if using the whole database
future::plan("multisession", workers = 4) 
cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat, variable.both = F)
#> The number of highly variable ligand-receptor pairs used for signaling inference is 648
 
# project gene expression data onto PPI (Optional: when running it, USER should set `raw.use = FALSE` in the function `computeCommunProb()` in order to use the projected data)
# cellchat <- projectData(cellchat, PPI.mouse)
ptm = Sys.time()
execution.time = Sys.time() - ptm
print(as.numeric(execution.time, units = "secs"))
#> [1] 18.24813

# Part II: Inference of cell-cell communication network


# Compute the communication probability and infer cellular communication network
## we set up with a high memory

options(future.globals.maxSize = 89128960000000) 

cellchat <- computeCommunProb(cellchat, type = "truncatedMean", trim = 0.1,
                              distance.use = TRUE, interaction.range = 250, scale.distance = 0.01,
                              contact.dependent = TRUE, contact.range = 100)
#> truncatedMean is used for calculating the average gene expression per cell group. 
#> [1] ">>> Run CellChat on spatial transcriptomics data using distances as constraints of the computed communication probability <<< [2024-02-22 10:39:52.098053]"
#> Molecules of the input L-R pairs are diffusible. Run CellChat in a diffusion manner based on the `interaction.range`.
#> [1] ">>> CellChat inference is done. Parameter values are stored in `object@options$parameter` <<< [2024-02-22 10:54:30.589965]"


cellchat <- filterCommunication(cellchat, min.cells = 2)
cellchat

# Infer the cell-cell communciation at a signaling pathway
cellchat <- computeCommunProbPathway(cellchat)

# Calculate the aggregated cell-cell communication network
cellchat <- aggregateNet(cellchat)
# execution.time = Sys.time() - ptm
# print(as.numeric(execution.time, units = "secs"))
# ptm = Sys.time()

groupSize <- as.numeric(table(cellchat@idents))
par(mfrow = c(1,2), xpd=TRUE)
netVisual_circle(cellchat@net$count, vertex.weight = rowSums(cellchat@net$count), weight.scale = T, label.edge= F, title.name = "Number of interactions")
pdf(file="Spatial_D7_Overall_netVisual_circle_strength.pdf",height = 5, width = 4)
netVisual_circle(cellchat@net$weight, vertex.weight = rowSums(cellchat@net$weight), weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")
dev.off()
#library(ComplexHeatmap)
pdf(file="Spatial_D7_Overall_NetVisual_interaction_counts.pdf",height = 5, width = 4)
netVisual_heatmap(cellchat, measure = "count", color.heatmap = "Blues")
dev.off()
#> Do heatmap based on a single object
CellChat::netVisual_heatmap(cellchat, measure = "weight", color.heatmap = "Blues")
# pathway
AllPotentialPathways<-cellchat@netP$pathways
AllPotentialPathways
pathways.show <- c("TNF")
par(mfrow=c(1,1), xpd = TRUE) # `xpd = TRUE` should be added to show the title
# save t
pdf(file="Spatial_D7_Overall_TNFSignalingNetVisual.pdf",height = 7, width = 5)
#netVisual_aggregate(cellchat, signaling = pathways.show, layout = "circle")
#dev.off()
# Compute the network centrality scores
cellchat <- netAnalysis_computeCentrality(cellchat, slot.name = "netP") # the slot 'netP' means the inferred intercellular communication network of signaling pathways
# Visualize the computed centrality scores using heatmap, allowing ready identification of major signaling roles of cell groups
par(mfrow=c(1,1))
netAnalysis_signalingRole_network(cellchat, signaling = pathways.show, width = 8, height = 2.5, font.size = 10)
dev.off()
CandidatePathways<- c("CCL","CXCL","TNF")
#CandidatePathways <-cellchat@netP$pathways

pdf(file="Spatial_D7_Overall_InnateImmuneSignalingNetVisual.pdf",height = 6, width = 4)
Test<-netAnalysis_contribution(cellchat, signaling = CandidatePathways, sources.use= c("N14"),return.data=T)

dev.off()

# generate the contribute of netanalyses for each cluster
RequestedClusters<- levels(Idents(Pyelonephritis_InfectionDay28))
RequestedClusters
RequestLigendReceptors<-list()
for (i in RequestedClusters){
  Test<-netAnalysis_contribution(cellchat, signaling = CandidatePathways, sources.use= i,return.data=T)
  RequestLigendReceptors[[i]] <- Test$LR.contribution
  
}

RequestLigendReceptors
FinalRequestedLigandReceptorPairs<-as.data.frame(do.call(rbind, lapply(RequestLigendReceptors, `length<-`, max(lengths(RequestLigendReceptors)))))
write.csv(FinalRequestedLigandReceptorPairs, file = "FinalRequestedLigandReceptorPairs.csv")

# split by the .
sapply(strsplit(rownames(FinalRequestedLigandReceptorPairs), "\\."), `[`, 1)


# generate the potential clusters contributions

ClsContributionPair<-FinalRequestedLigandReceptorPairs %>% mutate(ClusterName=sapply(strsplit(rownames(FinalRequestedLigandReceptorPairs), "\\."), `[`, 1) ) %>% ggplot(aes(x=name, y= contribution, fill =ClusterName)) +
  # Implement a grouped bar chart
  geom_bar(position = "dodge", stat = "identity") + theme_classic()+RotatedAxis() 
ClsContributionPair
# save the pairs
ggsave("Spatial_D7_Overall_BarplotClsContributionPair.pdf", plot = ClsContributionPair, height = 5, width = 10)

# we only select the N12, N13, N14, N15, N16, N17
LimitedClsContributionPair<-FinalRequestedLigandReceptorPairs %>% mutate(ClusterName=sapply(strsplit(rownames(FinalRequestedLigandReceptorPairs), "\\."), `[`, 1) ) %>% filter(ClusterName %in% c("N12","N13","N14","N15","N16","N17")) %>% ggplot(aes(x=name, y= contribution, fill =ClusterName)) +
  # Implement a grouped bar chart
  geom_bar(position = "dodge", stat = "identity") + theme_classic()+RotatedAxis() 
ClsContributionPair
# save the pairs
ggsave("Spatial_D7_Overall_BarplotClsContributionPair_Infected.pdf", plot = LimitedClsContributionPair, height = 5, width = 10)

LimitedClsContributionPairHeatmap<-FinalRequestedLigandReceptorPairs %>% mutate(ClusterName=sapply(strsplit(rownames(FinalRequestedLigandReceptorPairs), "\\."), `[`, 1) )%>% 
filter(ClusterName %in% c("N12","N13","N14","N15","N16","N17"))%>%
  ggplot(aes(x=name, y= ClusterName, fill =contribution))+scale_fill_distiller(palette = "Spectral") +geom_tile() +RotatedAxis() 
LimitedClsContributionPairHeatmap
ggsave("Spatial_D7_Overall_BarplotClsContributionPair_Infected_heatmap.pdf", plot = LimitedClsContributionPairHeatmap, height = 4, width = 8)

#saveRDS(cellchat, file = "cellchat_Spatial_D7_Overall.rds")

# check the pair of the interactions

# extract the enriched pairs
pairLR.Interested<-extractEnrichedLR(cellchat, signaling = CandidatePathways, geneLR.return = FALSE)
pairLR.Interested
netVisual_individual(cellchat, signaling = pathways.show,  pairLR.use = pairLR.Interested, vertex.receiver = vertex.receive)

# read the cell chat object
cellchat@net$prob
#cellchat <- readRDS("cellchat_Spatial_D7_Overall.rds")
pdf(file="Spatial_D7_Overall_InnateImmuneSignalingnetVisual_bubble.pdf",height = 8, width = 12)
netVisual_bubble(cellchat,  signaling = CandidatePathways, remove.isolate = FALSE)
dev.off()
cellchat@net
# Only for N11,N12, N13, N14, N15,N16,N17
pdf(file="Spatial_D7_Targets_Overall_InnateImmuneSignalingnetVisual_bubble.pdf",height = 8, width = 12)
netVisual_bubble(cellchat,  sources.use = c("N12","N13","N14","N15","N16","N17"), targets.use = c("N12","N13","N14","N15","N16","N17"), signaling = CandidatePathways, remove.isolate = T,sort.by.source = T,sort.by.target = T)
dev.off()
?netVisual_bubble
# For the TNF signaling
pdf(file="Spatial_D7_Overall_TnfspatialFeature.pdf",height = 4.5, width = 4)
spatialFeaturePlot(cellchat, features = c("Tnf","Tnfrsf1a"), point.size = 1.2, color.heatmap = "Reds", direction = 1)
dev.off()

pdf(file="Spatial_D7_Overall_Il1bspatialFeature.pdf",height = 4.5, width = 4)
spatialFeaturePlot(cellchat, features = c("Il1b","Il1r1"), point.size = 1.2, color.heatmap = "Reds", direction = 1)
dev.off()

# potential useful 
cellchat@meta%>% head()
library(NMF)
library(ggalluvial)
netAnalysis_signalingRole_scatter(cellchat)
netAnalysis_signalingRole_scatter(cellchat, signaling = CandidatePathways)
 ht1<-netAnalysis_signalingRole_heatmap(cellchat, pattern = "outgoing", signaling = CandidatePathways)
 ht1
ht2<- netAnalysis_signalingRole_heatmap(cellchat, pattern = "incoming", signaling = CandidatePathways)
pdf(file="Spatial_D7_Overall_outgoincome.pdf",height = 5, width = 12)
ht1+ht2
dev.off()

# we plot the CXCL for vlnPlot
pdf(file="Spatial_D7_Overall_outgoincome_Violn_CCL.pdf",height = 15, width = 9)

plotGeneExpression(cellchat, signaling = c("CCL"), enriched.only = TRUE, type = "violin")
dev.off()
pdf(file="Spatial_D7_Overall_outgoincome_Violn_TNF.pdf",height = 6, width = 9)

plotGeneExpression(cellchat, signaling = c("TNF"), enriched.only = TRUE, type = "violin")
dev.off()

pdf(file="Spatial_D7_Overall_outgoincome_Violn_CXCL.pdf",height = 9, width = 9)

plotGeneExpression(cellchat, signaling = c("CXCL"), enriched.only = TRUE, type = "violin")
dev.off()

# generate the spatial features for each of pairs
# select all the requested pairs
RequestPairs 

# check the expression pair in the spatial datasets
spatialFeaturePlot(cellchat, pairLR.use = "CCL5_CCR5", point.size = 1.2, do.binary = TRUE, cutoff = 0.05, enriched.only = F, color.heatmap = "Reds", direction = 1)

spatialFeaturePlot(cellchat, pairLR.use = "CCL5_CCR5", point.size = 1.2, do.binary = F, cutoff = 0.05, enriched.only = F, color.heatmap = "Reds", direction = 1)


InterespatialFeatureDir <- paste0(CNichComdir,"InterespatialFeature/")
dir.create(InterespatialFeatureDir)

for (j in pairLR.Interested$interaction_name){
  SpatialFeature<- spatialFeaturePlot(cellchat, pairLR.use = j, point.size = 1.2, do.binary = F, cutoff = 0.05, enriched.only = F, color.heatmap = "Reds", direction = 1)
  
  ggsave(paste0(InterespatialFeatureDir,j,"spatialFeaturePlots.pdf"), plot=SpatialFeature, height = 3, width = 6 )
  
}

```



# ############################## 
# Generate the biomarkers within timepoints
# ############################## 
```{r}
CBiomadir
setwd(CBiomadir)
# generate the Dotplot for the biomarkers
DefaultAssay(Pyelonephritis_InfectionDay28) <-"SCT"
Finalcelltypemarkers<- c("Lrp2","Slc34a1","Slc13a3", # Proximal tubule cells (PT)
                      "Slc12a3","Pvalb","Wnk1",# Distal connecting tublue (DCT)

                     "Umod","Slc12a1","Cldn10", # Thick ascending limb cells (TAL)
                      "Klk1","Slc8a1","Calb1", # Connecting tubule (CNT)
                     "Slc14a2","Fst","Bst1", # descending thin  limb cells (DTL)
                     #"Akr1b1","Sh3gl3","Prox1", # ascending thin limb cells (ATL)
                     #"Enox1","Thsd4", #Macula densa
                      "Aqp2","Hsd11b2","Scnn1g", # principal cells (PC)
                    "Atp6v1g3","Aqp6","Slc26a7", # Intercalated cells
                    "Krt5","Upk1b","Krt14", # Urothelium
                    "Cdh5","Igfbp3","Pecam1", # Endothelium
                    "Col3a1","Fbn1","Lum","Col1a1", # Fibroblast
                    "Ireb2","Alas2", # Plasmacytoid dentritic cells
                    "S100a8","S100a9","Il1b", # Neutrophil
                    "C1qa","C1qb","Aif1", # Macrophage
                    "Cxcr6","Cd247","Cd3e", # T cells
                    "Igkc","Cd79a","Cd79b", # B cells
                    "Ccl5","Nkg7","Cd7", # NK/Cd8 cells
                    "Cx3cr1", "Ccr2", "H2-Ab1", # reactivate 
                 "Nos2","Ifngr1","Il17a","Il1b","Cd80","Cd86","Fcgr1","Fcgr3","Fcgr2b", #Macrophage 1
                 "Arg1","Il13","Cd163","Mrc1", "Cd68" # Macrophage 2
)


 for (i in Finalcelltypemarkers){
   CellTypeMarkerSpatialFeature<- SpatialFeaturePlot(Pyelonephritis_InfectionDay28, features = i, pt.size.factor = 6000, crop = TRUE)
   #& ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(0,10))
  
  ggsave(filename = paste0(i,"_SpatialFeatureAllMarkersDotPlot_Splits.pdf"), plot = CellTypeMarkerSpatialFeature, height = 5, width = 25)
  
}

```


# ############################## 
# DEGs and GO enrichment
# ############################## 

# Biomarker detection from these cluster of PT
```{r}
setwd(CDEGdir)

DefaultAssay(Pyelonephritis_InfectionDay28) <- "SCT"


? FindAllMarkers
Pyelonephritis_InfectionDay28 <- ScaleData(object = Pyelonephritis_InfectionDay28, features = rownames(Pyelonephritis_InfectionDay28))
# MAST was used to calculate the cluster
CorMed_nichesBiomarkers <-
  FindAllMarkers(
    object = Pyelonephritis_InfectionDay28,
    logfc.threshold = 0.25,
    test.use = "MAST",
    min.pct = 0.25
  ) # only markers that are in 25% of cells
# save all biomarkers as a dataframe
write.csv(CorMed_nichesBiomarkers, "CorMed_nichesBiomarkers_Mast.csv")

CorMed_nichesBiomarkers
```

# Draw the dotplot, heatmap and featureplot for the biomarkers in each cluster
```{r}
# select the top 3 genes in each cluster and perform the heatmap
CorMed_nichesBiomarkersTop5 <-
  as.data.frame(CorMed_nichesBiomarkers) %>% filter (avg_log2FC > 0.25) %>% group_by(cluster) %>% slice_max(order_by = avg_log2FC, n = 5) %>% pull (gene)

CorMed_nichesBiomarkersTop5 %>% unique()

CorMed_nichesBiomarkersTop5Heatmap<- DoHeatmap(Pyelonephritis_InfectionDay28, features = CorMed_nichesBiomarkersTop5 , slot = "scale.data") + scale_fill_gradientn(colors = c("blue", "white", "red")) + theme(axis.text.y = element_text(size = 10))
ggsave(
  "CorMed_nichesBiomarkersTop5_heatmap.pdf",
  plot = CorMed_nichesBiomarkersTop5Heatmap,
  height = 5,
  width = 5
)


CorMedCorMedPTSubBiomarkersTop5DotPlot <-
  DotPlot(Pyelonephritis_InfectionDay28, features = CorMed_nichesBiomarkersTop5 %>% unique()) +
  RotatedAxis()

ggsave(
  "CorMedCorMedPTSubBiomarkersTop5DotPlot.pdf",
  plot = CorMedCorMedPTSubBiomarkersTop5DotPlot,
  height = 4,
  width = 8
)

# Select the top 5 biomarkers
Top5PTMarker <-
  as.data.frame(CorMed_nichesBiomarkers) %>% 
  filter (avg_log2FC > 0.25) %>% group_by(cluster) %>% 
  slice_max(order_by = avg_log2FC, n = 5) %>% pull (gene) %>% unique()
? FeaturePlot

for (i in Top5PTMarker) {
  
  PTmarkerSpatialFeature <-
    SpatialFeaturePlot(Pyelonephritis_InfectionDay28, features = i, slot = "scale.data",pt.size.factor = 4)
  
  ggsave(
    paste0(i, "_PTSubBiomarkerFeatureSplit.pdf"),
    plot = PTmarkerSpatialFeature,
    height = 4,
    width = 4.5
  )
  
}

```


```{r}
setwd(CGOEndir)
# Generate a table for figure
CorMed_nichesBiomarkers %>% head()
CorMed_infectionSpecifcUpgeneFreqFigure <-
  table(CorMed_nichesBiomarkers$cluster) %>% as.data.frame() %>%
  ggplot(aes(x = Var1, y = Freq)) + geom_bar(stat = "identity") + theme_classic() +
  xlab("CorMed Niches") + ylab("Differentially expressed gene number")

CorMed_infectionSpecifcUpgeneFreqFigure
ggsave(
  "CorMed_nichesBiomarkersnumbersFigure.pdf",
  plot = CorMed_infectionSpecifcUpgeneFreqFigure,
  height = 3,
  width = 3.5
)

# Create a GO comparison for the figure
CorMed_infectionSpecifcUpgene <- list()

Idents(Pyelonephritis_InfectionDay28)
for (n in Idents(Pyelonephritis_InfectionDay28)) {
  # put he DEG into the up-regulated gene for comparision
  CorMed_infectionSpecifcUpgene[[n]] <-
    CorMed_nichesBiomarkers %>% filter(cluster %in% n &
                                        avg_log2FC >= 0.25) %>% pull(gene)
  #}
}
CorMed_infectionSpecifcUpgene$N8%>% head()
CorMed_infectionSpecifcUpgene$N10%>% head()
CorMed_infectionSpecifcUpgene$N11%>% head()
CorMed_infectionSpecifcUpgene$N12%>% head()
CorMed_infectionSpecifcUpgene$N15%>% head()

CorMed_infectionSpecifcUpgeneReorder<- CorMed_infectionSpecifcUpgene[c("N8","N10","N11","N12","N15")]
# change the name order: 
#names(CorMed_infectionSpecifcUpgene) <- c("N8","N10","N11","N12","N15")

CorMed_infectionSpecifcUpgeneReorder$N8%>% head()
CorMed_infectionSpecifcUpgeneReorder$N10%>% head()
CorMed_infectionSpecifcUpgeneReorder$N11%>% head()
CorMed_infectionSpecifcUpgeneReorder$N12%>% head()
CorMed_infectionSpecifcUpgeneReorder$N15%>% head()

# Comparing the Cluster
CorMed_infectionSpecifcUpgeneGOenrich <-
  clusterProfiler::compareCluster(
    geneClusters = CorMed_infectionSpecifcUpgeneReorder,
    fun = "enrichGO",
    OrgDb = "org.Mm.eg.db",
    ont = "BP",
    keyType = 'SYMBOL'
  )

write.csv(CorMed_infectionSpecifcUpgeneGOenrich,
          "CorMed_infectionSpecifcUpgeneGOenrich.csv")

#CorMed_infectionSpecifcUpgeneGOenrich
#CorMed_infectionSpecifcUpgeneGOenrich@compareClusterResult %>% head()

# Here we only concentrate on selected features
selectedPathway<- c("regulation of innate immune response","regulation of adaptive immune response","response to lipopolysaccharide","defense response to bacterium","neutrophil activation", "pattern recognition receptor signaling pathway","phagocytosis", "I-kappaB kinase/NF-kappaB signaling","receptor signaling pathway via STAT","tumor necrosis factor production","T cell differentiation","extracellular matrix organization", "epithelium migration","keratinocyte differentiation","Wnt signaling pathway","regulation of apoptotic signaling pathway")


CorMed_infectionSpecifcUpgeneGOenrichDotPlot <-
  dotplot(
    CorMed_infectionSpecifcUpgeneGOenrich,
    font.size = 12,
    label_format = 80,
    showCategory = selectedPathway
  )
? dotplot
#low = "LightBlue", midpoint=0, mid="white", high = "FireBrick",
ggsave(
  "CorMed_infectionSpecifcUpgeneGOenrichDotPlot.pdf",
  plot = CorMed_infectionSpecifcUpgeneGOenrichDotPlot,
  height = 7,
  width = 7.5
)

```

# ############################## 
# signaling pathway
# ############################## 
```{r}
setwd(CProgenydir)
## we set up with a high memory

options(future.globals.maxSize = 89128960000000) 
## We compute the Progeny activity scores and add them to our Seurat object
## as a new assay called Progeny. 
Idents(Pyelonephritis_InfectionDay28)
# create cluster data frame with the cells, Celltypes
CorMedCluster<- data.frame(Cell = names(Idents(Pyelonephritis_InfectionDay28)),   
                              CellType = as.character(Idents(Pyelonephritis_InfectionDay28)),
                              stringsAsFactors = FALSE)

?progeny
Pyelonephritis_InfectionDay28 <- progeny(Pyelonephritis_InfectionDay28, scale=FALSE, organism="Mouse", top=500, perm=1, return_assay = TRUE,assay_name = "SCT")
Pyelonephritis_InfectionDay28
# setup the default assay as progeny
DefaultAssay(object = Pyelonephritis_InfectionDay28) <- "progeny"

Pyelonephritis_InfectionDay28 <- Seurat::ScaleData(Pyelonephritis_InfectionDay28, assay = "progeny") 
## We can now directly apply Seurat functions in our Progeny scores. 
## For instance, we scale the pathway activity scores. 
progeny_scores_df <- 
  as.data.frame(t(GetAssayData(Pyelonephritis_InfectionDay28, slot = "scale.data", 
                               assay = "progeny"))) %>%
  rownames_to_column("Cell") %>%
  gather(Pathway, Activity, -Cell) 

progeny_scores_df
#progeny_scores_df %>% filter(Cell =="0DPI_AAAAAAAACCACCAGA")
## We match Progeny scores with the cell clusters.
progeny_scores_df <- inner_join(progeny_scores_df, CorMedCluster)

progeny_scores_df %>% head()
## We summarize the Progeny scores by cellpopulation
summarized_progeny_scores <- progeny_scores_df %>% 
  group_by(Pathway, CellType) %>%
  summarise(avg = mean(Activity), std = sd(Activity))
summarized_progeny_scores


## We prepare the data for the plot
summarized_progeny_scores_df <- summarized_progeny_scores %>%
  dplyr::select(-std) %>%   
  spread(Pathway, avg) %>%
  data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 

paletteLength = 80
myColor = colorRampPalette(c("royalblue", "white","lightcoral"))(paletteLength)

progenyBreaks = c(seq(min(summarized_progeny_scores_df), 0, 
                      length.out=ceiling(paletteLength/2) + 1),
                  seq(max(summarized_progeny_scores_df)/paletteLength, 
                      max(summarized_progeny_scores_df), 
                      length.out=floor(paletteLength/2)))
CorMedForHeatmap<-as.data.frame(t(summarized_progeny_scores_df[,-1]))

# generate the heatmap for PT
pdf(file="CorMedForHeatmap_Progeny_ForHeatmap.pdf",height = 3, width = 3.5)
progeny_hmap = pheatmap(CorMedForHeatmap,fontsize=8, 
                        fontsize_row = 8,  show_rownames=T, cluster_cols=T, cluster_rows=T, scale="row",
                        cex=1, clustering_distance_rows="euclidean", cex=1,
                        # clustering_distance_cols="euclidean", clustering_method="complete", 
                        color=myColor, breaks = progenyBreaks, 
                        main = "PROGENy (500)", angle_col = 45,
                        border_color = NA,
                        treeheight_col = 0,   treeheight_row=0)

progeny_hmap
dev.off()
test<- FeaturePlot(Pyelonephritis_InfectionDay28, features = "NFkB")
ggsave("test_NFKB.pdf", plot = test, width = 5, height = 5)
# generate the featureplot 
ProgenyFeatures <- summarized_progeny_scores$Pathway %>% unique()
ProgenyFeatures
for (i in ProgenyFeatures ){
  p2 <-(FeaturePlot(Pyelonephritis_InfectionDay28, features = i) & 
  scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red'))& 
  ggtitle(paste0(i," activity"))
  
  ggsave(paste0(i, "_CorMedInfected_cluster_featurePlot.pdf"), plot = p2, width = 6, height = 6)
  
  p3 <- (SpatialFeaturePlot(Pyelonephritis_InfectionDay28, features = i, pt.size.factor = 4) +ggtitle(paste0(i," activity")))
  ggsave(paste0(i, "_CorMedInfected_cluster_SpatialfeaturePlot.pdf"), plot = p3, width = 6, height = 6)
  
}

```
# ############################## 
# Transcript factor
# ############################## 
```{r}

# Transcription factor activity inference from Immune scRNA-seq

# CollecTRI network
# set up the directory
setwd(CTFdir)
netTF <-  readRDS("/Users/XXW004/Documents/Projects/RuizRosado/PIPseq/Datasets/get_collectri_human_mouse20230716/get_collectri_mouse_process_table.rds")

Pyelonephritis_InfectionDay28@assays$SCT@data %>% head()
mat <- as.matrix(Pyelonephritis_InfectionDay28@assays$SCT@data)
# Run ulm
acts <- run_ulm(mat=mat, net=netTF, .source='source', .target='target',
                .mor='mor', minsize = 5)


saveRDS(acts, file="CorMed_Transcriptfactor.RDS")


#acts <- readRDS("PT_Transcriptfactor.RDS")
# Activity inference with Univariate Linear Model (ULM)
# Extract ulm and store it in tfsulm in pbmc
Pyelonephritis_InfectionDay28[['tfsulm']] <- acts %>%
  pivot_wider(id_cols = 'source', names_from = 'condition',
              values_from = 'score') %>%
  column_to_rownames('source') %>%
  Seurat::CreateAssayObject(.)

# Change assay
DefaultAssay(object = Pyelonephritis_InfectionDay28) <- "tfsulm"

# Scale the data
Pyelonephritis_InfectionDay28 <- ScaleData(Pyelonephritis_InfectionDay28)
Pyelonephritis_InfectionDay28@assays$tfsulm@data <- Pyelonephritis_InfectionDay28@assays$tfsulm@scale.data

# Explore the top 20 more varaible TFs
n_tfs <- 20

t(as.matrix(Pyelonephritis_InfectionDay28@assays$tfsulm@data))
# Extract activities from object as a long dataframe
df <- t(as.matrix(Pyelonephritis_InfectionDay28@assays$tfsulm@data)) %>%
  as.data.frame() %>%
  mutate(cluster = Idents(Pyelonephritis_InfectionDay28)) %>%
  pivot_longer(cols = -cluster, names_to = "source", values_to = "score") %>%
  group_by(cluster, source) %>%
  summarise(mean = mean(score))

df
df %>% filter(grepl("Egfr",source))
# Get top tfs with more variable means across clusters
tfs <- df %>%
  group_by(source) %>%
  summarise(std = sd(mean)) %>%
  arrange(-abs(std)) %>%
  head(n_tfs) %>%
  pull(source)

# drop "Znf335" due to not exist in the expression
#tfs2 <- tfs[!tfs == "Tnf"]

#tfs[!tfs == "Znf335"]
# Subset long data frame to top tfs and transform to wide matrix
top_acts_mat <- df %>%
  filter(source %in% tfs) %>%
  pivot_wider(id_cols = 'cluster', names_from = 'source',
              values_from = 'mean') %>%
  column_to_rownames('cluster') %>%
  as.matrix()

# Choose color palette
palette_length = 100
my_color = colorRampPalette(c("Darkblue", "white","red"))(palette_length)

my_breaks <- c(seq(-3, 0, length.out=ceiling(palette_length/2) + 1),
               seq(0.05, 3, length.out=floor(palette_length/2)))

# Plot
PTTFheatmap<- pheatmap(t(top_acts_mat),border_color = NA, cluster_cols = F, 
                        color=my_color,fontsize=12, fontsize_row = 10, 
                        breaks = my_breaks, 
                        #main = "Predicted TF activities", 
                        #angle_col = 90,
                        treeheight_col = 0, treeheight_row=0)


ggsave("CorMedInfectionHeatmap_Transcriptfactor_Progeny_Decouple.pdf", plot = PTTFheatmap, width = 3, height = 4)

# draw the transcriptor factor activity and their expression
# generate the figure of transcript factor activity:
for (m in tfs){
  p2 <- (SpatialFeaturePlot(Pyelonephritis_InfectionDay28, features = m,pt.size.factor = 4) +
  #scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red')) +
  ggtitle(paste0(m, ' activity')))
  
  ggsave(paste0("TFActivity_",m,"_CorMedInfection_SpatialFeaturePlot.pdf"), plot = p2, width = 6, height = 6)
}

Pyelonephritis_InfectionDay28
DefaultAssay(object = Pyelonephritis_InfectionDay28) <- "SCT"

for (n in tfs){
  p3 <- (FeaturePlot(Pyelonephritis_InfectionDay28, features = n,slot= "scale.data") + 
  #scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red')) +
  ggtitle(paste0(n, ' expression')))
  
  p4 <- (SpatialFeaturePlot(Pyelonephritis_InfectionDay28, features = n, pt.size.factor = 4, slot= "scale.data") +
 # scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red')) +
  ggtitle(paste0(n, ' expression (time)')))
 
  ggsave(paste0("TFExpression_",n,"_Pyelonephritis_InfectionDay28_FeaturePlot.pdf"), plot = p3, width = 6, height = 6)
  ggsave(paste0("TFExpression_",n,"_Pyelonephritis_InfectionDay28_SpatilFeaturePlot.pdf"), plot = p4, width = 6, height = 6)

}


# add their activity in Dotplot
Idents(Pyelonephritis_InfectionDay28)
# get the order of row name
GeneOrderByPheatmap <- rownames(t(top_acts_mat))[PTTFheatmap$tree_row$order]
#IMMTFheatmap$tree_row$order
GeneOrderByPheatmap
?DotPlot
TFDotPlot <- DotPlot(Pyelonephritis_InfectionDay28, features = rev(GeneOrderByPheatmap)) +
  RotatedAxis() + coord_flip()+ 
  #ggtitle("Predicted TF expression")+
   theme(legend.position = "right")
ggsave("CorMed_TFDotPlot.pdf", plot = TFDotPlot, width = 5, height = 6)

```



# Please ignore the following datasets


# ############################## 
# niche-niche communications
# ############################## 

# refernece: https://github.com/saeyslab/nichenetr/blob/master/vignettes/differential_nichenet.md
# Run the nichenet_seruratobj_aggregate for the different regions
```{r}
DefaultAssay(Pyelonephritis_InfectionDay28) <- "SCT"
Pyelonephritis_InfectionDay28@meta.data %>% head()
#Idents(Pyelonephritis_InfectionDay28) <- Pyelonephritis_InfectionDay28@meta.data$Sub2

# reset the identity, Combine the PT cells

Idents(Pyelonephritis_InfectionDay28)
Pyelonephritis_InfectionDay28@meta.data
# adjust the levels
# Idents(Pyelonephritis_InfectionDay28) <-
# ordered(factor(Idents(Pyelonephritis_InfectionDay28)), levels=c( "5_1","10_1","10_0","2_1","2_0","5_0","5_2","4","7","0","1","9","6","3","8"))

levels(Idents(Pyelonephritis_InfectionDay28))
```

# use the wrap 
```{r}
# sender
sender_cluster <- c("N8","N10","N11","N12","N15")

# define receive cluster

receiver_cluster <- c("N15")
Pyelonephritis_InfectionDay28@meta.data %>% head()
```


```{r}

ligand_target_matrix <-
    readRDS(paste0(data_path, "ligand_target_matrix_nsga2r_final_mouse.rds"))

lr_network <-
    readRDS(paste0(data_path, "lr_network_mouse_21122021.rds"))

ligand_target_matrix %>% head()
lr_network = lr_network %>% dplyr::distinct(from, to)

weighted_networks <-
    readRDS(paste0(data_path, "weighted_networks_nsga2r_final_mouse.rds"))

weighted_networks_lr <-
    weighted_networks$lr_sig %>%
    inner_join(lr_network %>% distinct(from, to),
               by = c("from", "to"))


Pyelonephritis_InfectionDay28 = alias_to_symbol_seurat(Pyelonephritis_InfectionDay28, "mouse")

Pyelonephritis_InfectionDay28@meta.data %>% head()
Pyelonephritis_InfectionDay28@meta.data$IntegrateAnnot %>% table() 

head(weighted_networks$gr) 
DimPlot(Pyelonephritis_InfectionDay28)

# If you use convert_to_alias before here, this one won't work
Pyelonephritis_InfectionDay28 = Seurat::PrepSCTFindMarkers(Pyelonephritis_InfectionDay28, assay = "SCT", verbose = FALSE)
lr_network = lr_network %>% dplyr::rename(ligand = from, receptor = to) %>% distinct(ligand, receptor)

# define niches
niches <- list()
niches = list(
    "Abcess_niche" = list(
      "sender" = c("N12","N11","N10","N8"),
      "receiver" = c("N15")),
    "Margen_niche" = list(
      "sender" = c("N15","N11","N10","N8"),
      "receiver" = c("N12"))
  )


lr_network %>% head()
Pyelonephritis_InfectionDay28 %>% subset(features = lr_network$ligand %>% intersect(rownames(Pyelonephritis_InfectionDay28)))
# calcualte the niches difference 
DE_sender = nichenetr::calculate_niche_de(seurat_obj  = Pyelonephritis_InfectionDay28 %>% subset(features = lr_network$ligand %>% intersect(rownames(Pyelonephritis_InfectionDay28))), niches = niches, type = "sender", assay_oi = "SCT") # only ligands important for sender cell types
DE_receiver = nichenetr::calculate_niche_de(seurat_obj  = Pyelonephritis_InfectionDay28 %>% subset(features = lr_network$receptor %>% unique()), niches = niches, type = "receiver", assay_oi = "SCT") # only receptors now, later on: DE analysis to find targets


DE_sender = DE_sender %>% mutate(avg_log2FC = ifelse(avg_log2FC == Inf, max(avg_log2FC[is.finite(avg_log2FC)]), ifelse(avg_log2FC == -Inf, min(avg_log2FC[is.finite(avg_log2FC)]), avg_log2FC)))
DE_receiver = DE_receiver %>% mutate(avg_log2FC = ifelse(avg_log2FC == Inf, max(avg_log2FC[is.finite(avg_log2FC)]), ifelse(avg_log2FC == -Inf, min(avg_log2FC[is.finite(avg_log2FC)]), avg_log2FC)))

# Process DE results:
expression_pct = 0.10
DE_sender_processed = nichenetr::process_niche_de(DE_table = DE_sender, niches = niches, expression_pct = expression_pct, type = "sender")
DE_receiver_processed =nichenetr::process_niche_de(DE_table = DE_receiver, niches = niches, expression_pct = expression_pct, type = "receiver")
# Combine sender-receiver DE based on L-R pairs:
specificity_score_LR_pairs = "min_lfc"
DE_sender_receiver = nichenetr::combine_sender_receiver_de(DE_sender_processed, DE_receiver_processed, lr_network, specificity_score = specificity_score_LR_pairs)
DE_sender_receiver

# 4. Calculate ligand activities and infer active ligand-target links

lfc_cutoff = 0.15 # recommended for 10x as min_lfc cutoff. 
specificity_score_targets = "min_lfc"

DE_receiver_targets = nichenetr::calculate_niche_de_targets(seurat_obj = Pyelonephritis_InfectionDay28, niches = niches, lfc_cutoff = lfc_cutoff, expression_pct = expression_pct, assay_oi = "SCT") 

DE_receiver_processed_targets = nichenetr::process_receiver_target_de(DE_receiver_targets = DE_receiver_targets, niches = niches, expression_pct = expression_pct, specificity_score = specificity_score_targets)
  
background = DE_receiver_processed_targets  %>% pull(target) %>% unique()

geneset_Abcess = DE_receiver_processed_targets %>% filter(receiver == niches$Abcess_niche$receiver & target_score >= lfc_cutoff & target_significant == 1 & target_present == 1) %>% pull(target) %>% unique()
geneset_Margin = DE_receiver_processed_targets %>% filter(receiver == niches$Margen_niche$receiver & target_score >= lfc_cutoff & target_significant == 1 & target_present == 1) %>% pull(target) %>% unique()

# Good idea to check which genes will be left out of the ligand activity analysis (=when not present in the rownames of the ligand-target matrix).
# If many genes are left out, this might point to some issue in the gene naming (eg gene aliases and old gene symbols, bad human-mouse mapping)
geneset_Abcess %>% setdiff(rownames(ligand_target_matrix))
### [1] "Wfdc17"        "AW112010"      "2900097C17Rik" "B430306N03Rik" "AC149090.1"
geneset_Margin %>% setdiff(rownames(ligand_target_matrix))


length(geneset_Abcess)
## [1] 17
length(geneset_Margin)



nichenet_output = nichenet_seuratobj_aggregate(
  seurat_obj = Pyelonephritis_InfectionDay28, 
  receiver = "PT", 
  condition_colname = "DataSet", condition_oi = "7DPI", condition_reference = "0DPI", 
  sender = c("FIB","M1M2","RecMAC","NEU","ResIMM","T"), 
  ligand_target_matrix = ligand_target_matrix,
  lr_network = lr_network,
  weighted_networks = weighted_networks)

nichenet_output$ligand_target_heatmap
nichenet_output$ligand_receptor_heatmap

ggsave(filename = "PIPseq_Cell_Cell_Communication_v1011.pdf",nichenet_output$ligand_activity_target_heatmap, width = 28, height = 9)

ggsave(filename = "PIPseq_Cell_Cell_Communication_ligandtargetheatmap_v1011.pdf",nichenet_output$ligand_target_heatmap, width = 15, height = 9)
ggsave(filename = "PIPseq_Cell_Cell_Communication_ligandreceptor_heatmapp_v1011.pdf",nichenet_output$ligand_receptor_heatmap, width = 15, height = 9)


```

# draw the Circos plots
```{r}
# Calculate average ligand expression in sender cells
# avg_expression_ligands = AverageExpression(seuratObj %>% subset(subset = aggregate == "LCMV"),features = nichenet_output$top_ligands) # if want to look specifically in LCMV-only cells
avg_expression_ligands = AverageExpression(Pyelonephritis_InfectionDay28, features = nichenet_output$top_ligands)

# Assign ligands to sender cells
sender_ligand_assignment = avg_expression_ligands$RNA %>% apply(1, function(ligand_expression){
  ligand_expression > (ligand_expression %>% mean() + ligand_expression %>% sd())
  }) %>% t()
sender_ligand_assignment = sender_ligand_assignment %>% apply(2, function(x){x[x == TRUE]}) %>% purrr::keep(function(x){length(x) > 0})
names(sender_ligand_assignment)

# Determine now which prioritized ligands are expressed by CAFs and or endothelial cells

all_assigned_ligands = sender_ligand_assignment %>% lapply(function(x){names(x)}) %>% unlist()
all_assigned_ligands
unique_ligands = all_assigned_ligands %>% table() %>% .[. == 1] %>% names()

general_ligands = nichenet_output$top_ligands %>% setdiff(unique_ligands)


sender_ligand_assignment$`FIB` %>% names()%>% setdiff(general_ligands)


FIB_specific_ligands = sender_ligand_assignment$`FIB` %>% names() %>% setdiff(general_ligands)

FIB_specific_ligands

PT_specific_ligands = sender_ligand_assignment$`PT` %>% names() %>% setdiff(general_ligands)

PT_specific_ligands

RecMAC_specific_ligands = sender_ligand_assignment$`RecMAC` %>% names() %>% setdiff(general_ligands)

RecMAC_specific_ligands

NEU_specific_ligands = sender_ligand_assignment$`NEU` %>% names() %>% setdiff(general_ligands)

NEU_specific_ligands

M1M2_specific_ligands = sender_ligand_assignment$`M1M2` %>% names() %>% setdiff(general_ligands)

M1M2_specific_ligands


#IC_specific_ligands = sender_ligand_assignment$IC %>% names() %>% setdiff(general_ligands)
#IC_specific_ligands

#PC_specific_ligands = sender_ligand_assignment$PC %>% names() %>% setdiff(general_ligands)
#PC_specific_ligands
PT_specific_ligands
ligand_type_indication_df = tibble(
  ligand_type = c(rep("PT-specific", times = PT_specific_ligands %>% length()),
                  rep("NEU-specific", times = NEU_specific_ligands %>% length()),
                  rep("RecMAC-specific", times = RecMAC_specific_ligands %>% length()),
                  rep("M1M2-specific", times = M1M2_specific_ligands %>% length()),
                   rep("FIB-specific", times = FIB_specific_ligands %>% length()),
                  #rep("IC-specific", times = IC_specific_ligands %>% length()),
                  rep("General", times = general_ligands %>% length())),
  ligand = c(PT_specific_ligands, NEU_specific_ligands, RecMAC_specific_ligands, M1M2_specific_ligands, FIB_specific_ligands, general_ligands))

ligand_type_indication_df
```

```{r}
# Define the ligand-target links of interest
active_ligand_target_links_df = nichenet_output$ligand_target_df %>% mutate(target_type = "7DPI") %>% inner_join(ligand_type_indication_df) # if you want ot make circos plots for multiple gene sets, combine the different data frames and differentiate which target belongs to which gene set via the target type

active_ligand_target_links_df

cutoff_include_all_ligands = active_ligand_target_links_df$weight %>% quantile(0.40)

active_ligand_target_links_df_circos = active_ligand_target_links_df %>% filter(weight > cutoff_include_all_ligands)

ligands_to_remove = setdiff(active_ligand_target_links_df$ligand %>% unique(), active_ligand_target_links_df_circos$ligand %>% unique())
targets_to_remove = setdiff(active_ligand_target_links_df$target %>% unique(), active_ligand_target_links_df_circos$target %>% unique())
  
```

# Prepare the circos visualization: give each segment of ligands and targets a specific color and order
```{r}
circos_links = active_ligand_target_links_df %>% filter(!target %in% targets_to_remove &!ligand %in% ligands_to_remove)
circos_links

grid_col_ligand =c("General" = "steelblue2",
            "RecMAC-specific" = "red",
            "NEU-specific" = "violet",
            "M1M2-specific" = "darkgreen",
            "FIB-specific" = "lawngreen",
            "PT-specific" = "royalblue"
            )
grid_col_target =c(
            "7DPI" = "darkgray")

grid_col_tbl_ligand = tibble(ligand_type = grid_col_ligand %>% names(), color_ligand_type = grid_col_ligand)
grid_col_tbl_target = tibble(target_type = grid_col_target %>% names(), color_target_type = grid_col_target)

circos_links = circos_links %>% mutate(ligand = paste(ligand," ")) # extra space: make a difference between a gene as ligand and a gene as target!
circos_links = circos_links %>% inner_join(grid_col_tbl_ligand) %>% inner_join(grid_col_tbl_target)

circos_links

links_circle = circos_links %>% dplyr::select(ligand,target, weight)

ligand_color = circos_links %>% distinct(ligand,color_ligand_type)
grid_ligand_color = ligand_color$color_ligand_type %>% set_names(ligand_color$ligand)
target_color = circos_links %>% distinct(target,color_target_type)
grid_target_color = target_color$color_target_type %>% set_names(target_color$target)

grid_col =c(grid_ligand_color,grid_target_color)


# give the option that links in the circos plot will be transparant ~ ligand-target potential score
transparency = circos_links %>% mutate(weight =(weight-min(weight))/(max(weight)-min(weight))) %>% mutate(transparency = 1-weight) %>% .$transparency 

transparency

# prepare circos
target_order = circos_links$target %>% unique()

ligand_order = c(PT_specific_ligands, NEU_specific_ligands, RecMAC_specific_ligands, M1M2_specific_ligands, FIB_specific_ligands, general_ligands) %>% c(paste(.," ")) %>% intersect(circos_links$ligand)
order = c(ligand_order,target_order)
# Prepare the circos visualization: define the gaps between the different segments
width_same_cell_same_ligand_type = 0.5
width_different_cell = 6
width_ligand_target = 15
width_same_cell_same_target_type = 0.5

gaps = c(
  # width_ligand_target,
  rep(width_same_cell_same_ligand_type, times = (circos_links %>% filter(ligand_type == "FIB-specific") %>% distinct(ligand) %>% nrow() -1)),
  width_different_cell,
  rep(width_same_cell_same_ligand_type, times = (circos_links %>% filter(ligand_type == "NEU-specific") %>% distinct(ligand) %>% nrow() -1)),
  width_different_cell,
  rep(width_same_cell_same_ligand_type, times = (circos_links %>% filter(ligand_type == "RecMAC-specific") %>% distinct(ligand) %>% nrow() -1)),
  width_different_cell,
  rep(width_same_cell_same_ligand_type, times = (circos_links %>% filter(ligand_type == "M1M2-specific") %>% distinct(ligand) %>% nrow() -1)),
  width_different_cell,
    rep(width_same_cell_same_ligand_type, times = (circos_links %>% filter(ligand_type == "PT-specific") %>% distinct(ligand) %>% nrow() -1)),
  width_different_cell,
  rep(width_same_cell_same_ligand_type, times = (circos_links %>% filter(ligand_type == "General") %>% distinct(ligand) %>% nrow() -1)),
  width_ligand_target,
  rep(width_same_cell_same_target_type, times = (circos_links %>% filter(target_type == "7DPI") %>% distinct(target) %>% nrow() -1)),
  width_ligand_target
  )
```

```{r}
circos.par(gap.degree = gaps)
chordDiagram(links_circle, directional = 1,order=order,link.sort = TRUE, link.decreasing = FALSE, grid.col = grid_col,transparency = 0, diffHeight = 0.005, direction.type = c("diffHeight", "arrows"),link.arr.type = "big.arrow", link.visible = links_circle$weight >= cutoff_include_all_ligands,annotationTrack = "grid", 
    preAllocateTracks = list(track.height = 0.075))
# we go back to the first track and customize sector labels
circos.track(track.index = 1, panel.fun = function(x, y) {
    circos.text(CELL_META$xcenter, CELL_META$ylim[1], CELL_META$sector.index,
        facing = "clockwise", niceFacing = TRUE, adj = c(0, 0.55), cex = 1)
}, bg.border = NA) #

circos.clear()
```

```{r}
svg("ligand_receptor_circos.svg", width = 12, height = 12)
circos.par(gap.degree = gaps)
chordDiagram(links_circle, directional = 1,order=order,link.sort = TRUE, link.decreasing = FALSE, grid.col = grid_col,transparency = transparency, diffHeight = 0.005, direction.type = c("diffHeight", "arrows"),link.arr.type = "big.arrow", link.visible = links_circle$weight >= cutoff_include_all_ligands,annotationTrack = "grid",
    preAllocateTracks = list(track.height = 0.075))
# we go back to the first track and customize sector labels
circos.track(track.index = 1, panel.fun = function(x, y) {
    circos.text(CELL_META$xcenter, CELL_META$ylim[1], CELL_META$sector.index,
        facing = "clockwise", niceFacing = TRUE, adj = c(0, 0.55), cex = 0.8)
}, bg.border = NA) #
circos.clear()
dev.off()
```












# ################
# Generate the Day 0
# scatter pie plot for different cell types at day 0
```{r}
setwd(Contrdir)
Pyelonephritis_InfectionSubslidesD0 <-
    subset (Pyelonephritis_Infection, orig.ident == "PyeloD0_1")
Pyelonephritis_InfectionSubslidesD0
  # there seems to be an error in the susetting, so we removed other igmages
  #names(Pyelonephritis_InfectionSubslides@images)
SelectedImageName <- names(Pyelonephritis_InfectionSubslidesD0@images)[1]
SelectedImageName
OtherImages <- Allimagmes[Allimagmes != SelectedImageName]
OtherImages
Pyelonephritis_InfectionSubslidesD0@images[OtherImages] = NULL

#nrow(Pyelonephritis_InfectionDay28@meta.data )
SpatialDimPlot(Pyelonephritis_InfectionSubslidesD0, crop = T)

# the prediction of the object
Pyelonephritis_InfectionSubslidesD0[["predictions"]]
InfectedMat_D0<- t(Pyelonephritis_InfectionSubslidesD0[["predictions"]]@data)
colnames(InfectedMat_D0) <- ordered (colnames(InfectedMat_D0), levels =CelltypeOrder)
# change the order
InfectedMat_D0<- InfectedMat_D0[ ,CelltypeOrder]
# for final celltype matrix: 
InfectedMat_D0CellType<- InfectedMat_D0[,!colnames(InfectedMat_D0) %in% c("max")]
InfectedMat_D0CellTypeNozero <- InfectedMat_D0CellType[,colSums(InfectedMat_D0CellType)>0]


# correlation
InfectedMat_D0CellTypeCorrelationMatrix<-plotCorrelationMatrix(InfectedMat_D0CellTypeNozero)
InfectedMat_D0CellTypeInteractions<- plotInteractions(InfectedMat_D0CellTypeNozero, "heatmap") + scale_fill_gradientn(colors = c("blue", "white", "red"))

ggsave("SpatialScatterPieDay0_CorrelationMatrix.pdf", plot =InfectedMat_D0CellTypeCorrelationMatrix, height = 6, width = 6 )

ggsave("SpatialScatterPieDay0_HeatmapColocalization.pdf", plot =InfectedMat_D0CellTypeInteractions, height = 6, width = 6 )
# scatterpie plot
# scatterpie plot
# 
ct <- colnames(InfectedMat_D0CellType)
ct
# we did not add the low proportion in the scatterpie
InfectedMat_D0CellType[InfectedMat_D0CellType < 0.1] <- 0
paletteMartin <- c(
    "#b6dbff", # HeaPT-S1
    "#99CCFF", # HeaPT-S2
    "#0099FF", # HeaPT-S3
    "#6699FF", # HeaPT-S4
    "#006ddb", # HeaPT-S5
    "#66CC99", # DegPT
    "#EE4E34", # MalaPT
    "#F5DEB3", # infiPT
    "#BEE5BF", # HybridPT
    "#FFD700", # CNT
    "#bdbdbd", # DTL
    "#b66dff", #TAL
    "#FF9900", #DCT-TAL
    "#F96167", #IC
    "#CCF381", #PC
    "#ff6db6", #URO
    "#C5DECD", #ENO
    "#FFCC00", #FIB
    "#F4FEC1", #PL
    "#CCFF33", #ResIMM
    "#99FF66", # RecMAC
    "#A1E8CC", # M1-M2
    "#FF00FF", #NEU
    "#DFF3E3", # T
    "#E2D1F9", #B
    "#CC0033" #NK
  )
#DefaultAssay(Pyelonephritis_InfectionDay28) <- "Spatial"

# choose three color
# pal<-colorRampPalette(c("LightBlue","white","FireBrick"))(length(ct))
# choose the selected color
pal <- colorRampPalette(paletteMartin)(length(ct))

# choose the spectral range colors:
# https://earlglynn.github.io/RNotes/package/RColorBrewer/index.html
#pal <- (colorRampPalette(brewer.pal(11,"Paired"))(length(ct)))

names(pal) <- ct
#pal
rownames(Pyelonephritis_InfectionSubslidesD0)
colnames(Pyelonephritis_InfectionSubslidesD0)
str(Pyelonephritis_InfectionSubslidesD0)
#SPOTlight:::. extract_image(Pyelonephritis_Infection )
#rownames(Pyelonephritis_Infection)
SpatialScatterPieDay0 <-plotSpatialScatterpie(
  x= Pyelonephritis_InfectionSubslidesD0,
  y= InfectedMat_D0CellType,
  cell_types = colnames(InfectedMat_D0CellType),
    img = TRUE,
    scatterpie_alpha = 1,
    pie_scale = 0.4) +
    scale_fill_manual(
        values = pal,
        breaks = names(pal)) +coord_sf(xlim = c(200, 450), ylim = c(150, 450))

ggsave("SpatialScatterPieDay0.pdf",plot = SpatialScatterPieDay0, height = 4,width = 5)
# colnames(DeconvolutedMatrix)
# ncol(Pyelonephritis_Infection)
# nrow(DeconvolutedMatrix)
InfectedMat_D0CellType
pdf(file = "SpatialScatterPieDay0_CellTypeNetwork.pdf", height = 8,width = 8)
plotInteractions(InfectedMat_D0CellType, "network")
dev.off()

Idents(Pyelonephritis_InfectionSubslidesD0)

# generate the Dotplot for the biomarkers
DefaultAssay(Pyelonephritis_InfectionSubslidesD0) <-"SCT"
Finalcelltypemarkers<- c("Lrp2","Slc34a1","Slc13a3", # Proximal tubule cells (PT)
                      "Slc12a3","Pvalb","Wnk1",# Distal connecting tublue (DCT)

                     "Umod","Slc12a1","Cldn10", # Thick ascending limb cells (TAL)
                      "Klk1","Slc8a1","Calb1", # Connecting tubule (CNT)
                     "Slc14a2","Fst","Bst1", # descending thin  limb cells (DTL)
                     #"Akr1b1","Sh3gl3","Prox1", # ascending thin limb cells (ATL)
                     #"Enox1","Thsd4", #Macula densa
                      "Aqp2","Hsd11b2","Scnn1g", # principal cells (PC)
                    "Atp6v1g3","Aqp6","Slc26a7", # Intercalated cells
                    "Krt5","Upk1b","Krt14", # Urothelium
                    "Cdh5","Igfbp3","Pecam1", # Endothelium
                    "Col3a1","Fbn1","Lum","Col1a1", # Fibroblast
                    "Ireb2","Alas2","C78334", # Plasmacytoid dentritic cells
                    "S100a8","S100a9","Il1b", # Neutrophil
                    "C1qa","C1qb","Aif1", # Macrophage
                    "Cxcr6","Cd247","Cd3e", # T cells
                    "Igkc","Cd79a","Cd79b", # B cells
                    "Ccl5","Nkg7","Cd7" # NK/Cd8 cells
)

SpatialD0MarkersDotPlot<- DotPlot(Pyelonephritis_InfectionSubslidesD0, features = Finalcelltypemarkers,cols = "Spectral", dot.scale = 4) +RotatedAxis() +theme(legend.text = element_text(size = 12), legend.title = element_text(size = 8))

ggsave("SpatialD0MarkersDotPlot.pdf", plot = SpatialD0MarkersDotPlot, height = 6, width = 14)
Pyelonephritis_InfectionSubslidesD0

for (i in Finalcelltypemarkers){
   CellTypeMarkerSpatialFeature<- SpatialFeaturePlot(Pyelonephritis_InfectionSubslidesD0, features = i, pt.size.factor = 1.6, crop = TRUE) 
   #&
       #   ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(0,1))
  
  ggsave(filename = paste0(i,"_SpatialFeaturePrediction.pdf"), plot = CellTypeMarkerSpatialFeature, height = 5, width = 5)
  
}

```


# ################
# Generate the Day 7
# scatter pie plot for different cell types at day 7
```{r}
setwd(Day7dir)

Pyelonephritis_Infection$orig.ident
Pyelonephritis_InfectionDay28 <-
    subset (Pyelonephritis_Infection, orig.ident == "PyeloD7_1")
Pyelonephritis_InfectionDay28
  # there seems to be an error in the susetting, so we removed other igmages
  #names(Pyelonephritis_InfectionSubslides@images)
SelectedImageName <- names(Pyelonephritis_InfectionDay28@images)[5]
SelectedImageName
OtherImages <- Allimagmes[Allimagmes != SelectedImageName]
OtherImages
Pyelonephritis_InfectionDay28@images[OtherImages] = NULL

#nrow(Pyelonephritis_InfectionDay28@meta.data )
Pyelonephritis_InfectionDay28Dimplot <-SpatialDimPlot(Pyelonephritis_InfectionDay28, crop = T)

ggsave("Pyelonephritis_InfectionDay28Dimplot.pdf", plot =Pyelonephritis_InfectionDay28Dimplot, height = 6, width = 6 )

# the prediction of the object
Pyelonephritis_InfectionDay28[["predictions"]]
InfectedMat_D7<- t(Pyelonephritis_InfectionDay28[["predictions"]]@data)
colnames(InfectedMat_D7) <- ordered (colnames(InfectedMat_D7), levels =CelltypeOrder)
# change the order
InfectedMat_D7<- InfectedMat_D7[ ,CelltypeOrder]
InfectedMat_D7
# for final celltype matrix: 
InfectedMat_D7CellType<- InfectedMat_D7[,!colnames(InfectedMat_D7) %in% c("max")]
InfectedMat_D7CellTypeNozero <- InfectedMat_D7CellType[,colSums(InfectedMat_D7CellType)>0]


# correlation
InfectedMat_D7CellTypeCorrelationMatrix<-plotCorrelationMatrix(InfectedMat_D7CellTypeNozero)
InfectedMat_D7CellTypeInteractions<- plotInteractions(InfectedMat_D7CellTypeNozero, "heatmap") + scale_fill_gradientn(colors = c("blue", "white", "red"))

ggsave("SpatialScatterPieDay7_CorrelationMatrix.pdf", plot =InfectedMat_D7CellTypeCorrelationMatrix, height = 6, width = 6 )

ggsave("SpatialScatterPieDay7_HeatmapColocalization.pdf", plot =InfectedMat_D7CellTypeInteractions, height = 6, width = 6 )
# scatterpie plot
# scatterpie plot
# 
ct <- colnames(InfectedMat_D7CellType)
ct
# we did not add the low proportion in the scatterpie
InfectedMat_D7CellType[InfectedMat_D7CellType < 0.01] <- 0
paletteMartin <- c(
    "#b6dbff", # HeaPT-S1
    "#99CCFF", # HeaPT-S2
    "#0099FF", # HeaPT-S3
    "#6699FF", # HeaPT-S4
    "#006ddb", # HeaPT-S5
    "#66CC99", # DegPT
    "#EE4E34", # MalaPT
    "#F5DEB3", # infiPT
    "#BEE5BF", # HybridPT
    "#FFD700", # CNT
    "#bdbdbd", # DTL
    "#b66dff", #TAL
    "#FF9900", #DCT-TAL
    "#F96167", #IC
    "#CCF381", #PC
    "#ff6db6", #URO
    "#C5DECD", #ENO
    "#FFCC00", #FIB
    "#F4FEC1", #PL
    "#CCFF33", #ResIMM
    "#99FF66", # RecMAC
    "#A1E8CC", # M1-M2
    "#FF00FF", #NEU
    "#DFF3E3", # T
    "#E2D1F9", #B
    "#CC0033" #NK
  )
#DefaultAssay(Pyelonephritis_InfectionDay28) <- "Spatial"

# choose three color
# pal<-colorRampPalette(c("LightBlue","white","FireBrick"))(length(ct))
# choose the selected color
pal <- colorRampPalette(paletteMartin)(length(ct))

# choose the spectral range colors:
# https://earlglynn.github.io/RNotes/package/RColorBrewer/index.html
#pal <- (colorRampPalette(brewer.pal(11,"Paired"))(length(ct)))

names(pal) <- ct
#pal
rownames(Pyelonephritis_InfectionDay28)
colnames(Pyelonephritis_InfectionDay28)
str(Pyelonephritis_InfectionDay28)
#SPOTlight:::. extract_image(Pyelonephritis_Infection )
#rownames(Pyelonephritis_Infection)
SpatialScatterPieDay7 <-plotSpatialScatterpie(
  x= Pyelonephritis_InfectionDay28,
  y= InfectedMat_D7CellType,
  cell_types = colnames(InfectedMat_D7CellType),
    img = TRUE,
    scatterpie_alpha = 1,
    pie_scale = 0.4) +
    scale_fill_manual(
        values = pal,
        breaks = names(pal)) +coord_sf(xlim = c(100, 500), ylim = c(100, 500))

ggsave("SpatialScatterPieDay7.pdf",plot = SpatialScatterPieDay7, height = 4,width = 5)
# colnames(DeconvolutedMatrix)
# ncol(Pyelonephritis_Infection)
# nrow(DeconvolutedMatrix)
InfectedMat_D7CellType
pdf(file = "SpatialScatterPieDay7_CellTypeNetwork.pdf", height = 8,width = 8)
plotInteractions(InfectedMat_D7CellType, "network")
dev.off()

Idents(Pyelonephritis_InfectionDay28)

# generate the Dotplot for the biomarkers
DefaultAssay(Pyelonephritis_InfectionDay28) <-"SCT"
Finalcelltypemarkers<- c("Lrp2","Slc34a1","Slc13a3", # Proximal tubule cells (PT)
                      "Slc12a3","Pvalb","Wnk1",# Distal connecting tublue (DCT)

                     "Umod","Slc12a1","Cldn10", # Thick ascending limb cells (TAL)
                      "Klk1","Slc8a1","Calb1", # Connecting tubule (CNT)
                     "Slc14a2","Fst","Bst1", # descending thin  limb cells (DTL)
                     #"Akr1b1","Sh3gl3","Prox1", # ascending thin limb cells (ATL)
                     #"Enox1","Thsd4", #Macula densa
                      "Aqp2","Hsd11b2","Scnn1g", # principal cells (PC)
                    "Atp6v1g3","Aqp6","Slc26a7", # Intercalated cells
                    "Krt5","Upk1b","Krt14", # Urothelium
                    "Cdh5","Igfbp3","Pecam1", # Endothelium
                    "Col3a1","Fbn1","Lum","Col1a1", # Fibroblast
                    "Ireb2","Alas2","C78334", # Plasmacytoid dentritic cells
                    "S100a8","S100a9","Il1b", # Neutrophil
                    "C1qa","C1qb","Aif1", # Macrophage
                    "Cxcr6","Cd247","Cd3e", # T cells
                    "Igkc","Cd79a","Cd79b", # B cells
                    "Ccl5","Nkg7","Cd7" # NK/Cd8 cells
)

SpatialD7MarkersDotPlot<- DotPlot(Pyelonephritis_InfectionDay28, features = Finalcelltypemarkers,cols = "Spectral", dot.scale = 4) +RotatedAxis() +theme(legend.text = element_text(size = 12), legend.title = element_text(size = 8))

ggsave("SpatialD7MarkersDotPlot.pdf", plot = SpatialD7MarkersDotPlot, height = 6, width = 14)
Pyelonephritis_InfectionDay28

for (i in Finalcelltypemarkers){
   CellTypeMarkerSpatialFeature<- SpatialFeaturePlot(Pyelonephritis_InfectionDay28, features = i, pt.size.factor = 1.6, crop = TRUE) 
   #&
       #   ggplot2::scale_fill_gradientn(colours = myPalette(4), limits=c(0,1))
  
  ggsave(filename = paste0(i,"_SpatialFeaturePrediction_day7.pdf"), plot = CellTypeMarkerSpatialFeature, height = 5, width = 5)
  
}

Pdate ="PyeloD7_1"
pdf(paste0(Pdate,"_MeanProp_tidyHeatmap.pdf"),height = 5, width = 6)
  
   MeanPropForHeatmap <-FinalIntergratedTable %>% filter(orig.ident==Pdate) %>% dplyr::select(-c("Row.names",'orig.ident',"max")) %>% group_by(Annotation) %>% summarise_all("mean") %>% remove_rownames %>% column_to_rownames(var="Annotation") %>% as.matrix()
   # reorder the matrix
   MeanPropForHeatmapReordered <- MeanPropForHeatmap[,CelltypeOrder]
   MeanPropForHeatmapReordered
   # eliminate the annotation with zero sum
   MeanPropForHeatmapNoZeroHeatmap <-MeanPropForHeatmapReordered[,colSums(MeanPropForHeatmapReordered)>0]%>% Heatmap(cluster_rows =F, cluster_columns = F, show_row_dend=F, show_column_dend = F,column_names_rot = 45,name = "mean",col= col_fun, border = T)  
  print(MeanPropForHeatmapNoZeroHeatmap)
  dev.off()

```

# we generate pie plot for day7
```{r}
# set up the day 7
setwd(InfCPiedir)
Pyelonephritis_Infection$orig.ident
Pyelonephritis_InfectionDay28 <-
    subset (Pyelonephritis_Infection, orig.ident == "PyeloD7_1")
Pyelonephritis_InfectionDay28
  # there seems to be an error in the susetting, so we removed other igmages
  #names(Pyelonephritis_InfectionSubslides@images)
SelectedImageName <- names(Pyelonephritis_InfectionDay28@images)[5]
SelectedImageName
OtherImages <- Allimagmes[Allimagmes != SelectedImageName]
OtherImages
Pyelonephritis_InfectionDay28@images[OtherImages] = NULL

# we here only select the list at day 7 (slide 8)

Pyelonephritis_InfectionDay28 <-
    subset (Pyelonephritis_Infection, orig.ident == "PyeloD7_1")
Pyelonephritis_InfectionDay28
  # there seems to be an error in the susetting, so we removed other igmages
  #names(Pyelonephritis_InfectionSubslides@images)
SelectedImageName <- names(Pyelonephritis_InfectionDay28@images)[5]
 
  OtherImages <- Allimagmes[Allimagmes != SelectedImageName]
  
Pyelonephritis_InfectionDay28@images[OtherImages] = NULL

#nrow(Pyelonephritis_InfectionDay28@meta.data )
SpatialDimPlot(Pyelonephritis_InfectionDay28)

# the prediction of the object

Pyelonephritis_InfectionDay28[["predictions"]]
InfectedMat_D7<- t(Pyelonephritis_InfectionDay28[["predictions"]]@data)
InfectedMat_D7
# for final celltype matrix: 
InfectedMat_D7CellType<- InfectedMat_D7[,!colnames(InfectedMat_D7) %in% c("max")]

# correlation
plotCorrelationMatrix(InfectedMat_D7CellType)
plotInteractions(InfectedMat_D7CellType, "heatmap")

# scatterpie plot
# 
ct <- colnames(InfectedMat_D7CellType)
ct
# we did not add the low proportion in the scatterpie
InfectedMat_D7CellType[InfectedMat_D7CellType < 0.1] <- 0
# paletteMartin <- c(
#     "#000000", "#004949", "#009292", "#ff6db6", "#ffb6db", 
#     "#490092", "#006ddb", "#b66dff", "#6db6ff", "#b6dbff", 
#     "#920000", "#924900", "#db6d00", "#24ff24", "#ffff6d")

paletteMartin <- c(
    "#b6dbff", # HeaPT-S1
    "#99CCFF", # HeaPT-S2
    "#0099FF", # HeaPT-S3
    "#6699FF", # HeaPT-S4
    "#006ddb", # HeaPT-S5
    "#66CC99", # DegPT
    "#EE4E34", # MalaPT
    "#F5DEB3", # infiPT
    "#BEE5BF", # HybridPT
    "#FFD700", # CNT
    "#bdbdbd", # DTL
    "#b66dff", #TAL
    "#FF9900", #DCT-TAL
    "#F96167", #IC
    "#CCF381", #PC
    "#ff6db6", #URO
    "#C5DECD", #ENO
    "#FFCC00", #FIB
    "#F4FEC1", #PL
    "#CCFF33", #ResIMM
    "#99FF66", # RecMAC
    "#A1E8CC", # M1-M2
    "#FF00FF", #NEU
    "#DFF3E3", # T
    "#E2D1F9", #B
    "#CC0033" #NK
  )
#DefaultAssay(Pyelonephritis_InfectionDay28) <- "Spatial"

pal <- colorRampPalette(paletteMartin)(length(ct))
names(pal) <- ct
#pal
rownames(Pyelonephritis_InfectionDay28)
colnames(Pyelonephritis_InfectionDay28)
str(Pyelonephritis_InfectionDay28)
#SPOTlight:::. extract_image(Pyelonephritis_Infection )
#rownames(Pyelonephritis_Infection)
plotSpatialScatterpie(
  x= Pyelonephritis_InfectionDay28,
  y= InfectedMat_D7CellType,
  cell_types = colnames(InfectedMat_D7CellType),
    img = TRUE,
    scatterpie_alpha = 1,
    pie_scale = 0.4) +
    scale_fill_manual(
        values = pal,
        breaks = names(pal))
# colnames(DeconvolutedMatrix)
# ncol(Pyelonephritis_Infection)
# nrow(DeconvolutedMatrix)

plotInteractions(InfectedMat_D7CellType, "network")

# we only focused on the specific clusters
Idents(Pyelonephritis_InfectionDay28)
```

# ###################################
# For interesting infection regions
# ####################################

# #######
# For interesting infection regions : for the abscess of the CorMed
# we generate pie plot for day7 under infection regions for the CorMed abscess regions
```{r}
setwd(InfCorMeddir)
Pyelonephritis_Infection$orig.ident
Pyelonephritis_InfectionDay28 <-
    subset (Pyelonephritis_Infection, orig.ident == "PyeloD7_1")
Pyelonephritis_InfectionDay28
  # there seems to be an error in the susetting, so we removed other igmages
  #names(Pyelonephritis_InfectionSubslides@images)
SelectedImageName <- names(Pyelonephritis_InfectionDay28@images)[5]
SelectedImageName
OtherImages <- Allimagmes[Allimagmes != SelectedImageName]
OtherImages
Pyelonephritis_InfectionDay28@images[OtherImages] = NULL

# for a specific regions and specific clusters
# get the coordate of these cluster
coords <- GetTissueCoordinates(Pyelonephritis_InfectionDay28)
coords
Pyelonephritis_InfectionDay28@meta.data %>% head()
summary(coords)
Pyelonephritis_InfectionDay28@images$slice8

# we select the regions for the cluster
Pyelonephritis_InfectionDay28 <- subset (Pyelonephritis_InfectionDay28, slice8_imagerow > 140 | slice8_imagecol > 200, invert = TRUE)
Pyelonephritis_InfectionDay28 <- subset (Pyelonephritis_InfectionDay28, idents = c("N13","N14","N15","N16","N17"))
InfectedCorMedRegionD1Spatial <-SpatialDimPlot(Pyelonephritis_InfectionDay28,crop = TRUE, label = FALSE)
InfectedCorMedRegionD1Spatial
ggsave("Spatial_infected_day7_SpatialDimplot.pdf",
      plot = InfectedCorMedRegionD1Spatial,
      height = 3,
      width = 4.5 )

Pyelonephritis_InfectionDay28@meta.data[,c("orig.ident","Annotation")]
## merge is way too troublesome, therefore
FinalIntergratedTableAbcess<- merge(Pyelonephritis_InfectionDay28@meta.data[,c("orig.ident","Annotation")], t(Pyelonephritis_InfectionDay28[["predictions"]]@data), all=TRUE, by="row.names") 
?Heatmap
col_fun <- circlize::colorRamp2(c(0, 0.01, 1), c("white", "blue", "red"))

# generate the mean value for each column
pdf(file = "Spatial_infected_day7_FinalIntergratedTableAbcessHeatmap.pdf", height = 4, width = 4)
FinalIntergratedTableAbcessForHeatmap <-FinalIntergratedTableAbcess %>% dplyr::select(-c("Row.names",'orig.ident',"max")) %>% group_by(Annotation) %>% summarise_all("mean") %>% remove_rownames %>% column_to_rownames(var="Annotation") %>% as.matrix()

FinalIntergratedTableAbcessHeatmap<-FinalIntergratedTableAbcessForHeatmap[,colSums(FinalIntergratedTableAbcessForHeatmap)>0] %>% Heatmap(show_row_dend=T, show_column_dend = F,column_names_rot = 45,name = "Mean",col= col_fun, border = T)  

print(FinalIntergratedTableAbcessHeatmap)
dev.off()
#FinalIntergratedTableAbcess %>% group_by(Annotation) %>% summarize(the_medians = median(as.numeric(value), na.rm = TRUE))
#colorRamp2(c(0, 0.2, 1), c("gray", "blue", "red"))
```


```{r}
# generate the spatial pie plot for this region
ct <- colnames(InfectedMat_D7CellType)
ct
# we did not add the low proportion in the scatterpie
InfectedMat_D7CellType[InfectedMat_D7CellType < 0.1] <- 0
# paletteMartin <- c(
#     "#000000", "#004949", "#009292", "#ff6db6", "#ffb6db", 
#     "#490092", "#006ddb", "#b66dff", "#6db6ff", "#b6dbff", 
#     "#920000", "#924900", "#db6d00", "#24ff24", "#ffff6d")

paletteMartin <- c(
    "#b6dbff", # HeaPT-S1
    "#99CCFF", # HeaPT-S2
    "#0099FF", # HeaPT-S3
    "#6699FF", # HeaPT-S4
    "#006ddb", # HeaPT-S5
    "#66CC99", # DegPT
    "#EE4E34", # MalaPT
    "#F5DEB3", # infiPT
    "#BEE5BF", # HybridPT
    "#FFD700", # CNT
    "#bdbdbd", # DTL
    "#b66dff", #TAL
    "#FF9900", #DCT-TAL
    "#F96167", #IC
    "#CCF381", #PC
    "#ff6db6", #URO
    "#C5DECD", #ENO
    "#FFCC00", #FIB
    "#F4FEC1", #PL
    "#CCFF33", #ResIMM
    "#99FF66", # RecMAC
    "#A1E8CC", # M1-M2
    "#FF00FF", #NEU
    "#DFF3E3", # T
    "#E2D1F9", #B
    "#CC0033" #NK
  )
#DefaultAssay(Pyelonephritis_InfectionDay28) <- "Spatial"

pal <- colorRampPalette(paletteMartin)(length(ct))
names(pal) <- ct
Pyelonephritis_InfectionDay28CellTypeMatrix <-t(Pyelonephritis_InfectionDay28[["predictions"]]@data)

Pyelonephritis_InfectionDay28CellTypeMatrixNoMax <- Pyelonephritis_InfectionDay28CellTypeMatrix[,!colnames(Pyelonephritis_InfectionDay28CellTypeMatrix) %in% c("max")]
Pyelonephritis_InfectionDay28CellTypeMatrixNoMax
# generate the scatter pie plot
colnames(Pyelonephritis_InfectionDay28CellTypeMatrixNoMax)
#D0subclusterCellTypeMatrixNoMax[D0subclusterCellTypeMatrixNoMax < 0.1] <- 0
InfectedCorMedRegionD1SpatialPie<- plotSpatialScatterpie(
  x= Pyelonephritis_InfectionDay28,
  y= Pyelonephritis_InfectionDay28CellTypeMatrixNoMax,
  cell_types = colnames(Pyelonephritis_InfectionDay28CellTypeMatrixNoMax),
    img = TRUE,
    scatterpie_alpha = 1,
    pie_scale = 0.7,axis ="v") +
    scale_fill_manual(
        values = pal,
        breaks = names(pal)) +
  coord_sf(xlim = c(150, 300), ylim = c(300, 450))
?plotSpatialScatterpie
ggsave("Spatial_infected_day7_SpatialCellType_Scatterpie.pdf",
      plot = InfectedCorMedRegionD1SpatialPie,
      height = 3,
      width = 4.5 )

# correlation
InfectedCorMedRegionD1SpatialPieCorrelationMatrix <- plotCorrelationMatrix(Pyelonephritis_InfectionDay28CellTypeMatrixNoMax)
ggsave("Spatial_infected_day7_InfectedCorMedRegionD1SpatialPieCorrelationMatrix.pdf",
      plot = InfectedCorMedRegionD1SpatialPieCorrelationMatrix,
      height = 4.5,
      width = 4.5 )
# we only select the matrix that rowsum >0
Pyelonephritis_InfectionDay28CellTypeMatrixNoMaxLarger<- Pyelonephritis_InfectionDay28CellTypeMatrixNoMax[,colSums(Pyelonephritis_InfectionDay28CellTypeMatrixNoMax[])>0]
InfectedPevRegionD1Interactions<- plotInteractions(Pyelonephritis_InfectionDay28CellTypeMatrixNoMaxLarger, "heatmap") + scale_fill_gradientn(colors = c("blue", "white", "red"))
?plotInteractions
ggsave("Spatial_infected_day7_InfectedPevRegionD1Interactions.pdf",
      plot = InfectedPevRegionD1Interactions,
      height = 4.5,
      width = 4.5 )
# 
# Test<-D0subclusterCellTypeMatrixNoMax[rowSums(D0subclusterCellTypeMatrixNoMax[])>0,]
# Test2<-Test[colSums(Test[])>0,]
# Test2
# plotCorrelationMatrix(Test2)
# plotInteractions(Test2, "heatmap")

```




# we generate pie plot for day7 under infection regions for the CorMed abscess regions
```{r}
setwd(InfCPiedir)
Pyelonephritis_Infection$orig.ident
Pyelonephritis_InfectionDay28 <-
    subset (Pyelonephritis_Infection, orig.ident == "PyeloD7_1")
Pyelonephritis_InfectionDay28
  # there seems to be an error in the susetting, so we removed other igmages
  #names(Pyelonephritis_InfectionSubslides@images)
SelectedImageName <- names(Pyelonephritis_InfectionDay28@images)[5]
SelectedImageName
OtherImages <- Allimagmes[Allimagmes != SelectedImageName]
OtherImages
Pyelonephritis_InfectionDay28@images[OtherImages] = NULL
Pyelonephritis_InfectionDay28@meta.data %>% head()
# for a specific regions and specific clusters
# get the coordate of these cluster
coords <- GetTissueCoordinates(Pyelonephritis_InfectionDay28)
coords %>% head()
Pyelonephritis_InfectionDay28@meta.data %>% head()
summary(coords)
Pyelonephritis_InfectionDay28@images$slice8

# we select the regions for the cluster : , invert = TRUE
Pyelonephritis_InfectionDay28 <- subset (Pyelonephritis_InfectionDay28, (slice8_imagerow >310 & slice8_imagerow <450 )  & (slice8_imagecol < 330 & slice8_imagecol> 220) )
Pyelonephritis_InfectionDay28 <- subset (Pyelonephritis_InfectionDay28, idents = c("N8","N10","N11","N12","N15"))
InfectedCorMedRegionD1Spatial <-SpatialDimPlot(Pyelonephritis_InfectionDay28,label = TRUE)
InfectedCorMedRegionD1Spatial
ggsave("Spatial_infected_day7_SpatialDimplot.pdf",
      plot = InfectedCorMedRegionD1Spatial,
      height = 3,
      width = 4.5 )
```

# we generated a specific marker genes 
```{r}
library(magrittr)
library(dplyr)
#install.packages("remotes")
remotes::install_github("jbergenstrahle/STUtility",force = T)
library(STutility)
?
?FindAllMarkers
InfectedCorMedRegionD1Spatial
Pyelonephritis_InfectionDay28
DefaultAssay(Pyelonephritis_InfectionDay28) <- "integrated"
nbs_2.markers <- FindAllMarkers(Pyelonephritis_InfectionDay28,test.use ="MAST")
nbs_2.markers
??SubsetSTData
#nbs_2.markers$gene <- rownames(nbs_2.markers)

nbs_2.markers
#se.subset <- SubsetSTData(Pyelonephritis_InfectionDay28, expression = nbs_2 %in% c("N12", "N15"))
sorted.marks <- nbs_2.markers %>% group_by(cluster) %>% arrange(-avg_log2FC) %>% top_n(n = 10, wt = abs(avg_log2FC)) 
sorted.marks

# Here we only selected several marker genes: 
DotPlot(Pyelonephritis_InfectionDay28, features = sorted.marks$gene %>% unique()) +RotatedAxis()
#DoHeatmap(Pyelonephritis_InfectionDay28, features = sorted.marks$gene, group.colors = c("red", "lightgray"))

Pyelonephritis_InfectionDay28@meta.data[,c("orig.ident","Annotation")]
## merge is way too troublesome, therefore
FinalIntergratedTableAbcess<- merge(Pyelonephritis_InfectionDay28@meta.data[,c("orig.ident","Annotation")], t(Pyelonephritis_InfectionDay28[["predictions"]]@data), all=TRUE, by="row.names") 
?Heatmap
col_fun <- circlize::colorRamp2(c(0, 0.01, 1), c("white", "blue", "red"))

# generate the mean value for each column
pdf(file = "Spatial_infected_day7_FinalIntergratedTableAbcessHeatmap.pdf", height = 4, width = 4)
FinalIntergratedTableAbcessForHeatmap <-FinalIntergratedTableAbcess %>% dplyr::select(-c("Row.names",'orig.ident',"max")) %>% group_by(Annotation) %>% summarise_all("mean") %>% remove_rownames %>% column_to_rownames(var="Annotation") %>% as.matrix()

FinalIntergratedTableAbcessHeatmap<-FinalIntergratedTableAbcessForHeatmap[,colSums(FinalIntergratedTableAbcessForHeatmap)>0] %>% Heatmap(show_row_dend=T, show_column_dend = F,column_names_rot = 45,name = "Mean",col= col_fun, border = T)  

print(FinalIntergratedTableAbcessHeatmap)
dev.off()
#FinalIntergratedTableAbcess %>% group_by(Annotation) %>% summarize(the_medians = median(as.numeric(value), na.rm = TRUE))
#colorRamp2(c(0, 0.2, 1), c("gray", "blue", "red"))
```

# generate the spatial pie plot for this region
```{r}

# scatterpie plot
# 
ct <- colnames(InfectedMat_D7CellType)
ct
# we did not add the low proportion in the scatterpie
InfectedMat_D7CellType[InfectedMat_D7CellType < 0.1] <- 0
# paletteMartin <- c(
#     "#000000", "#004949", "#009292", "#ff6db6", "#ffb6db", 
#     "#490092", "#006ddb", "#b66dff", "#6db6ff", "#b6dbff", 
#     "#920000", "#924900", "#db6d00", "#24ff24", "#ffff6d")

paletteMartin <- c(
    "#b6dbff", # HeaPT-S1
    "#99CCFF", # HeaPT-S2
    "#0099FF", # HeaPT-S3
    "#6699FF", # HeaPT-S4
    "#006ddb", # HeaPT-S5
    "#66CC99", # DegPT
    "#EE4E34", # MalaPT
    "#F5DEB3", # infiPT
    "#BEE5BF", # HybridPT
    "#FFD700", # CNT
    "#bdbdbd", # DTL
    "#b66dff", #TAL
    "#FF9900", #DCT-TAL
    "#F96167", #IC
    "#CCF381", #PC
    "#ff6db6", #URO
    "#C5DECD", #ENO
    "#FFCC00", #FIB
    "#F4FEC1", #PL
    "#CCFF33", #ResIMM
    "#99FF66", # RecMAC
    "#A1E8CC", # M1-M2
    "#FF00FF", #NEU
    "#DFF3E3", # T
    "#E2D1F9", #B
    "#CC0033" #NK
  )
#DefaultAssay(Pyelonephritis_InfectionDay28) <- "Spatial"

pal <- colorRampPalette(paletteMartin)(length(ct))
names(pal) <- ct
#pal
Pyelonephritis_InfectionDay28CellTypeMatrix <-t(Pyelonephritis_InfectionDay28[["predictions"]]@data)

Pyelonephritis_InfectionDay28CellTypeMatrixNoMax <- Pyelonephritis_InfectionDay28CellTypeMatrix[,!colnames(Pyelonephritis_InfectionDay28CellTypeMatrix) %in% c("max")]
# generate the scatter pie plot
colnames(Pyelonephritis_InfectionDay28CellTypeMatrixNoMax)
#D0subclusterCellTypeMatrixNoMax[D0subclusterCellTypeMatrixNoMax < 0.1] <- 0
InfectedCorMedRegionD1SpatialPie<- plotSpatialScatterpie(
  x= Pyelonephritis_InfectionDay28,
  y= Pyelonephritis_InfectionDay28CellTypeMatrixNoMax,
  cell_types = colnames(Pyelonephritis_InfectionDay28CellTypeMatrixNoMax),
    img = TRUE,
    scatterpie_alpha = 1,
    pie_scale = 0.7,axis ="v") +
    scale_fill_manual(
        values = pal,
        breaks = names(pal)) +
  coord_sf(xlim = c(150, 300), ylim = c(300, 450))
?plotSpatialScatterpie
ggsave("Spatial_infected_day7_SpatialCellType_Scatterpie.pdf",
      plot = InfectedCorMedRegionD1SpatialPie,
      height = 3,
      width = 4.5 )

# correlation
InfectedCorMedRegionD1SpatialPieCorrelationMatrix <- plotCorrelationMatrix(Pyelonephritis_InfectionDay28CellTypeMatrixNoMax)
ggsave("Spatial_infected_day7_InfectedCorMedRegionD1SpatialPieCorrelationMatrix.pdf",
      plot = InfectedCorMedRegionD1SpatialPieCorrelationMatrix,
      height = 4.5,
      width = 4.5 )
# we only select the matrix that rowsum >0
Pyelonephritis_InfectionDay28CellTypeMatrixNoMaxLarger<- Pyelonephritis_InfectionDay28CellTypeMatrixNoMax[,colSums(Pyelonephritis_InfectionDay28CellTypeMatrixNoMax[])>0]
InfectedPevRegionD1Interactions<- plotInteractions(Pyelonephritis_InfectionDay28CellTypeMatrixNoMaxLarger, "heatmap") + scale_fill_gradientn(colors = c("blue", "white", "red"))
?plotInteractions
ggsave("Spatial_infected_day7_InfectedPevRegionD1Interactions.pdf",
      plot = InfectedPevRegionD1Interactions,
      height = 4.5,
      width = 4.5 )
# 
# Test<-D0subclusterCellTypeMatrixNoMax[rowSums(D0subclusterCellTypeMatrixNoMax[])>0,]
# Test2<-Test[colSums(Test[])>0,]
# Test2
# plotCorrelationMatrix(Test2)
# plotInteractions(Test2, "heatmap")

```





# ###############################################################
# unfinished section:
# ###############################################################



```{r}
SpatialDimPlot(Pyelonephritis_InfectionD0)

DefaultAssay(Pyelonephritis_Infection)  <- "SCT"
Pyelonephritis_InfectionInfection <- subset(Pyelonephritis_Infection, idents = c("10_1"))

Pyelonephritis_InfectionD0 <- subset(Pyelonephritis_Infection, orig.ident == c("PyeloD0_1"))
Pyelonephritis_InfectionD0
?Seurat::subset
DefaultAssay(Pyelonephritis_InfectionD0)  <- "Spatial"

SpatialDimPlot(Pyelonephritis_InfectionD0, images = "slice1")

# plot the 


InfectedMat_10_1<- t(Pyelonephritis_InfectionInfection[["predictions"]]@data)
# we only select the slide 7 from the predictionPyelonephritis_Infection
#InfectedMat[,!colnames(mat) %in% c("max")]
#subset(Pyelonephritis_Infection, subset = images == "slice1")
InfectedMat_10_1[,!colnames(mat) %in% c("max")]

plotCorrelationMatrix(InfectedMat_10_1[,!colnames(mat) %in% c("max")])
plotInteractions(InfectedMat_10_1[,!colnames(mat) %in% c("max")], "heatmap")

Pyelonephritis_InfectionInfection_5_1 <- subset(Pyelonephritis_Infection, idents = c("5_1"))

InfectedMat_5_1<- t(Pyelonephritis_InfectionInfection_5_1[["predictions"]]@data)
# we only select the slide 7 from the predictionPyelonephritis_Infection
#InfectedMat[,!colnames(mat) %in% c("max")]
#subset(Pyelonephritis_Infection, subset = images == "slice1")
plotCorrelationMatrix(InfectedMat_5_1[,!colnames(mat) %in% c("max")])
plotInteractions(InfectedMat_5_1[,!colnames(mat) %in% c("max")], "heatmap")

# decovonvoluted matrix
mat<- t(Pyelonephritis_Infection[["predictions"]]@data)

colnames(mat)
# remove the max 
DeconvolutedMatrix <- mat[,!colnames(mat) %in% c("max")]
plotCorrelationMatrix(DeconvolutedMatrix)

plotInteractions(DeconvolutedMatrix, "heatmap")
plotInteractions(DeconvolutedMatrix, "network")

GetTissueCoordinates(Pyelonephritis_Infection)
#Pyelonephritis_Infection
length(ct)
# Scatterpie


ct <- colnames(DeconvolutedMatrix)
ct
# we did not add the low proportion in the scatterpie
DeconvolutedMatrix[DeconvolutedMatrix < 0.1] <- 0
paletteMartin <- c(
    "#000000", "#004949", "#009292", "#ff6db6", "#ffb6db", 
    "#490092", "#006ddb", "#b66dff", "#6db6ff", "#b6dbff", 
    "#920000", "#924900", "#db6d00", "#24ff24", "#ffff6d")

DefaultAssay(Pyelonephritis_Infection) <- "Spatial"

pal <- colorRampPalette(paletteMartin)(length(ct))
names(pal) <- ct
#pal
rownames(Pyelonephritis_Infection)
colnames(Pyelonephritis_Infection)
str(Pyelonephritis_Infection)
#SPOTlight:::. extract_image(Pyelonephritis_Infection )

nrow(Pyelonephritis_Infection)
nrow(DeconvolutedMatrix)
#rownames(Pyelonephritis_Infection)
plotSpatialScatterpie(
  x= Pyelonephritis_Infection,
  y= DeconvolutedMatrix,
  cell_types = colnames(DeconvolutedMatrix),
    img = TRUE,
    scatterpie_alpha = 1,
    pie_scale = 0.4) +
    scale_fill_manual(
        values = pal,
        breaks = names(pal))
# colnames(DeconvolutedMatrix)
# ncol(Pyelonephritis_Infection)
# nrow(DeconvolutedMatrix)
```

 ```{r}
# GetAssayData(Pyelonephritis_Infection, assay = "SCT" )
# # scale the object into 0 to 1
# 
# library(magrittr)
# rownames(Pyelonephritis_Infection)
# rownames(Pyelonephritis_Infection) %>% seq()
# #rownames(Pyelonephritis_Infection[["SCT"]]) %>% seq()
# Pyelonephritis_Infection_scaledMatrix<- sapply(rownames(Pyelonephritis_Infection) %>% seq(),  
#        function (i) { 
#            scales::rescale(GetAssayData(Pyelonephritis_Infection, 
#                               assay = "predictions")[rownames(Pyelonephritis_Infection)[i],], to = c(0,1)) }
#   ) %>% t
# 
# Pyelonephritis_Infection_scaledMatrix[1:10,10:20]
# 
# # check expression ranges
# quantile(Pyelonephritis_Infection_scaledMatrix, c(0.01, 1))
# 
# # add gene names
# dimnames(Pyelonephritis_Infection_scaledMatrix)[[1]] <- rownames(Pyelonephritis_Infection)
# 
# # set new scaled matrix to the slt 
# Pyelonephritis_Infection <- SetAssayData(Pyelonephritis_Infection, slot = "scale.data", Pyelonephritis_Infection_scaledMatrix)
# 
# Pyelonephritis_Infection[["predictions"]]
# 
# ?SpatialFeaturePlot
```

```{r}
# create a list to store the objects of subslide
Pyelonephritis_InfectionSubslides <- list()

i=0
for (SubSlide in  AllIdentities) {
  i=i+1
  # select the subslide object
  Pyelonephritis_InfectionSubslides[[i]] <-
    subset (Pyelonephritis_Infection, orig.ident == SubSlide)

  # there seems to be an error in the susetting, so we removed other igmages
  #names(Pyelonephritis_InfectionSubslides@images)
  SelectedImageName <- names(Pyelonephritis_InfectionSubslides[[i]]@images)[i]
 # SelectedImageName
  OtherImages <- Allimagmes[Allimagmes != SelectedImageName]
  
  Pyelonephritis_InfectionSubslides[[i]]@images[OtherImages] = NULL
   
}

```



```{r}
# measure the different frequency of each cluster

######################################################################
# ### measure the proportion of each spots of cell types
### this stored in the matrix of Pyelonephritis.integrated[["predictions"]]@data
#################################################################
# DefaultAssay(Pyelonephritis_Infection) <-"predictions"
# the matrix for the predicted cell types and spot id
row.names(Pyelonephritis_Infection[["predictions"]]@data)

colnames(Pyelonephritis_Infection[["predictions"]]@data)

#Pyelonephritis.integrated[["predictions"]]@data[, ]
################
# the matrix for the cluster of the spot ID
Pyelonephritis_Infection@meta.data

# cluster ID Pyelonephritis.integrated$seurat_clusters

## merge the Pyelonephritis.integrated@meta.data and Pyelonephritis.integrated[["predictions"]]@data

t(Pyelonephritis_Infection[["predictions"]]@data)

Pyelonephritis_Infection@meta.data[c("AAACCGTTCGTCCAGG-1_1"),]
row.names(t(Pyelonephritis_Infection[["predictions"]]@data))
Pyelonephritis_Infection@meta.data %>% head()

Table1 <- Pyelonephritis_Infection@meta.data[,c("orig.ident","Annotation","ConditionGroup")]

## merge is way too troublesome, therefore
FinalIntergratedTable<- merge(Pyelonephritis_Infection@meta.data[,c("orig.ident","Annotation","ConditionGroup")], t(Pyelonephritis_Infection[["predictions"]]@data), all=TRUE, by="row.names")

FinalIntergratedTable
colnames(FinalIntergratedTable)

# cluster 5_1
FinalIntergratedTable %>% filter(Annotation == "5_1")%>% slice_head(n=10)
# cluster 10_1
FinalIntergratedTable %>% filter(Annotation == "10_1" ) %>% slice_head(n=20)
# cluster 5_0
FinalIntergratedTable %>% filter(Annotation == "5_0")%>% slice_head(n=10) 
#%>% dplyr::select("FIB","M1-M2" ,"END","NEU","RecMAC", "T", "HybridPT", "B","TAL","InfiPT","ResIMM")
# cluster 10_0
FinalIntergratedTable %>% filter(Annotation == "10_0" ) %>% slice_head(n=20)
```

# generare the deconvolution using RCTD
```{r}
library(spacexr)
# set up reference
ref <- readRDS(file = paste0(Scindir, "PIPSeq_ImmuneSub_Annotated_v0823.RDS"))
ref <- UpdateSeuratObject(ref)

ref@meta.data %>% head()
Idents(ref) <- "ImmuneSub"


# extract information to pass to the RCTD Reference function
counts <- as.data.frame(ref[["RNA"]]@counts)
counts[1:4,1:4]
cluster <- as.factor(ref$ImmuneSub)
names(cluster) <- colnames(ref)

nUMI <- ref$nCount_RNA

names(nUMI) <- colnames(ref)
reference <- spacexr::Reference(counts, cluster, nUMI)

str(reference)
# set up query with the RCTD function SpatialRNA
#Pyelonephritis_Infection<-readRDS(file = paste0(Spindir, "FinalAbcessClustersObj.rds"))

counts <- as.data.frame(Pyelonephritis_Infection[["Spatial"]]@counts)
counts[2,1:4]
coords <- GetTissueCoordinates(Pyelonephritis_Infection)
coords
colnames(coords) <- c("x", "y")
coords[is.na(colnames(coords))] <- NULL
#total counts per pixel is nUMI
query <- spacexr::SpatialRNA(coords, counts, colSums(counts))
str(query)

```

# Using the reference and query object, we annotate the dataset and add the cell type labels to the query Seurat object. RCTD parallelizes well, so multiple cores can be specified for faster performance.

```{r}
# Optional: spacexr runs faster with more cores, but you may need to adjust system environment settings
Sys.setenv("OPENBLAS_NUM_THREADS"=2)

RCTD <- spacexr::create.RCTD(query, reference, max_cores = 2)
RCTD <- spacexr::run.RCTD(RCTD, doublet_mode = "full")
slide.seq <- AddMetaData(slide.seq, metadata = RCTD@results$results_df)
# Next, plot the RCTD annotations. Because we ran RCTD in doublet mode, the algorithm assigns a first_type and second_type for each barcode or spot.

p1 <- SpatialDimPlot(slide.seq, group.by = "first_type")
p2 <- SpatialDimPlot(slide.seq, group.by = "second_type")
p1 | p2


```




# generate the scatterpie from the SPOTlight
```{r}
ct <- colnames(FinalIntergratedTable)

FinalIntergratedTable[FinalIntergratedTable <0.1] <-0
paletteMartin <- c(
    "#000000", "#004949", "#009292", "#ff6db6", "#ffb6db", 
    "#490092", "#006ddb", "#b66dff", "#6db6ff", "#b6dbff", 
    "#920000", "#924900", "#db6d00", "#24ff24", "#ffff6d")

pal <- colorRampPalette(paletteMartin)(length(ct))
names(pal) <- ct
library(SPOTlight)

plotSpatialScatterpie(
    x = Pyelonephritis_Infection,
    y = FinalIntergratedTable,
    cell_types = colnames(y),
    img = FALSE,
    scatterpie_alpha = 1,
    pie_scale = 0.4) +
    scale_fill_manual(
        values = pal,
        breaks = names(pal))

```


```{r}
##########################################################################################
# The heat map figure to show the different compostion within timepoints in the cluster spots
###########################################################################################
### We combined all the dataset without group by orig.ident and measure the different composition

ClusterAllCluster_Median<-FinalIntergratedTable %>% group_by(Annotation) %>% summarise(ProximaltubuleI=median(`Proximal tubule I`),  ProximaltubuleII=median(`Proximal tubule II`), ProximaltubuleIII=median(`Proximal tubule III`), ProximaltubuleIV= median(`Proximal tubule IV`), ProximaltubuleVI=median(`Proximal tubule VI`), Distaltubule=median(`Distal tubule`),  ThinlimbofLOH=median(`Thin limb of LOH`),Urothelium= median(`Urothelium`),   ThickascendinglimbofLOH=median(`Thick ascending limb of LOH`), 
 Intercalatedcells=median(`Intercalated cells`), Principalcells=median(`Principal cells`), 
                                                                                                        Fibroblast=median(`Fibroblast`), Endothelium=median(`Endothelium`), Myeloblast=median(`Myeloblast`), Tcells=median(`T cells`),  Bcells=median(`B cells`),  Plasmacytoiddendriticcells=median(`Plasmacytoid dendritic cells`),
                                                                                                           NKCells=median(`NK cells`),  .groups= 'drop') %>% as.data.frame()



ClusterAllCluster_Median

write.csv(ClusterAllCluster_Median, file = "ClusterAllCluster_Median.csv")

### we calculate the median value for each cluster and also based on the timepotins
### we used the median value to  scale the cell-type compositions within each niche. 
ClusterTimepointGroup_Median<-FinalIntergratedTable %>% group_by(orig.ident, Annotation) %>% summarise(ProximaltubuleI=median(`Proximal tubule I`),  ProximaltubuleII=median(`Proximal tubule II`), ProximaltubuleIII=median(`Proximal tubule III`), ProximaltubuleIV= median(`Proximal tubule IV`), ProximaltubuleVI=median(`Proximal tubule VI`), Distaltubule=median(`Distal tubule`),  ThinlimbofLOH=median(`Thin limb of LOH`), ThickascendinglimbofLOH=median(`Thick ascending limb of LOH`), 
 Intercalatedcells=median(`Intercalated cells`), Principalcells=median(`Principal cells`),  Urothelium= median(`Urothelium`),                                                                                                        Fibroblast=median(`Fibroblast`), Endothelium=median(`Endothelium`), Myeloblast=median(`Myeloblast`), Tcells=median(`T cells`),  Bcells=median(`B cells`),  Plasmacytoiddendriticcells=median(`Plasmacytoid dendritic cells`),
                                                                                                           NKCells=median(`NK cells`),  .groups= 'drop')  %>% as.data.frame()
ClusterTimepointGroup_Median
write.csv(ClusterTimepointGroup_Median, file = "ClusterTimepointGroup_Median.csv")

### draw the heatmap for these combined clusters

Combined_Median_tidy <- ClusterAllCluster_Median %>% mutate_at( vars(-Annotation),scale) %>% pivot_longer(cols = -c(Annotation), names_to = "Property", values_to = "Value") 
Combined_Median_tidy
names(Combined_Median_tidy)
### reorder the levels
Combined_Median_tidy$Property<- ordered(factor(Combined_Median_tidy$Property),
                                levels=c("ProximaltubuleI","ProximaltubuleII",
                                         "ProximaltubuleIII", "ProximaltubuleIV", 
                                         "ProximaltubuleVI","Principalcells", "Intercalatedcells",  "ThickascendinglimbofLOH", "ThinlimbofLOH",  "Urothelium","Distaltubule",  "Fibroblast", "Endothelium","Myeloblast" ,  "Tcells" ,"Bcells" , "NKCells"))

pdf("Combined_Median_tidy_Heatmap.pdf",height = 6, width = 8)
Combined_Median_tidyHeatmap<- na.omit(Combined_Median_tidy) %>% heatmap(Property, Annotation, Value, scale="row",cluster_rows = FALSE, cluster_columns = TRUE ,palette_value=circlize::colorRamp2(seq(-4, 4, length.out = 11), RColorBrewer::brewer.pal(11, "RdBu") ) )
Combined_Median_tidyHeatmap
dev.off()


## setting up the conditions 
Condition<- levels(ClusterTimepointGroup_Median$orig.ident)
## run the for liips to generate the heatmap figures

### generate multiple heatmap plots



#pdf("Combined_heatmaps_Seurat_0320.pdf", height = 4, width = 40)

## attention par() and layout() not work for ggplot and 
# par(mfrow=c(11,1))
## instead we should use grid.arrange()
## grid.arrange(plot1, plot2, ncol=2)
ClusterTimepointGroup_Median
## create a new empty list and new number
i=0
my_vect <- c()
for (Pdate in Condition){
  ### generate the transfered objects for each condition
  ### 
 i =i+1
  Pdate_Median_tidy<- ClusterTimepointGroup_Median %>% filter(orig.ident==Pdate)  %>% mutate_at(vars (-orig.ident, -Annotation),scale) %>% pivot_longer(cols = -c(orig.ident, Annotation), names_to = "Property", values_to ="Value") 
  
  ### change the levels of Property at each condition
  ### tranfer the ClusterTimepointGroup_Median into a tide "element-feature-independent variables"  data frame, where the independent variable
  ### similar function to melt, but this will transfer the objects
  levels(factor(Pdate_Median_tidy$Property))
  Pdate_Median_tidy$Property<- ordered(factor(Pdate_Median_tidy$Property), 
                                       levels=c("ProximaltubuleI","ProximaltubuleII",
                                         "ProximaltubuleIII", "ProximaltubuleIV", 
                                         "ProximaltubuleVI","Principalcells", "Intercalatedcells",  "ThickascendinglimbofLOH", "ThinlimbofLOH",  "Urothelium","Distaltubule",  "Fibroblast", "Endothelium","Myeloblast" ,  "Tcells" ,"Bcells" , "NKCells"))
  
  ## change the NA value to Zero
  #PyeloD0_1_Median_tidy$Value[is.nan(PyeloD0_1_Median_tidy$Value)]<-0
  # ignore the na.omit
  na.omit(Pdate_Median_tidy)
  
 pdf(paste0(Pdate,"_Median_tidyHeatmap.pdf"),height = 6, width = 8)
  #PyeloD3_1_Median_tidyHeatmap<- na.omit(PyeloD3_1_Median_tidy) %>% heatmap(Property, seurat_clusters, Value, scale="row",cluster_rows = FALSE, cluster_columns = TRUE ,palette_value = c( "#E6E6FA", "#FFFFFF","#F1948A")) 
  
 Pdate_Median_tidy
  Pdate_Median_tidyHeatmap<- na.omit(Pdate_Median_tidy) %>% heatmap(Property, Annotation, Value, scale="row",cluster_rows = FALSE, cluster_columns = FALSE ,palette_value=circlize::colorRamp2(seq(-4, 4, length.out = 11), RColorBrewer::brewer.pal(11, "RdBu") ) )
  
  #PyeloD3_1_Median_tidyHeatmap<- na.omit(PyeloD3_1_Median_tidy) %>% heatmap(Property, seurat_clusters, Value, scale="row",cluster_rows = FALSE, cluster_columns = FALSE , palette_value = c("blue","white","red"), color_legend_min=-4, color_legend_max=4)
  
  #my_vect<- append(my_vect,Pdate_Median_tidyHeatmap)
  #plots[[i]]<- Pdate_Median_tidyHeatmap
  print(Pdate_Median_tidyHeatmap)
  dev.off()
}


###############################################################################################
# spearman correlation of the celltype matrix 
###############################################################################################
ClusterTimepointGroup_Median

for (Pdate in Condition){
  ### generate the transfered objects for each condition
  ### 
  #i =i+1
  #Pdate=c("PyeloD0_1")
  Pdate_Median_tidy_matrix <-ClusterTimepointGroup_Median %>% filter(orig.ident==Pdate) %>% dplyr::select(-c('orig.ident','Annotation'))
  
  #Pdate_Median_tidy_matrix
  #Pdate_Median_tidy_matrix<- ClusterTimepointGroup_Median %>% filter(orig.ident==Pdate) 
  
  #%>% mutate_at(vars (-orig.ident, -seurat_clusters),scale) %>% pivot_longer(cols = -c(orig.ident, seurat_clusters), names_to = "Property", values_to ="Value") 
  
 
  Pdate_Median_tidy_matrix_cor<-cor(as.matrix(Pdate_Median_tidy_matrix),method = "spearman")
  
  ## replace the NA into zerro
  Pdate_Median_tidy_matrix_cor <- replace(Pdate_Median_tidy_matrix_cor, is.na(Pdate_Median_tidy_matrix_cor),0)
  
  #testRes = cor.mtest(BiomarkerExpCor, conf.level = 0.95)
  Pdate_Median_tidy_matrix_cor_melt = melt(Pdate_Median_tidy_matrix_cor)
  
  ## create a dataframe
  
  #Pdate_Median_tidy_matrix_cor_melt$name= rep(Pdate, length(Pdate_Median_tidy_matrix_cor_melt$Var1))
 
  ## merge different dataframe
  
  pdf(paste0(Pdate,"_Correlation_tidyHeatmap.pdf"),height = 6, width = 8)
  #pdf("BioMarkerExpCorrelation.pdf", width = 10, height =10)
  ## generate the coeffecient figure
  corrplot(Pdate_Median_tidy_matrix_cor, order = 'hclust', type='upper',tl.col = 'black', insig='blank',
           addCoef.col ='black', number.cex = 0.5, tl.cex= 0.7, tl.srt = 45, diag=FALSE,col = colorRampPalette(c("LightBlue","white","FireBrick"))(100))
  dev.off()
  
  ## other way to plot
  # Pdate_Median_tidy_matrix_cor = rcorr(as.matrix(Pdate_Median_tidy_matrix), type='spearman')
  # 
  # Pdate_Median_tidy_matrix_cor_melt = melt(Pdate_Median_tidy_matrix_cor$r)
  # 
  # 
  # pdf(paste0(Pdate,"_Correlation_tidyHeatmap.pdf"),height = 6, width = 8)
  # 
  # #txtsize <- par('din')[2] / 2
  # ggplot(Pdate_Median_tidy_matrix_cor_melt, aes(x=Var1, y=Var2, fill=value)) + geom_tile() + 
  #   theme(axis.text.x = element_text(angle=90, hjust=TRUE))
  # dev.off()
   #+
   # xlab("") + ylab("") + 
   # geom_text(label=Pdate_Median_tidy_matrix_cor_melt$label, size=txtsize) + 
   # geom_text(label=Pdate_Median_tidy_matrix_cor_melt$strike, size=txtsize * 4, color="red", alpha=0.4) 

 
}


```

# redefine the default assay by the SCT using the scale data
```{r}
Pyelonephritis_Infection[["LOL"]] <-
  CreateAssayObject(counts = GetAssayData(Pyelonephritis_Infection,
                                          assay =
                                            "SCT", slot = "data")) #Copy "SCT data" to "LOL counts" and "LOL data"
Pyelonephritis_Infection <-
  SetAssayData(
    Pyelonephritis_Infection,
    assay = "LOL",
    slot = "scale.data",
    new.data = GetAssayData(Pyelonephritis_Infection, assay =
                              "SCT",
                            slot =
                              "scale.data")
  )
gc() #"Copy "SCT scale.data" to "LOL scale.data"
DefaultAssay(Pyelonephritis_Infection) <- "LOL"
Pyelonephritis_Infection <-
  DietSeurat(
    Pyelonephritis_Infection,
    assays = "LOL",
    counts = T,
    data = T,
    scale.data = T
  )
Pyelonephritis_Infection <-
  RenameAssays(Pyelonephritis_Infection, LOL = 'SCT')
## set up default of FinalAbcessClustersObjSubSlides

DefaultAssay(Pyelonephritis_Infection) <- "SCT"
 # save the RDS of Pyelonephritis_Infection
 #Pyelonephritis_Infection
Pyelonephritis_Infection
 saveRDS(filename = "Pyelonephritis_Infection_WithCellType.RDS",Pyelonephritis_Infection)
```


